// @kubb/plugin-client/dist/clients/fetch.js
//#region src/clients/fetch.ts
let _config = {};
const getConfig = () => _config;
const setConfig = (config) => {
	_config = config;
	return getConfig();
};
const client = async (paramsConfig) => {
	const normalizedParams = new URLSearchParams();
	const config = {
		...getConfig(),
		...paramsConfig
	};
	Object.entries(config.params || {}).forEach(([key, value]) => {
		if (value !== void 0) normalizedParams.append(key, value === null ? "null" : value.toString());
	});
	let targetUrl = [config.baseURL, config.url].filter(Boolean).join("");
	if (config.params) targetUrl += `?${normalizedParams}`;
	const response = await fetch(targetUrl, {
		credentials: config.credentials || "same-origin",
		method: config.method?.toUpperCase(),
		body: JSON.stringify(config.data),
		signal: config.signal,
		headers: config.headers
	});
	return {
		data: [
			204,
			205,
			304
		].includes(response.status) || !response.body ? {} : await response.json(),
		status: response.status,
		statusText: response.statusText,
		headers: response.headers
	};
};
client.getConfig = getConfig;
client.setConfig = setConfig;
var fetch_default = client;

//#endregion
export { client, fetch_default as default, getConfig, setConfig };
