// @kubb/react-fabric/dist/index.cjs
const require_react$1 = require('./react-Bq0UOw6S.cjs');
require('./globals-8sJ940pg.cjs');
const require_Runtime = require('./Runtime-JI640uK4.cjs');
const require_jsx_runtime$1 = require('./jsx-runtime-d2dmYscH.cjs');
let dedent = require("dedent");
dedent = require_react$1.__toESM(dedent);
let __kubb_fabric_core = require("@kubb/fabric-core");
__kubb_fabric_core = require_react$1.__toESM(__kubb_fabric_core);
let natural_orderby = require("natural-orderby");
natural_orderby = require_react$1.__toESM(natural_orderby);

//#region src/components/App.tsx
var import_react = /* @__PURE__ */ require_react$1.__toESM(require_react$1.require_react());
var import_jsx_runtime = /* @__PURE__ */ require_react$1.__toESM(require_jsx_runtime$1.require_jsx_runtime());
const AppContext = (0, import_react.createContext)(void 0);
function App({ meta, children }) {
	const { exit } = (0, import_react.useContext)(require_Runtime.RootContext);
	return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(AppContext.Provider, {
		value: {
			exit,
			meta
		},
		children
	});
}
App.Context = AppContext;
App.displayName = "KubbApp";

//#endregion
//#region src/utils/createJSDoc.ts
function createJSDoc({ comments }) {
	const filteredComments = comments.filter(Boolean);
	if (!filteredComments.length) return "";
	return `/**\n * ${filteredComments.join("\n * ")}\n */`;
}

//#endregion
//#region src/components/Const.tsx
function Const({ name, export: canExport, type, JSDoc, asConst, children }) {
	return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
		(JSDoc === null || JSDoc === void 0 ? void 0 : JSDoc.comments) && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [createJSDoc({ comments: JSDoc === null || JSDoc === void 0 ? void 0 : JSDoc.comments }), /* @__PURE__ */ (0, import_jsx_runtime.jsx)("br", {})] }),
		canExport && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: "export " }),
		"const ",
		name,
		" ",
		type && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
			":",
			type,
			" "
		] }),
		"= ",
		children,
		asConst && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: " as const" })
	] });
}
Const.displayName = "KubbConst";

//#endregion
//#region src/components/File.tsx
const FileContext = (0, import_react.createContext)({});
function File({ children,...rest }) {
	if (!rest.baseName || !rest.path) return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children });
	return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("kubb-file", {
		...rest,
		children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(FileContext.Provider, {
			value: {
				baseName: rest.baseName,
				path: rest.path,
				meta: rest.meta
			},
			children
		})
	});
}
File.displayName = "KubbFile";
function FileSource({ isTypeOnly, name, isExportable, isIndexable, children }) {
	return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("kubb-source", {
		name,
		isTypeOnly,
		isExportable,
		isIndexable,
		children
	});
}
FileSource.displayName = "KubbFileSource";
function FileExport({ name, path, isTypeOnly, asAlias }) {
	return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("kubb-export", {
		name,
		path,
		isTypeOnly: isTypeOnly || false,
		asAlias
	});
}
FileExport.displayName = "KubbFileExport";
function FileImport({ name, root, path, isTypeOnly, isNameSpace }) {
	return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("kubb-import", {
		name,
		root,
		path,
		isNameSpace,
		isTypeOnly: isTypeOnly || false
	});
}
FileImport.displayName = "KubbFileImport";
File.Export = FileExport;
File.Import = FileImport;
File.Source = FileSource;
File.Context = FileContext;

//#endregion
//#region ../../node_modules/.pnpm/indent-string@5.0.0/node_modules/indent-string/index.js
function indentString(string, count = 1, options = {}) {
	const { indent = " ", includeEmptyLines = false } = options;
	if (typeof string !== "string") throw new TypeError(`Expected \`input\` to be a \`string\`, got \`${typeof string}\``);
	if (typeof count !== "number") throw new TypeError(`Expected \`count\` to be a \`number\`, got \`${typeof count}\``);
	if (count < 0) throw new RangeError(`Expected \`count\` to be at least 0, got \`${count}\``);
	if (typeof indent !== "string") throw new TypeError(`Expected \`options.indent\` to be a \`string\`, got \`${typeof indent}\``);
	if (count === 0) return string;
	const regex = includeEmptyLines ? /^/gm : /^(?!\s*$)/gm;
	return string.replace(regex, indent.repeat(count));
}

//#endregion
//#region src/components/Indent.tsx
/**
* Indents all children by `size` spaces.
* Collapses consecutive <br /> tags to at most 2.
*/
function Indent({ size = 2, children }) {
	if (!children) return null;
	const filtered = import_react.Children.toArray(children).filter(Boolean);
	const result = [];
	let prevWasBr = false;
	let brCount = 0;
	filtered.forEach((child) => {
		if (import_react.isValidElement(child) && child.type === "br") {
			if (!prevWasBr || brCount < 2) {
				result.push(child);
				brCount++;
			}
			prevWasBr = true;
		} else {
			prevWasBr = false;
			brCount = 0;
			result.push(child);
		}
	});
	return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: result.map((child) => {
		if (typeof child === "string") return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: indentString((0, dedent.default)(child), size) });
		return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [" ".repeat(size), child] });
	}) });
}

//#endregion
//#region src/components/Function.tsx
function Function({ name, default: isDefault, export: canExport, async, generics, params, returnType, JSDoc, children }) {
	return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
		(JSDoc === null || JSDoc === void 0 ? void 0 : JSDoc.comments) && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [createJSDoc({ comments: JSDoc === null || JSDoc === void 0 ? void 0 : JSDoc.comments }), /* @__PURE__ */ (0, import_jsx_runtime.jsx)("br", {})] }),
		canExport && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: "export " }),
		isDefault && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: "default " }),
		async && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: "async " }),
		"function ",
		name,
		generics && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
			"<",
			Array.isArray(generics) ? generics.join(", ").trim() : generics,
			">"
		] }),
		"(",
		params,
		")",
		returnType && !async && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [": ", returnType] }),
		returnType && async && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
			": Promise",
			"<",
			returnType,
			">"
		] }),
		" {",
		/* @__PURE__ */ (0, import_jsx_runtime.jsx)("br", {}),
		/* @__PURE__ */ (0, import_jsx_runtime.jsx)(Indent, {
			size: 2,
			children
		}),
		/* @__PURE__ */ (0, import_jsx_runtime.jsx)("br", {}),
		"}"
	] });
}
Function.displayName = "KubbFunction";
function ArrowFunction({ name, default: isDefault, export: canExport, async, generics, params, returnType, JSDoc, singleLine, children }) {
	return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
		(JSDoc === null || JSDoc === void 0 ? void 0 : JSDoc.comments) && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [createJSDoc({ comments: JSDoc === null || JSDoc === void 0 ? void 0 : JSDoc.comments }), /* @__PURE__ */ (0, import_jsx_runtime.jsx)("br", {})] }),
		canExport && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: "export " }),
		isDefault && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: "default " }),
		"const ",
		name,
		" = ",
		async && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: "async " }),
		generics && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
			"<",
			Array.isArray(generics) ? generics.join(", ").trim() : generics,
			">"
		] }),
		"(",
		params,
		")",
		returnType && !async && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [": ", returnType] }),
		returnType && async && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
			": Promise",
			"<",
			returnType,
			">"
		] }),
		singleLine && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
			" => ",
			children,
			/* @__PURE__ */ (0, import_jsx_runtime.jsx)("br", {})
		] }),
		!singleLine && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
			" => {",
			/* @__PURE__ */ (0, import_jsx_runtime.jsx)("br", {}),
			/* @__PURE__ */ (0, import_jsx_runtime.jsx)(Indent, {
				size: 2,
				children
			}),
			/* @__PURE__ */ (0, import_jsx_runtime.jsx)("br", {}),
			"}",
			/* @__PURE__ */ (0, import_jsx_runtime.jsx)("br", {})
		] })
	] });
}
ArrowFunction.displayName = "KubbArrowFunction";
Function.Arrow = ArrowFunction;

//#endregion
//#region src/components/Type.tsx
function Type({ name, export: canExport, JSDoc, children }) {
	if (name.charAt(0).toUpperCase() !== name.charAt(0)) throw new Error("Name should start with a capital letter(see TypeScript types)");
	return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
		(JSDoc === null || JSDoc === void 0 ? void 0 : JSDoc.comments) && /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [createJSDoc({ comments: JSDoc === null || JSDoc === void 0 ? void 0 : JSDoc.comments }), /* @__PURE__ */ (0, import_jsx_runtime.jsx)("br", {})] }),
		canExport && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: "export " }),
		"type ",
		name,
		" = ",
		children
	] });
}
Type.displayName = "KubbType";

//#endregion
//#region src/composables/useApp.ts
/**
* `useApp` will return the current App with plugin, pluginManager, fileManager and mode.
*/
function useApp() {
	const app = (0, import_react.useContext)(App.Context);
	if (!app) throw new Error("<App /> should be set");
	return app;
}

//#endregion
//#region src/composables/useFile.ts
/**
* `useFile` will return the current file when <File/> is used.
*/
function useFile() {
	return (0, import_react.useContext)(File.Context);
}

//#endregion
//#region src/composables/useLifecycle.tsx
/**
* `useLifecycle` will return some helpers to exit/restart the generation.
*/
function useLifecycle() {
	const { exit } = (0, import_react.useContext)(require_Runtime.Root.Context);
	return { exit: () => {
		setTimeout(() => {
			exit();
		}, 0);
	} };
}

//#endregion
//#region src/utils/getFunctionParams.ts
function order(items) {
	return (0, natural_orderby.orderBy)(items.filter(Boolean), [([_key, item]) => {
		if (item === null || item === void 0 ? void 0 : item.children) return;
		return !(item === null || item === void 0 ? void 0 : item.default);
	}, ([_key, item]) => {
		if (item === null || item === void 0 ? void 0 : item.children) return;
		return !(item === null || item === void 0 ? void 0 : item.optional);
	}], ["desc", "desc"]);
}
function parseChild(key, item, options) {
	const entries = order(Object.entries(item.children));
	const types = [];
	const names = [];
	const optional = entries.every(([_key, item$1]) => item$1 === null || item$1 === void 0 ? void 0 : item$1.optional);
	entries.forEach(([key$1, entryItem]) => {
		if (entryItem) {
			const name$1 = parseItem(key$1, {
				...entryItem,
				type: void 0
			}, options);
			if (entryItem.children) {
				const subTypes = Object.entries(entryItem.children).map(([key$2]) => {
					return key$2;
				}).join(", ");
				if (subTypes) names.push(`${name$1}: { ${subTypes} }`);
				else names.push(name$1);
			} else if (options.type === "call" && options.transformName) names.push(`${key$1}: ${name$1}`);
			else names.push(name$1);
			if (entries.some(([_key, item$1]) => item$1 === null || item$1 === void 0 ? void 0 : item$1.type)) types.push(parseItem(key$1, {
				...entryItem,
				default: void 0
			}, options));
		}
	});
	const name = item.mode === "inline" ? key : names.length ? `{ ${names.join(", ")} }` : void 0;
	const type = item.type ? item.type : types.length ? `{ ${types.join("; ")} }` : void 0;
	if (!name) return null;
	return parseItem(name, {
		type,
		default: item.default,
		optional: !item.default ? optional : void 0
	}, options);
}
function parseItem(name, item, options) {
	const acc = [];
	const transformedName = options.transformName ? options.transformName(name) : name;
	const transformedType = options.transformType && item.type ? options.transformType(item.type) : item.type;
	if (options.type === "object") return transformedName;
	if (options.type === "objectValue") return item.value ? `${transformedName}: ${item.value}` : transformedName;
	if (item.type && options.type === "constructor") if (item.optional) acc.push(`${transformedName}?: ${transformedType}`);
	else acc.push(`${transformedName}: ${transformedType}${item.default ? ` = ${item.default}` : ""}`);
	else if (item.default && options.type === "constructor") acc.push(`${transformedName} = ${item.default}`);
	else if (item.value) acc.push(`${transformedName} : ${item.value}`);
	else if (item.mode === "inlineSpread") acc.push(`... ${transformedName}`);
	else acc.push(transformedName);
	return acc[0];
}
function getFunctionParams(params, options) {
	return order(Object.entries(params)).reduce((acc, [key, item]) => {
		if (!item) return acc;
		if (item.children) {
			if (Object.keys(item.children).length === 0) return acc;
			if (item.mode === "inlineSpread") return [...acc, getFunctionParams(item.children, options)];
			const parsedItem$1 = parseChild(key, item, options);
			if (!parsedItem$1) return acc;
			return [...acc, parsedItem$1];
		}
		const parsedItem = parseItem(key, item, options);
		return [...acc, parsedItem];
	}, []).join(", ");
}
function createFunctionParams(params) {
	return params;
}
var _params = /* @__PURE__ */ new WeakMap();
var FunctionParams = class FunctionParams {
	static factory(params) {
		return new FunctionParams(params);
	}
	constructor(params) {
		require_Runtime._classPrivateFieldInitSpec(this, _params, void 0);
		require_Runtime._classPrivateFieldSet2(_params, this, params);
	}
	get params() {
		return require_Runtime._classPrivateFieldGet2(_params, this);
	}
	get flatParams() {
		const flatter = (acc, [key, item]) => {
			if (item === null || item === void 0 ? void 0 : item.children) return Object.entries(item.children).reduce(flatter, acc);
			if (item) acc[key] = item;
			return acc;
		};
		return Object.entries(require_Runtime._classPrivateFieldGet2(_params, this)).reduce(flatter, {});
	}
	toCall({ transformName, transformType } = {}) {
		return getFunctionParams(require_Runtime._classPrivateFieldGet2(_params, this), {
			type: "call",
			transformName,
			transformType
		});
	}
	toObject() {
		return getFunctionParams(require_Runtime._classPrivateFieldGet2(_params, this), { type: "object" });
	}
	toObjectValue() {
		return getFunctionParams(require_Runtime._classPrivateFieldGet2(_params, this), { type: "objectValue" });
	}
	toConstructor() {
		return getFunctionParams(require_Runtime._classPrivateFieldGet2(_params, this), { type: "constructor" });
	}
};

//#endregion
//#region src/index.ts
const createContext = import_react.createContext;
const createElement = import_react.createElement;
const useContext = import_react.useContext;
const useEffect = import_react.useEffect;
const useState = import_react.useState;
const useRef = import_react.useRef;
const use = import_react.use;
const useReducer = import_react.useReducer;

//#endregion
exports.App = App;
exports.Const = Const;
exports.File = File;
exports.Function = Function;
exports.FunctionParams = FunctionParams;
exports.Indent = Indent;
exports.Runtime = require_Runtime.Runtime;
exports.Type = Type;
Object.defineProperty(exports, 'createApp', {
  enumerable: true,
  get: function () {
    return __kubb_fabric_core.createApp;
  }
});
exports.createContext = createContext;
exports.createElement = createElement;
exports.createFunctionParams = createFunctionParams;
exports.use = use;
exports.useApp = useApp;
exports.useContext = useContext;
exports.useEffect = useEffect;
exports.useFile = useFile;
exports.useLifecycle = useLifecycle;
exports.useReducer = useReducer;
exports.useRef = useRef;
exports.useState = useState;
