// @kubb/react-fabric/dist/plugins.cjs
const require_react$1 = require('./react-Bq0UOw6S.cjs');
const require_Runtime = require('./Runtime-JI640uK4.cjs');
require('./jsx-runtime-d2dmYscH.cjs');
let __kubb_fabric_core_plugins = require("@kubb/fabric-core/plugins");
__kubb_fabric_core_plugins = require_react$1.__toESM(__kubb_fabric_core_plugins);

//#region src/plugins/reactPlugin.ts
var import_react = /* @__PURE__ */ require_react$1.__toESM(require_react$1.require_react(), 1);
const reactPlugin = (0, __kubb_fabric_core_plugins.createPlugin)({
	name: "react",
	install() {},
	inject(app, options = {}) {
		const runtime = new require_Runtime.Runtime({
			fileManager: app.context.fileManager,
			...options
		});
		return {
			async render(App) {
				runtime.render((0, import_react.createElement)(App));
				await app.context.events.emit("start");
			},
			async renderToString(App) {
				return runtime.renderToString((0, import_react.createElement)(App));
			},
			async waitUntilExit() {
				await runtime.waitUntilExit();
				await app.context.events.emit("end");
			}
		};
	}
});

//#endregion
exports.reactPlugin = reactPlugin;
Object.keys(__kubb_fabric_core_plugins).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return __kubb_fabric_core_plugins[k]; }
  });
});
