// @kubb/fabric-core/dist/index.js
import { a as _classPrivateFieldInitSpec, n as createFile, r as _classPrivateFieldGet2, t as _defineProperty } from "./defineProperty-BtekiGIK.js";
import { t as trimExtName } from "./trimExtName-CeOVQVbu.js";
import "./createParser-17uGjfu3.js";
import { t as defaultParser } from "./defaultParser--HzU9LPa.js";
import { orderBy } from "natural-orderby";
import pLimit from "p-limit";
import { EventEmitter } from "node:events";

//#region src/utils/Cache.ts
var _buffer = /* @__PURE__ */ new WeakMap();
var Cache = class {
	constructor() {
		_classPrivateFieldInitSpec(this, _buffer, /* @__PURE__ */ new Map());
	}
	get(key) {
		var _this$buffer$get;
		return (_this$buffer$get = _classPrivateFieldGet2(_buffer, this).get(key)) !== null && _this$buffer$get !== void 0 ? _this$buffer$get : null;
	}
	set(key, value) {
		_classPrivateFieldGet2(_buffer, this).set(key, value);
	}
	delete(key) {
		_classPrivateFieldGet2(_buffer, this).delete(key);
	}
	clear() {
		_classPrivateFieldGet2(_buffer, this).clear();
	}
	keys() {
		return [..._classPrivateFieldGet2(_buffer, this).keys()];
	}
	values() {
		return [..._classPrivateFieldGet2(_buffer, this).values()];
	}
	flush() {}
};

//#endregion
//#region ../../node_modules/.pnpm/remeda@2.32.0/node_modules/remeda/dist/isFunction-BJjFuZR7.js
const e = (e$1) => typeof e$1 == `function`;

//#endregion
//#region src/utils/AsyncEventEmitter.ts
var _emitter = /* @__PURE__ */ new WeakMap();
var AsyncEventEmitter = class {
	constructor(maxListener = 100) {
		_classPrivateFieldInitSpec(this, _emitter, new EventEmitter());
		_classPrivateFieldGet2(_emitter, this).setMaxListeners(maxListener);
	}
	async emit(eventName, ...eventArgs) {
		const listeners = _classPrivateFieldGet2(_emitter, this).listeners(eventName);
		if (listeners.length === 0) return;
		await Promise.all(listeners.map(async (listener) => {
			try {
				return await listener(...eventArgs);
			} catch (err) {
				console.error(`Error in async listener for "${eventName}":`, err);
			}
		}));
	}
	on(eventName, handler) {
		_classPrivateFieldGet2(_emitter, this).on(eventName, handler);
	}
	onOnce(eventName, handler) {
		const wrapper = (...args) => {
			this.off(eventName, wrapper);
			handler(...args);
		};
		this.on(eventName, wrapper);
	}
	off(eventName, handler) {
		_classPrivateFieldGet2(_emitter, this).off(eventName, handler);
	}
	removeAll() {
		_classPrivateFieldGet2(_emitter, this).removeAllListeners();
	}
};

//#endregion
//#region src/FileProcessor.ts
var _limit = /* @__PURE__ */ new WeakMap();
var FileProcessor = class {
	constructor({ events = new AsyncEventEmitter() } = {}) {
		_classPrivateFieldInitSpec(this, _limit, pLimit(100));
		_defineProperty(this, "events", void 0);
		this.events = events;
		return this;
	}
	async parse(file, { parsers, extension } = {}) {
		const parseExtName = (extension === null || extension === void 0 ? void 0 : extension[file.extname]) || void 0;
		if (!parsers) {
			console.warn("No parsers provided, using default parser. If you want to use a specific parser, please provide it in the options.");
			return defaultParser.parse(file, { extname: parseExtName });
		}
		if (!file.extname) return defaultParser.parse(file, { extname: parseExtName });
		let parser;
		for (const item of parsers) {
			var _item$extNames;
			if ((_item$extNames = item.extNames) === null || _item$extNames === void 0 ? void 0 : _item$extNames.includes(file.extname)) {
				parser = item;
				break;
			}
		}
		if (!parser) return defaultParser.parse(file, { extname: parseExtName });
		return parser.parse(file, { extname: parseExtName });
	}
	async run(files, { parsers, mode = "sequential", dryRun, extension } = {}) {
		await this.events.emit("process:start", { files });
		let processed = 0;
		const total = files.length;
		const processOne = async (resolvedFile, index) => {
			const percentage = processed / total * 100;
			await this.events.emit("file:start", {
				file: resolvedFile,
				index,
				total
			});
			const source = dryRun ? void 0 : await this.parse(resolvedFile, {
				extension,
				parsers
			});
			await this.events.emit("process:progress", {
				file: resolvedFile,
				source,
				processed,
				percentage,
				total
			});
			processed++;
			await this.events.emit("file:end", {
				file: resolvedFile,
				index,
				total
			});
		};
		if (mode === "sequential") {
			async function* asyncFiles() {
				for (let index = 0; index < files.length; index++) yield [files[index], index];
			}
			for await (const [file, index] of asyncFiles()) if (file) await processOne(file, index);
		} else {
			const promises = files.map((resolvedFile, index) => _classPrivateFieldGet2(_limit, this).call(this, () => processOne(resolvedFile, index)));
			await Promise.all(promises);
		}
		await this.events.emit("process:end", { files });
		return files;
	}
};

//#endregion
//#region src/FileManager.ts
function mergeFile(a, b) {
	return {
		...a,
		sources: [...a.sources || [], ...b.sources || []],
		imports: [...a.imports || [], ...b.imports || []],
		exports: [...a.exports || [], ...b.exports || []]
	};
}
var _cache = /* @__PURE__ */ new WeakMap();
var FileManager = class {
	constructor({ events = new AsyncEventEmitter() } = {}) {
		_classPrivateFieldInitSpec(this, _cache, new Cache());
		_defineProperty(this, "events", void 0);
		_defineProperty(this, "processor", void 0);
		this.processor = new FileProcessor({ events });
		this.events = events;
		return this;
	}
	async add(...files) {
		const resolvedFiles = [];
		const mergedFiles = /* @__PURE__ */ new Map();
		files.forEach((file) => {
			const existing = mergedFiles.get(file.path);
			if (existing) mergedFiles.set(file.path, mergeFile(existing, file));
			else mergedFiles.set(file.path, file);
		});
		for (const file of mergedFiles.values()) {
			const existing = _classPrivateFieldGet2(_cache, this).get(file.path);
			const resolvedFile = createFile(existing ? mergeFile(existing, file) : file);
			_classPrivateFieldGet2(_cache, this).set(resolvedFile.path, resolvedFile);
			this.flush();
			resolvedFiles.push(resolvedFile);
		}
		await this.events.emit("file:add", { files: resolvedFiles });
		return resolvedFiles;
	}
	flush() {
		_classPrivateFieldGet2(_cache, this).flush();
	}
	getByPath(path) {
		return _classPrivateFieldGet2(_cache, this).get(path);
	}
	deleteByPath(path) {
		_classPrivateFieldGet2(_cache, this).delete(path);
	}
	clear() {
		_classPrivateFieldGet2(_cache, this).clear();
	}
	get files() {
		return orderBy(_classPrivateFieldGet2(_cache, this).keys(), [(v) => v.length, (v) => trimExtName(v).endsWith("index")]).map((key) => _classPrivateFieldGet2(_cache, this).get(key)).filter(Boolean);
	}
	async write(options) {
		await this.events.emit("write:start", { files: this.files });
		const resolvedFiles = await this.processor.run(this.files, options);
		this.clear();
		await this.events.emit("write:end", { files: resolvedFiles });
		return resolvedFiles;
	}
};

//#endregion
//#region src/defineApp.ts
function defineApp(instance) {
	function createApp$1(options) {
		const events = new AsyncEventEmitter();
		const installedPlugins = /* @__PURE__ */ new Set();
		const installedParsers = /* @__PURE__ */ new Set();
		const fileManager = new FileManager({ events });
		const app = {
			context: {
				events,
				options,
				fileManager,
				installedPlugins,
				installedParsers
			},
			get files() {
				return fileManager.files;
			},
			async addFile(...newFiles) {
				await fileManager.add(...newFiles);
			},
			async use(pluginOrParser, ...options$1) {
				const args = options$1;
				if (pluginOrParser.type === "plugin") {
					if (installedPlugins.has(pluginOrParser)) console.warn("Plugin has already been applied to target app.");
					else installedPlugins.add(pluginOrParser);
					if (pluginOrParser.inject && e(pluginOrParser.inject)) {
						const injecter = pluginOrParser.inject;
						const extraApp = injecter(app, ...args);
						Object.assign(app, extraApp);
					}
				}
				if (pluginOrParser.type === "parser") if (installedParsers.has(pluginOrParser)) console.warn("Parser has already been applied to target app.");
				else installedParsers.add(pluginOrParser);
				if (pluginOrParser && e(pluginOrParser.install)) {
					const installer = pluginOrParser.install;
					await installer(app, ...args);
				}
				return app;
			}
		};
		if (instance) instance(app);
		return app;
	}
	return createApp$1;
}

//#endregion
//#region src/createApp.ts
const createApp = defineApp();

//#endregion
export { FileManager, FileProcessor, createApp, createFile, defineApp };
