// @kubb/fabric-core/dist/plugins.js
import { a as _classPrivateFieldInitSpec, i as _assertClassBrand, n as createFile, r as _classPrivateFieldGet2, t as _defineProperty } from "./defineProperty-BtekiGIK.js";
import "./trimExtName-CeOVQVbu.js";
import { t as getRelativePath } from "./getRelativePath-CERJmYkp.js";
import path, { relative, resolve } from "node:path";
import { switcher } from "js-runtime";
import fs from "fs-extra";
import { Presets, SingleBar } from "cli-progress";
import process from "node:process";
import http from "node:http";
import open from "tiny-open";
import handler from "serve-handler";

//#region src/plugins/createPlugin.ts
function createPlugin(plugin) {
	return {
		type: "plugin",
		...plugin
	};
}

//#endregion
//#region src/plugins/fsPlugin.ts
async function write(path$1, data, options = {}) {
	return switcher({
		node: async (path$2, data$1, { sanity }) => {
			if (!data$1 || (data$1 === null || data$1 === void 0 ? void 0 : data$1.trim()) === "") return;
			try {
				const oldContent = await fs.readFile(resolve(path$2), { encoding: "utf-8" });
				if ((oldContent === null || oldContent === void 0 ? void 0 : oldContent.toString()) === (data$1 === null || data$1 === void 0 ? void 0 : data$1.toString())) return;
			} catch (_err) {}
			await fs.outputFile(resolve(path$2), data$1.trim(), { encoding: "utf-8" });
			if (sanity) {
				const savedData = await fs.readFile(resolve(path$2), { encoding: "utf-8" });
				if ((savedData === null || savedData === void 0 ? void 0 : savedData.toString()) !== (data$1 === null || data$1 === void 0 ? void 0 : data$1.toString())) throw new Error(`Sanity check failed for ${path$2}\n\nData[${data$1.length}]:\n${data$1}\n\nSaved[${savedData.length}]:\n${savedData}\n`);
				return savedData;
			}
			return data$1;
		},
		bun: async (path$2, data$1, { sanity }) => {
			if (!data$1 || (data$1 === null || data$1 === void 0 ? void 0 : data$1.trim()) === "") return;
			await Bun.write(resolve(path$2), data$1.trim());
			if (sanity) {
				const savedData = await Bun.file(resolve(path$2)).text();
				if ((savedData === null || savedData === void 0 ? void 0 : savedData.toString()) !== (data$1 === null || data$1 === void 0 ? void 0 : data$1.toString())) throw new Error(`Sanity check failed for ${path$2}\n\nData[${data$1.length}]:\n${data$1}\n\nSaved[${savedData.length}]:\n${savedData}\n`);
				return savedData;
			}
			return data$1;
		}
	}, "node")(path$1, data, options);
}
const fsPlugin = createPlugin({
	name: "fs",
	install(app, options = {}) {
		if (options.clean) fs.removeSync(options.clean.path);
		app.context.events.on("process:progress", async ({ file, source }) => {
			if (options.onBeforeWrite) await options.onBeforeWrite(file.path, source);
			await write(file.path, source, { sanity: false });
		});
	},
	inject(app, { dryRun } = {}) {
		return { async write(options = { extension: { ".ts": ".ts" } }) {
			var _app$context$options;
			await app.context.fileManager.write({
				mode: (_app$context$options = app.context.options) === null || _app$context$options === void 0 ? void 0 : _app$context$options.mode,
				extension: options.extension,
				dryRun,
				parsers: app.context.installedParsers
			});
		} };
	}
});

//#endregion
//#region \0@oxc-project+runtime@0.95.0/helpers/classPrivateFieldSet2.js
function _classPrivateFieldSet2(s, a, r) {
	return s.set(_assertClassBrand(s, a), r), r;
}

//#endregion
//#region src/utils/TreeNode.ts
var _cachedLeaves = /* @__PURE__ */ new WeakMap();
var TreeNode = class TreeNode {
	constructor(data, parent) {
		_defineProperty(this, "data", void 0);
		_defineProperty(this, "parent", void 0);
		_defineProperty(this, "children", []);
		_classPrivateFieldInitSpec(this, _cachedLeaves, void 0);
		this.data = data;
		this.parent = parent;
	}
	addChild(data) {
		const child = new TreeNode(data, this);
		this.children.push(child);
		_classPrivateFieldSet2(_cachedLeaves, this, void 0);
		return child;
	}
	get leaves() {
		if (_classPrivateFieldGet2(_cachedLeaves, this)) return _classPrivateFieldGet2(_cachedLeaves, this);
		if (this.children.length === 0) return [this];
		const stack = [...this.children];
		const result = [];
		for (const node of stack) if (node.children.length) for (const child of node.children) stack.push(child);
		else result.push(node);
		_classPrivateFieldSet2(_cachedLeaves, this, result);
		return result;
	}
	forEach(callback) {
		const stack = [this];
		for (const node of stack) {
			callback(node);
			if (node.children.length) for (const child of node.children) stack.push(child);
		}
		return this;
	}
	findDeep(predicate) {
		for (const leaf of this.leaves) if (predicate(leaf)) return leaf;
	}
	static toGraph(root) {
		const nodes = [];
		const edges = [];
		root.forEach((node) => {
			nodes.push({
				id: node.data.path,
				label: node.data.name
			});
			for (const child of node.children) edges.push({
				from: node.data.path,
				to: child.data.path
			});
		});
		return {
			nodes,
			edges
		};
	}
	static fromFiles(files, rootFolder = "") {
		const normalizePath = (p) => p.replace(/\\/g, "/");
		const normalizedRoot = normalizePath(rootFolder);
		const rootPrefix = normalizedRoot.endsWith("/") ? normalizedRoot : `${normalizedRoot}/`;
		const filteredFiles = files.filter((file) => {
			const filePath = normalizePath(file.path);
			return !filePath.endsWith(".json") && (!rootFolder || filePath.startsWith(rootPrefix));
		});
		if (filteredFiles.length === 0) return null;
		const treeNode = new TreeNode({
			name: rootFolder || "",
			path: rootFolder || "",
			file: void 0
		});
		for (const file of filteredFiles) {
			const parts = normalizePath(file.path).slice(rootPrefix.length).split("/");
			let current = treeNode;
			let currentPath = rootFolder;
			for (const [index, part] of parts.entries()) {
				const isLast = index === parts.length - 1;
				currentPath += (currentPath.endsWith("/") ? "" : "/") + part;
				let next;
				for (const child of current.children) if (child.data.name === part) {
					next = child;
					break;
				}
				if (!next) next = current.addChild({
					name: part,
					path: currentPath,
					file: isLast ? file : void 0
				});
				current = next;
			}
		}
		return treeNode;
	}
};

//#endregion
//#region src/plugins/barrelPlugin.ts
function getBarrelFiles({ files, root, mode }) {
	if (mode === "propagate" || mode === false) return [];
	const cachedFiles = /* @__PURE__ */ new Map();
	const dedupe = /* @__PURE__ */ new Map();
	const treeNode = TreeNode.fromFiles(files, root);
	if (!treeNode) return [];
	treeNode.forEach((node) => {
		var _node$parent;
		if (!(node === null || node === void 0 ? void 0 : node.children) || !((_node$parent = node.parent) === null || _node$parent === void 0 ? void 0 : _node$parent.data.path)) return;
		const parentPath = node.parent.data.path;
		const barrelPath = path.join(parentPath, "index.ts");
		let barrelFile = cachedFiles.get(barrelPath);
		if (!barrelFile) {
			barrelFile = createFile({
				path: barrelPath,
				baseName: "index.ts",
				exports: [],
				sources: []
			});
			cachedFiles.set(barrelPath, barrelFile);
			dedupe.set(barrelPath, /* @__PURE__ */ new Set());
		}
		const seen = dedupe.get(barrelPath);
		node.leaves.forEach((leaf) => {
			const file = leaf.data.file;
			if (!file) return;
			(file.sources || []).forEach((source) => {
				if (!file.path || !source.isIndexable || !source.name) return;
				const key = `${source.name}|${source.isTypeOnly ? "1" : "0"}`;
				if (seen.has(key)) return;
				seen.add(key);
				barrelFile.exports.push({
					name: [source.name],
					path: getRelativePath(parentPath, file.path),
					isTypeOnly: source.isTypeOnly
				});
				barrelFile.sources.push({
					name: source.name,
					isTypeOnly: source.isTypeOnly,
					value: "",
					isExportable: mode === "all" || mode === "named",
					isIndexable: mode === "all" || mode === "named"
				});
			});
		});
	});
	const result = [...cachedFiles.values()];
	if (mode === "all") return result.map((file) => {
		var _file$exports;
		return {
			...file,
			exports: (_file$exports = file.exports) === null || _file$exports === void 0 ? void 0 : _file$exports.map((e) => ({
				...e,
				name: void 0
			}))
		};
	});
	return result;
}
const barrelPlugin = createPlugin({
	name: "barrel",
	install(app, options) {
		if (!options) throw new Error("Barrel plugin requires options.root and options.mode");
		if (!options.mode) return;
		app.context.events.on("write:start", async ({ files }) => {
			const root = options.root;
			const barrelFiles = getBarrelFiles({
				files,
				root,
				mode: options.mode
			});
			await app.context.fileManager.add(...barrelFiles);
		});
	},
	inject(app, options) {
		if (!options) throw new Error("Barrel plugin requires options.root and options.mode");
		return { async writeEntry({ root, mode }) {
			var _app$context$options;
			if (!mode || mode === "propagate") return;
			const rootPath = path.resolve(root, "index.ts");
			const entryFile = createFile({
				path: rootPath,
				baseName: "index.ts",
				exports: app.files.filter((file) => {
					return file.sources.some((source) => source.isIndexable);
				}).flatMap((file) => {
					var _file$sources;
					const containsOnlyTypes = file.sources.every((source) => source.isTypeOnly);
					return (_file$sources = file.sources) === null || _file$sources === void 0 ? void 0 : _file$sources.map((source) => {
						if (!file.path || !source.isIndexable) return;
						return {
							name: mode === "all" ? void 0 : [source.name],
							path: getRelativePath(rootPath, file.path),
							isTypeOnly: mode === "all" ? containsOnlyTypes : source.isTypeOnly
						};
					}).filter(Boolean);
				}).filter(Boolean),
				sources: []
			});
			await app.context.fileManager.add(entryFile);
			await app.context.fileManager.write({
				mode: (_app$context$options = app.context.options) === null || _app$context$options === void 0 ? void 0 : _app$context$options.mode,
				dryRun: options.dryRun,
				parsers: app.context.installedParsers
			});
		} };
	}
});

//#endregion
//#region src/plugins/progressPlugin.ts
const progressPlugin = createPlugin({
	name: "progress",
	install(app) {
		const progressBar = new SingleBar({
			format: "{bar} {percentage}% | {value}/{total} | {message}",
			barCompleteChar: "█",
			barIncompleteChar: "░",
			hideCursor: true,
			clearOnComplete: true
		}, Presets.shades_grey);
		app.context.events.on("process:start", async ({ files }) => {
			progressBar.start(files.length, 0, { message: "Starting..." });
		});
		app.context.events.on("process:progress", async ({ file }) => {
			const message = `Writing ${relative(process.cwd(), file.path)}`;
			progressBar.increment(1, { message });
		});
		app.context.events.on("process:end", async ({ files }) => {
			progressBar.update(files.length, { message: "Done ✅" });
			progressBar.stop();
		});
	}
});

//#endregion
//#region src/plugins/graphPlugin.ts
function getGraph({ files, root }) {
	const treeNode = TreeNode.fromFiles(files, root);
	if (!treeNode) return;
	return TreeNode.toGraph(treeNode);
}
const html = `
      <!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>File Graph</title>
    <script type="module">
      import { Network } from 'https://cdn.jsdelivr.net/npm/vis-network/standalone/esm/vis-network.min.js'

      async function main() {
        const res = await fetch('./graph.json')
        const { nodes, edges } = await res.json()
        const container = document.getElementById('graph')

        const network = new Network(
          container,
          { nodes, edges },
          {
            layout: { hierarchical: { direction: 'UD', sortMethod: 'directed' } },
            nodes: { shape: 'box', font: { face: 'monospace' } },
            edges: { arrows: 'to' },
            physics: false,
          },
        )
      }

      main()
    <\/script>
    <style>
      html, body, #graph { height: 100%; margin: 0; }
    </style>
  </head>
  <body>
    <div id="graph"></div>
  </body>
</html>
`;
async function serve(root) {
	const server = http.createServer((req, res) => {
		return handler(req, res, {
			public: root,
			cleanUrls: true
		});
	});
	server.listen(0, async () => {
		const { port } = server.address();
		console.log(`Running on http://localhost:${port}/graph.html`);
		await open(`http://localhost:${port}/graph.html`);
	});
}
const graphPlugin = createPlugin({
	name: "graph",
	install(app, options) {
		if (!options) throw new Error("Graph plugin requires options.root and options.mode");
		app.context.events.on("write:start", async ({ files }) => {
			const root = options.root;
			const graph = getGraph({
				files,
				root
			});
			if (!graph) return;
			const graphFile = createFile({
				baseName: "graph.json",
				path: path.join(root, "graph.json"),
				sources: [{
					name: "graph",
					value: JSON.stringify(graph, null, 2)
				}]
			});
			const graphHtmlFile = createFile({
				baseName: "graph.html",
				path: path.join(root, "graph.html"),
				sources: [{
					name: "graph",
					value: html
				}]
			});
			await app.context.fileManager.add(graphFile, graphHtmlFile);
			if (options.open) await serve(root);
		});
	}
});

//#endregion
export { barrelPlugin, createPlugin, fsPlugin, graphPlugin, progressPlugin };
