// use-mask-input/dist/index.js
import { getUnmaskedValue, resolveInputRef, withMask, isServer_default, setUnmaskedValue, makeMaskCacheKey, flow, setPrevRef, applyMaskToElement } from './chunk-XSTQDKDU.js';
export { withMask } from './chunk-XSTQDKDU.js';
import { useRef, useCallback, useEffect, useLayoutEffect, useMemo } from 'react';

function useMaskInput(props) {
  const { mask, register, options } = props;
  const ref = useRef(null);
  const maskRef = useRef(mask);
  const optionsRef = useRef(options);
  const unmaskedValue = useCallback(() => getUnmaskedValue(ref.current), []);
  const refCallback = useCallback((input) => {
    if (!input) {
      ref.current = null;
      return;
    }
    ref.current = resolveInputRef(input);
    withMask(maskRef.current, optionsRef.current)(ref.current);
  }, []);
  useEffect(() => {
    if (isServer_default || !ref.current || !register) return;
    register(ref.current);
  }, [register]);
  if (isServer_default) {
    const noop = (() => {
    });
    return setUnmaskedValue(noop, () => "");
  }
  return setUnmaskedValue(refCallback, unmaskedValue);
}
function useHookFormMask(registerFn) {
  const entryCacheRef = useRef(/* @__PURE__ */ new Map());
  useLayoutEffect(() => {
    entryCacheRef.current.forEach((entry) => {
      const currentEntry = entry;
      if (!currentEntry.element || !currentEntry.latestRHFRef) return;
      if (currentEntry.latestRHFRef !== currentEntry.syncedRHFRef) {
        currentEntry.latestRHFRef(currentEntry.element);
        currentEntry.syncedRHFRef = currentEntry.latestRHFRef;
      }
    });
  });
  return useMemo(() => {
    entryCacheRef.current = /* @__PURE__ */ new Map();
    return (fieldName, mask, options) => {
      if (!registerFn) throw new Error("registerFn is required");
      const registerReturn = registerFn(fieldName, options);
      const { ref } = registerReturn;
      const cacheKey = makeMaskCacheKey(fieldName, mask);
      let entry = entryCacheRef.current.get(cacheKey);
      if (!entry) {
        const nextEntry = {
          element: null,
          latestRHFRef: ref,
          syncedRHFRef: void 0,
          stableRef: null
        };
        const applyMaskToRef = (_ref) => {
          nextEntry.element = _ref;
          if (_ref) applyMaskToElement(_ref, mask, options);
          return _ref;
        };
        nextEntry.stableRef = nextEntry.latestRHFRef ? flow(applyMaskToRef, (_ref) => nextEntry.latestRHFRef?.(_ref)) : applyMaskToRef;
        entry = nextEntry;
        entryCacheRef.current.set(cacheKey, nextEntry);
      } else {
        entry.latestRHFRef = ref;
      }
      const result = {
        ...registerReturn,
        ref: entry.stableRef
      };
      setUnmaskedValue(result, () => getUnmaskedValue(entry?.element ?? null));
      setPrevRef(result, ref);
      return result;
    };
  }, [registerFn]);
}

// src/api/withTanStackFormMask.ts
var refCache = /* @__PURE__ */ new WeakMap();
function withTanStackFormMask(inputProps, mask, options) {
  const { ref } = inputProps;
  if (!ref) {
    let currentElement = null;
    const result2 = {
      ...inputProps,
      ref: ((input) => {
        currentElement = input;
        if (input) applyMaskToElement(input, mask, options);
      })
    };
    setUnmaskedValue(result2, () => getUnmaskedValue(currentElement));
    setPrevRef(result2, ref);
    return result2;
  }
  if (!refCache.has(ref)) {
    refCache.set(ref, /* @__PURE__ */ new Map());
  }
  const maskCache = refCache.get(ref);
  const cacheKey = makeMaskCacheKey(inputProps.name ?? "", mask);
  if (!maskCache?.has(cacheKey)) {
    const maskedRef2 = ((input) => {
      maskedRef2.currentElement = input;
      if (input) applyMaskToElement(input, mask, options);
      ref(input);
    });
    maskCache?.set(cacheKey, maskedRef2);
  }
  const maskedRef = maskCache?.get(cacheKey);
  const result = {
    ...inputProps,
    ref: maskedRef
  };
  setUnmaskedValue(result, () => getUnmaskedValue(maskedRef?.currentElement ?? null));
  setPrevRef(result, ref);
  return result;
}

// src/api/useTanStackFormMask.ts
function useTanStackFormMask() {
  return useMemo(
    () => (mask, inputProps, options) => withTanStackFormMask(inputProps, mask, options),
    []
  );
}

// src/api/withHookFormMask.ts
var refCache2 = /* @__PURE__ */ new WeakMap();
function withHookFormMask(register, mask, options) {
  const { ref } = register;
  if (!ref) {
    const result2 = {
      ...register,
      ref: null
    };
    setUnmaskedValue(result2, () => "");
    setPrevRef(result2, ref);
    return result2;
  }
  if (!refCache2.has(ref)) {
    refCache2.set(ref, /* @__PURE__ */ new Map());
  }
  const maskCache = refCache2.get(ref);
  const cacheKey = makeMaskCacheKey(register.name, mask);
  if (!maskCache?.has(cacheKey)) {
    const maskedRef2 = ((input) => {
      maskedRef2.currentElement = input;
      if (input) applyMaskToElement(input, mask, options);
      return ref(input);
    });
    maskCache?.set(cacheKey, maskedRef2);
  }
  const maskedRef = maskCache?.get(cacheKey);
  const result = {
    ...register,
    ref: maskedRef
  };
  setUnmaskedValue(result, () => getUnmaskedValue(maskedRef?.currentElement ?? null));
  setPrevRef(result, ref);
  return result;
}

export { useHookFormMask, useMaskInput, useTanStackFormMask, withHookFormMask, withTanStackFormMask };
