// use-mask-input/dist/index.cjs
'use strict';

var chunkPMBRAXS4_cjs = require('./chunk-PMBRAXS4.cjs');
var react = require('react');

function useMaskInput(props) {
  const { mask, register, options } = props;
  const ref = react.useRef(null);
  const maskRef = react.useRef(mask);
  const optionsRef = react.useRef(options);
  const unmaskedValue = react.useCallback(() => chunkPMBRAXS4_cjs.getUnmaskedValue(ref.current), []);
  const refCallback = react.useCallback((input) => {
    if (!input) {
      ref.current = null;
      return;
    }
    ref.current = chunkPMBRAXS4_cjs.resolveInputRef(input);
    chunkPMBRAXS4_cjs.withMask(maskRef.current, optionsRef.current)(ref.current);
  }, []);
  react.useEffect(() => {
    if (chunkPMBRAXS4_cjs.isServer_default || !ref.current || !register) return;
    register(ref.current);
  }, [register]);
  if (chunkPMBRAXS4_cjs.isServer_default) {
    const noop = (() => {
    });
    return chunkPMBRAXS4_cjs.setUnmaskedValue(noop, () => "");
  }
  return chunkPMBRAXS4_cjs.setUnmaskedValue(refCallback, unmaskedValue);
}
function useHookFormMask(registerFn) {
  const entryCacheRef = react.useRef(/* @__PURE__ */ new Map());
  react.useLayoutEffect(() => {
    entryCacheRef.current.forEach((entry) => {
      const currentEntry = entry;
      if (!currentEntry.element || !currentEntry.latestRHFRef) return;
      if (currentEntry.latestRHFRef !== currentEntry.syncedRHFRef) {
        currentEntry.latestRHFRef(currentEntry.element);
        currentEntry.syncedRHFRef = currentEntry.latestRHFRef;
      }
    });
  });
  return react.useMemo(() => {
    entryCacheRef.current = /* @__PURE__ */ new Map();
    return (fieldName, mask, options) => {
      if (!registerFn) throw new Error("registerFn is required");
      const registerReturn = registerFn(fieldName, options);
      const { ref } = registerReturn;
      const cacheKey = chunkPMBRAXS4_cjs.makeMaskCacheKey(fieldName, mask);
      let entry = entryCacheRef.current.get(cacheKey);
      if (!entry) {
        const nextEntry = {
          element: null,
          latestRHFRef: ref,
          syncedRHFRef: void 0,
          stableRef: null
        };
        const applyMaskToRef = (_ref) => {
          nextEntry.element = _ref;
          if (_ref) chunkPMBRAXS4_cjs.applyMaskToElement(_ref, mask, options);
          return _ref;
        };
        nextEntry.stableRef = nextEntry.latestRHFRef ? chunkPMBRAXS4_cjs.flow(applyMaskToRef, (_ref) => nextEntry.latestRHFRef?.(_ref)) : applyMaskToRef;
        entry = nextEntry;
        entryCacheRef.current.set(cacheKey, nextEntry);
      } else {
        entry.latestRHFRef = ref;
      }
      const result = {
        ...registerReturn,
        ref: entry.stableRef
      };
      chunkPMBRAXS4_cjs.setUnmaskedValue(result, () => chunkPMBRAXS4_cjs.getUnmaskedValue(entry?.element ?? null));
      chunkPMBRAXS4_cjs.setPrevRef(result, ref);
      return result;
    };
  }, [registerFn]);
}

// src/api/withTanStackFormMask.ts
var refCache = /* @__PURE__ */ new WeakMap();
function withTanStackFormMask(inputProps, mask, options) {
  const { ref } = inputProps;
  if (!ref) {
    let currentElement = null;
    const result2 = {
      ...inputProps,
      ref: ((input) => {
        currentElement = input;
        if (input) chunkPMBRAXS4_cjs.applyMaskToElement(input, mask, options);
      })
    };
    chunkPMBRAXS4_cjs.setUnmaskedValue(result2, () => chunkPMBRAXS4_cjs.getUnmaskedValue(currentElement));
    chunkPMBRAXS4_cjs.setPrevRef(result2, ref);
    return result2;
  }
  if (!refCache.has(ref)) {
    refCache.set(ref, /* @__PURE__ */ new Map());
  }
  const maskCache = refCache.get(ref);
  const cacheKey = chunkPMBRAXS4_cjs.makeMaskCacheKey(inputProps.name ?? "", mask);
  if (!maskCache?.has(cacheKey)) {
    const maskedRef2 = ((input) => {
      maskedRef2.currentElement = input;
      if (input) chunkPMBRAXS4_cjs.applyMaskToElement(input, mask, options);
      ref(input);
    });
    maskCache?.set(cacheKey, maskedRef2);
  }
  const maskedRef = maskCache?.get(cacheKey);
  const result = {
    ...inputProps,
    ref: maskedRef
  };
  chunkPMBRAXS4_cjs.setUnmaskedValue(result, () => chunkPMBRAXS4_cjs.getUnmaskedValue(maskedRef?.currentElement ?? null));
  chunkPMBRAXS4_cjs.setPrevRef(result, ref);
  return result;
}

// src/api/useTanStackFormMask.ts
function useTanStackFormMask() {
  return react.useMemo(
    () => (mask, inputProps, options) => withTanStackFormMask(inputProps, mask, options),
    []
  );
}

// src/api/withHookFormMask.ts
var refCache2 = /* @__PURE__ */ new WeakMap();
function withHookFormMask(register, mask, options) {
  const { ref } = register;
  if (!ref) {
    const result2 = {
      ...register,
      ref: null
    };
    chunkPMBRAXS4_cjs.setUnmaskedValue(result2, () => "");
    chunkPMBRAXS4_cjs.setPrevRef(result2, ref);
    return result2;
  }
  if (!refCache2.has(ref)) {
    refCache2.set(ref, /* @__PURE__ */ new Map());
  }
  const maskCache = refCache2.get(ref);
  const cacheKey = chunkPMBRAXS4_cjs.makeMaskCacheKey(register.name, mask);
  if (!maskCache?.has(cacheKey)) {
    const maskedRef2 = ((input) => {
      maskedRef2.currentElement = input;
      if (input) chunkPMBRAXS4_cjs.applyMaskToElement(input, mask, options);
      return ref(input);
    });
    maskCache?.set(cacheKey, maskedRef2);
  }
  const maskedRef = maskCache?.get(cacheKey);
  const result = {
    ...register,
    ref: maskedRef
  };
  chunkPMBRAXS4_cjs.setUnmaskedValue(result, () => chunkPMBRAXS4_cjs.getUnmaskedValue(maskedRef?.currentElement ?? null));
  chunkPMBRAXS4_cjs.setPrevRef(result, ref);
  return result;
}

Object.defineProperty(exports, "withMask", {
  enumerable: true,
  get: function () { return chunkPMBRAXS4_cjs.withMask; }
});
exports.useHookFormMask = useHookFormMask;
exports.useMaskInput = useMaskInput;
exports.useTanStackFormMask = useTanStackFormMask;
exports.withHookFormMask = withHookFormMask;
exports.withTanStackFormMask = withTanStackFormMask;
