// @kubb/plugin-react-query/dist/index.js
import { a as MutationKey, l as QueryKey } from "./components-Dc5XkKq5.js";
import { i as queryGenerator, n as infiniteQueryGenerator, r as mutationGenerator, t as suspenseQueryGenerator } from "./generators-CqUlfy01.js";
import path from "node:path";
import { PluginManager, createPlugin, getBarrelFiles, getMode } from "@kubb/core";
import { camelCase, pascalCase } from "@kubb/core/transformers";
import { OperationGenerator, pluginOasName } from "@kubb/plugin-oas";
import { pluginTsName } from "@kubb/plugin-ts";
import { pluginZodName } from "@kubb/plugin-zod";

//#region src/plugin.ts
const pluginReactQueryName = "plugin-react-query";
const pluginReactQuery = createPlugin((options) => {
	const { output = {
		path: "hooks",
		barrelType: "named"
	}, group, exclude = [], include, override = [], parser = "client", suspense = {}, infinite = false, transformers = {}, paramsType = "inline", pathParamsType = paramsType === "object" ? "object" : options.pathParamsType || "inline", generators = [
		queryGenerator,
		suspenseQueryGenerator,
		infiniteQueryGenerator,
		mutationGenerator
	].filter(Boolean), mutation = {}, query = {}, mutationKey = MutationKey.getTransformer, queryKey = QueryKey.getTransformer, paramsCasing, contentType } = options;
	return {
		name: pluginReactQueryName,
		options: {
			output,
			client: {
				importPath: "@kubb/plugin-client/clients/axios",
				dataReturnType: "data",
				pathParamsType,
				...options.client
			},
			infinite: infinite ? {
				queryParam: "id",
				initialPageParam: 0,
				cursorParam: void 0,
				...infinite
			} : false,
			suspense,
			queryKey,
			query: query === false ? false : {
				methods: ["get"],
				importPath: "@tanstack/react-query",
				...query
			},
			mutationKey,
			mutation: {
				methods: [
					"post",
					"put",
					"patch",
					"delete"
				],
				importPath: "@tanstack/react-query",
				...mutation
			},
			paramsType,
			pathParamsType,
			parser,
			paramsCasing,
			group
		},
		pre: [
			pluginOasName,
			pluginTsName,
			parser === "zod" ? pluginZodName : void 0
		].filter(Boolean),
		resolvePath(baseName, pathMode, options$1) {
			const root = path.resolve(this.config.root, this.config.output.path);
			if ((pathMode ?? getMode(path.resolve(root, output.path))) === "single")
 /**
			* when output is a file then we will always append to the same file(output file), see fileManager.addOrAppend
			* Other plugins then need to call addOrAppend instead of just add from the fileManager class
			*/
			return path.resolve(root, output.path);
			if (group && (options$1?.group?.path || options$1?.group?.tag)) {
				const groupName = group?.name ? group.name : (ctx) => {
					if (group?.type === "path") return `${ctx.group.split("/")[1]}`;
					return `${camelCase(ctx.group)}Controller`;
				};
				return path.resolve(root, output.path, groupName({ group: group.type === "path" ? options$1.group.path : options$1.group.tag }), baseName);
			}
			return path.resolve(root, output.path, baseName);
		},
		resolveName(name, type) {
			let resolvedName = camelCase(name);
			if (type === "file" || type === "function") resolvedName = camelCase(name, { isFile: type === "file" });
			if (type === "type") resolvedName = pascalCase(name);
			if (type) return transformers?.name?.(resolvedName, type) || resolvedName;
			return resolvedName;
		},
		async buildStart() {
			const [swaggerPlugin] = PluginManager.getDependedPlugins(this.plugins, [pluginOasName]);
			const oas = await swaggerPlugin.context.getOas();
			const root = path.resolve(this.config.root, this.config.output.path);
			const mode = getMode(path.resolve(root, output.path));
			const baseURL = await swaggerPlugin.context.getBaseURL();
			if (baseURL) this.plugin.options.client.baseURL = baseURL;
			const files = await new OperationGenerator(this.plugin.options, {
				oas,
				pluginManager: this.pluginManager,
				plugin: this.plugin,
				contentType,
				exclude,
				include,
				override,
				mode
			}).build(...generators);
			await this.addFile(...files);
			const barrelFiles = await getBarrelFiles(this.fileManager.files, {
				type: output.barrelType ?? "named",
				root,
				output,
				meta: { pluginKey: this.plugin.key },
				logger: this.logger
			});
			await this.addFile(...barrelFiles);
		}
	};
});

//#endregion
export { pluginReactQuery, pluginReactQueryName };
