// @kubb/plugin-react-query/dist/index.cjs
const require_components = require('./components-DUCi70hD.cjs');
const require_generators = require('./generators-DaV-5jJ5.cjs');
let node_path = require("node:path");
node_path = require_components.__toESM(node_path);
let __kubb_core = require("@kubb/core");
__kubb_core = require_components.__toESM(__kubb_core);
let __kubb_core_transformers = require("@kubb/core/transformers");
__kubb_core_transformers = require_components.__toESM(__kubb_core_transformers);
let __kubb_plugin_oas = require("@kubb/plugin-oas");
__kubb_plugin_oas = require_components.__toESM(__kubb_plugin_oas);
let __kubb_plugin_ts = require("@kubb/plugin-ts");
__kubb_plugin_ts = require_components.__toESM(__kubb_plugin_ts);
let __kubb_plugin_zod = require("@kubb/plugin-zod");
__kubb_plugin_zod = require_components.__toESM(__kubb_plugin_zod);

//#region src/plugin.ts
const pluginReactQueryName = "plugin-react-query";
const pluginReactQuery = (0, __kubb_core.createPlugin)((options) => {
	const { output = {
		path: "hooks",
		barrelType: "named"
	}, group, exclude = [], include, override = [], parser = "client", suspense = {}, infinite = false, transformers = {}, paramsType = "inline", pathParamsType = paramsType === "object" ? "object" : options.pathParamsType || "inline", generators = [
		require_generators.queryGenerator,
		require_generators.suspenseQueryGenerator,
		require_generators.infiniteQueryGenerator,
		require_generators.mutationGenerator
	].filter(Boolean), mutation = {}, query = {}, mutationKey = require_components.MutationKey.getTransformer, queryKey = require_components.QueryKey.getTransformer, paramsCasing, contentType } = options;
	return {
		name: pluginReactQueryName,
		options: {
			output,
			client: {
				importPath: "@kubb/plugin-client/clients/axios",
				dataReturnType: "data",
				pathParamsType,
				...options.client
			},
			infinite: infinite ? {
				queryParam: "id",
				initialPageParam: 0,
				cursorParam: void 0,
				...infinite
			} : false,
			suspense,
			queryKey,
			query: query === false ? false : {
				methods: ["get"],
				importPath: "@tanstack/react-query",
				...query
			},
			mutationKey,
			mutation: {
				methods: [
					"post",
					"put",
					"patch",
					"delete"
				],
				importPath: "@tanstack/react-query",
				...mutation
			},
			paramsType,
			pathParamsType,
			parser,
			paramsCasing,
			group
		},
		pre: [
			__kubb_plugin_oas.pluginOasName,
			__kubb_plugin_ts.pluginTsName,
			parser === "zod" ? __kubb_plugin_zod.pluginZodName : void 0
		].filter(Boolean),
		resolvePath(baseName, pathMode, options$1) {
			const root = node_path.default.resolve(this.config.root, this.config.output.path);
			if ((pathMode ?? (0, __kubb_core.getMode)(node_path.default.resolve(root, output.path))) === "single")
 /**
			* when output is a file then we will always append to the same file(output file), see fileManager.addOrAppend
			* Other plugins then need to call addOrAppend instead of just add from the fileManager class
			*/
			return node_path.default.resolve(root, output.path);
			if (group && (options$1?.group?.path || options$1?.group?.tag)) {
				const groupName = group?.name ? group.name : (ctx) => {
					if (group?.type === "path") return `${ctx.group.split("/")[1]}`;
					return `${(0, __kubb_core_transformers.camelCase)(ctx.group)}Controller`;
				};
				return node_path.default.resolve(root, output.path, groupName({ group: group.type === "path" ? options$1.group.path : options$1.group.tag }), baseName);
			}
			return node_path.default.resolve(root, output.path, baseName);
		},
		resolveName(name, type) {
			let resolvedName = (0, __kubb_core_transformers.camelCase)(name);
			if (type === "file" || type === "function") resolvedName = (0, __kubb_core_transformers.camelCase)(name, { isFile: type === "file" });
			if (type === "type") resolvedName = (0, __kubb_core_transformers.pascalCase)(name);
			if (type) return transformers?.name?.(resolvedName, type) || resolvedName;
			return resolvedName;
		},
		async buildStart() {
			const [swaggerPlugin] = __kubb_core.PluginManager.getDependedPlugins(this.plugins, [__kubb_plugin_oas.pluginOasName]);
			const oas = await swaggerPlugin.context.getOas();
			const root = node_path.default.resolve(this.config.root, this.config.output.path);
			const mode = (0, __kubb_core.getMode)(node_path.default.resolve(root, output.path));
			const baseURL = await swaggerPlugin.context.getBaseURL();
			if (baseURL) this.plugin.options.client.baseURL = baseURL;
			const files = await new __kubb_plugin_oas.OperationGenerator(this.plugin.options, {
				oas,
				pluginManager: this.pluginManager,
				plugin: this.plugin,
				contentType,
				exclude,
				include,
				override,
				mode
			}).build(...generators);
			await this.addFile(...files);
			const barrelFiles = await (0, __kubb_core.getBarrelFiles)(this.fileManager.files, {
				type: output.barrelType ?? "named",
				root,
				output,
				meta: { pluginKey: this.plugin.key },
				logger: this.logger
			});
			await this.addFile(...barrelFiles);
		}
	};
});

//#endregion
exports.pluginReactQuery = pluginReactQuery;
exports.pluginReactQueryName = pluginReactQueryName;
