// @redocly/openapi-core/lib/types/entity.js
import { isPlainObject } from '../utils/is-plain-object.js';
import { getNodeTypesFromJSONSchema } from './json-schema-adapter.js';
export const ENTITY_DISCRIMINATOR_PROPERTY_NAME = 'type';
export const ENTITY_TYPES_WITH_API_SUPPORT = ['api-description', 'api-operation', 'data-schema'];
export function createEntityTypes(entitySchema, entityDefaultSchema) {
    const { ctx: defaultNodeTypes } = getNodeTypesFromJSONSchema('EntityFileDefault', entityDefaultSchema);
    const { ctx: namedNodeTypes, discriminatorResolver: namedDiscriminatorResolver } = getNodeTypesFromJSONSchema('EntityFile', entitySchema);
    const arrayNodeType = {
        properties: {},
        items: (value) => {
            if (isPlainObject(value)) {
                const typeValue = value[ENTITY_DISCRIMINATOR_PROPERTY_NAME];
                if (!typeValue)
                    return 'Entity';
                const resolvedType = namedDiscriminatorResolver?.(value, typeValue);
                if (resolvedType && typeof resolvedType === 'string') {
                    return resolvedType;
                }
            }
            return 'Entity';
        },
    };
    return {
        entityTypes: {
            ...defaultNodeTypes,
            ...namedNodeTypes,
            EntityFileArray: arrayNodeType,
        },
        discriminatorResolver: namedDiscriminatorResolver,
    };
}
