// @redocly/openapi-core/lib/types/json-schema-draft7.shared.js
import { isMappingRef } from '../ref-utils.js';
import { listOf } from './index.js';
export const Schema = {
    properties: {
        $id: { type: 'string' },
        $schema: { type: 'string' },
        definitions: 'NamedSchemas',
        externalDocs: 'ExternalDocs',
        discriminator: 'Discriminator',
        title: { type: 'string' },
        multipleOf: { type: 'number', minimum: 0 },
        maximum: { type: 'number' },
        minimum: { type: 'number' },
        exclusiveMaximum: { type: 'number' },
        exclusiveMinimum: { type: 'number' },
        maxLength: { type: 'integer', minimum: 0 },
        minLength: { type: 'integer', minimum: 0 },
        pattern: { type: 'string' },
        maxItems: { type: 'integer', minimum: 0 },
        minItems: { type: 'integer', minimum: 0 },
        uniqueItems: { type: 'boolean' },
        maxProperties: { type: 'integer', minimum: 0 },
        minProperties: { type: 'integer', minimum: 0 },
        required: { type: 'array', items: { type: 'string' } },
        enum: { type: 'array' },
        type: (value) => {
            return Array.isArray(value)
                ? {
                    type: 'array',
                    items: { enum: ['object', 'array', 'string', 'number', 'integer', 'boolean', 'null'] },
                }
                : {
                    enum: ['object', 'array', 'string', 'number', 'integer', 'boolean', 'null'],
                };
        },
        allOf: listOf('Schema'),
        anyOf: listOf('Schema'),
        oneOf: listOf('Schema'),
        not: 'Schema',
        if: 'Schema',
        then: 'Schema',
        else: 'Schema',
        contains: 'Schema',
        patternProperties: { type: 'object' },
        propertyNames: 'Schema',
        properties: 'SchemaProperties',
        items: (value) => {
            return Array.isArray(value) ? listOf('Schema') : 'Schema';
        },
        additionalProperties: (value) => {
            return typeof value === 'boolean' ? { type: 'boolean' } : 'Schema';
        },
        description: { type: 'string' },
        format: { type: 'string' },
        contentEncoding: { type: 'string' },
        contentMediaType: { type: 'string' },
        default: null,
        readOnly: { type: 'boolean' },
        writeOnly: { type: 'boolean' },
        examples: { type: 'array' },
        example: { isExample: true },
        deprecated: { type: 'boolean' },
        const: null,
        $comment: { type: 'string' },
        additionalItems: (value) => {
            return typeof value === 'boolean' ? { type: 'boolean' } : 'Schema';
        },
        dependencies: 'Dependencies',
    },
};
export const SchemaProperties = {
    properties: {},
    additionalProperties: (value) => {
        return typeof value === 'boolean' ? { type: 'boolean' } : 'Schema';
    },
};
export const Dependencies = {
    properties: {},
    additionalProperties: (value) => {
        return Array.isArray(value) ? { type: 'array', items: { type: 'string' } } : 'Schema';
    },
};
export const DiscriminatorMapping = {
    properties: {},
    additionalProperties: (value) => {
        if (isMappingRef(value)) {
            return { type: 'string', directResolveAs: 'Schema' };
        }
        else {
            return { type: 'string' };
        }
    },
};
export const Discriminator = {
    properties: {
        propertyName: { type: 'string' },
        mapping: 'DiscriminatorMapping',
    },
    required: ['propertyName'],
};
