// @redocly/config/lib-esm/asyncapi-config-schema.js
import { feedbackConfigSchema } from './feedback-config-schema';
import { downloadUrlsSchema } from './redoc-config-schema';
export const asyncapiConfigSchema = {
    type: 'object',
    properties: {
        downloadUrls: downloadUrlsSchema,
        apiLogo: {
            type: 'object',
            properties: {
                imageUrl: {
                    type: 'string',
                },
                href: {
                    type: 'string',
                },
                altText: {
                    type: 'string',
                },
                backgroundColor: {
                    type: 'string',
                },
            },
        },
        jsonSamplesDepth: {
            type: 'number',
        },
        samplesMaxInlineArgs: {
            type: 'number',
        },
        fieldExpandLevel: {
            type: 'number',
        },
        baseUrlPath: {
            type: 'string',
        },
        metadata: {
            type: 'object',
            properties: {
                apiId: {
                    type: 'string',
                },
            },
            additionalProperties: true,
        },
        feedback: feedbackConfigSchema,
        layout: { type: 'string', enum: ['stacked', 'three-panel'] },
    },
    additionalProperties: false,
};
