// @redocly/config/lib-esm/common.js
const rulesSchema = {
    nodeTypeName: undefined,
    type: 'object',
    additionalProperties: {
        oneOf: [{ type: 'string' }, { type: 'object' }],
    },
    description: 'The rules configuration blocks set up linting rules and their severity. You can configure built-in rules, add configurable rules, and rules from plugins.',
    documentationLink: 'https://redocly.com/docs/cli/configuration/reference/rules#rules',
};
export const ruleTypes = [
    'rules',
    'oas2Rules',
    'oas3_0Rules',
    'oas3_1Rules',
    'oas3_2Rules',
    'async2Rules',
    'async3Rules',
    'arazzo1Rules',
    'overlay1Rules',
    'openrpc1Rules',
];
const ruleSchemas = {
    rules: rulesSchema,
    oas2Rules: rulesSchema,
    oas3_0Rules: rulesSchema,
    oas3_1Rules: rulesSchema,
    oas3_2Rules: rulesSchema,
    async2Rules: rulesSchema,
    async3Rules: rulesSchema,
    arazzo1Rules: rulesSchema,
    overlay1Rules: rulesSchema,
    openrpc1Rules: rulesSchema,
};
export const preprocessorTypes = [
    'preprocessors',
    'oas2Preprocessors',
    'oas3_0Preprocessors',
    'oas3_1Preprocessors',
    'oas3_2Preprocessors',
    'async2Preprocessors',
    'async3Preprocessors',
    'arazzo1Preprocessors',
    'overlay1Preprocessors',
    'openrpc1Preprocessors',
];
const preprocessorsSchema = {
    nodeTypeName: undefined,
    type: 'object',
    additionalProperties: true,
};
const preprocessorSchemas = {
    preprocessors: preprocessorsSchema,
    oas2Preprocessors: preprocessorsSchema,
    oas3_0Preprocessors: preprocessorsSchema,
    oas3_1Preprocessors: preprocessorsSchema,
    oas3_2Preprocessors: preprocessorsSchema,
    async2Preprocessors: preprocessorsSchema,
    async3Preprocessors: preprocessorsSchema,
    arazzo1Preprocessors: preprocessorsSchema,
    overlay1Preprocessors: preprocessorsSchema,
    openrpc1Preprocessors: preprocessorsSchema,
};
const decoratorsSchema = {
    nodeTypeName: undefined,
    type: 'object',
    additionalProperties: true,
};
export const decoratorTypes = [
    'decorators',
    'oas2Decorators',
    'oas3_0Decorators',
    'oas3_1Decorators',
    'oas3_2Decorators',
    'async2Decorators',
    'async3Decorators',
    'arazzo1Decorators',
    'overlay1Decorators',
    'openrpc1Decorators',
];
const decoratorSchemas = {
    decorators: decoratorsSchema,
    oas2Decorators: decoratorsSchema,
    oas3_0Decorators: decoratorsSchema,
    oas3_1Decorators: decoratorsSchema,
    oas3_2Decorators: decoratorsSchema,
    async2Decorators: decoratorsSchema,
    async3Decorators: decoratorsSchema,
    arazzo1Decorators: decoratorsSchema,
    overlay1Decorators: decoratorsSchema,
    openrpc1Decorators: decoratorsSchema,
};
const extendsSchema = {
    nodeTypeName: undefined,
    type: 'array',
    items: { type: 'string' },
    description: 'Use extends to inherit rules and their configurations from other rulesets.',
    documentationLink: 'https://redocly.com/docs/cli/configuration/reference/extends',
};
export const configGovernanceSchema = Object.assign(Object.assign(Object.assign(Object.assign({}, ruleSchemas), decoratorSchemas), preprocessorSchemas), { extends: extendsSchema });
