// oas/dist/chunk-BRYRBTD7.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; } function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }

var _chunkO3GIPZLCcjs = require('./chunk-O3GIPZLC.cjs');










var _chunkB5WP4BJMcjs = require('./chunk-B5WP4BJM.cjs');


var _chunkAYA3UT4Lcjs = require('./chunk-AYA3UT4L.cjs');


var _chunkVQBEI5WIcjs = require('./chunk-VQBEI5WI.cjs');

// src/operation/index.ts
var _jsonschemarefparser = require('@apidevtools/json-schema-ref-parser');

// src/lib/build-discriminator-one-of.ts
function hasDiscriminatorWithoutPolymorphism(schema) {
  if (!schema || typeof schema !== "object") return false;
  if (!("discriminator" in schema)) return false;
  if ("oneOf" in schema || "anyOf" in schema) return false;
  return true;
}
function allOfReferencesSchema(schema, targetSchemaName) {
  if (!schema || typeof schema !== "object") return false;
  if (!("allOf" in schema) || !Array.isArray(schema.allOf)) return false;
  return schema.allOf.some((item) => {
    if (_chunkVQBEI5WIcjs.isRef.call(void 0, item)) {
      const refParts = item.$ref.split("/");
      const refSchemaName = refParts[refParts.length - 1];
      return refSchemaName === targetSchemaName;
    }
    return false;
  });
}
function findDiscriminatorChildren(api) {
  const childrenMap = /* @__PURE__ */ new Map();
  const invertedChildrenMap = /* @__PURE__ */ new Map();
  if (!_optionalChain([api, 'optionalAccess', _2 => _2.components, 'optionalAccess', _3 => _3.schemas]) || typeof api.components.schemas !== "object") {
    return { children: childrenMap, inverted: invertedChildrenMap };
  }
  const schemas = api.components.schemas;
  const schemaNames = Object.keys(schemas);
  const discriminatorSchemas = schemaNames.filter((name) => {
    return hasDiscriminatorWithoutPolymorphism(schemas[name]);
  });
  for (const baseName of discriminatorSchemas) {
    const baseSchema = schemas[baseName];
    const discriminator = baseSchema.discriminator;
    let childSchemaNames;
    if (discriminator.mapping && typeof discriminator.mapping === "object") {
      const mappingRefs = Object.values(discriminator.mapping);
      if (mappingRefs.length > 0) {
        childSchemaNames = mappingRefs.map((ref) => {
          const parts = ref.split("/");
          return parts[parts.length - 1];
        });
      }
    }
    if (!childSchemaNames || childSchemaNames.length === 0) {
      childSchemaNames = schemaNames.filter((name) => {
        if (name === baseName) return false;
        return allOfReferencesSchema(schemas[name], baseName);
      });
    }
    if (childSchemaNames.length > 0) {
      childrenMap.set(baseName, childSchemaNames);
    }
  }
  for (const [key, values] of childrenMap) {
    for (const value of values) {
      if (invertedChildrenMap.has(value)) {
        _optionalChain([invertedChildrenMap, 'access', _4 => _4.get, 'call', _5 => _5(value), 'optionalAccess', _6 => _6.push, 'call', _7 => _7(key)]);
      } else {
        invertedChildrenMap.set(value, [key]);
      }
    }
  }
  return { children: childrenMap, inverted: invertedChildrenMap };
}
function buildDiscriminatorOneOf(api, childrenMap) {
  if (!_optionalChain([api, 'optionalAccess', _8 => _8.components, 'optionalAccess', _9 => _9.schemas]) || typeof api.components.schemas !== "object") {
    return;
  } else if (childrenMap.size === 0) {
    return;
  }
  for (const [schemaName, childNames] of childrenMap) {
    const schema = api.components.schemas[schemaName];
    if (!schema) continue;
    const oneOf = [];
    for (const childName of childNames) {
      if (api.components.schemas[childName]) {
        oneOf.push(_chunkB5WP4BJMcjs.cloneObject.call(void 0, api.components.schemas[childName]));
      }
    }
    if (oneOf.length > 0) {
      schema.oneOf = oneOf;
    }
  }
  for (const [parentSchemaName, childNames] of childrenMap) {
    for (const childName of childNames) {
      const childSchema = api.components.schemas[childName];
      if (!childSchema || !("allOf" in childSchema) || !Array.isArray(childSchema.allOf)) {
        continue;
      }
      for (let i = 0; i < childSchema.allOf.length; i++) {
        const item = childSchema.allOf[i];
        if (item && typeof item === "object" && "x-readme-ref-name" in item && item["x-readme-ref-name"] === parentSchemaName && "oneOf" in item) {
          const clonedItem = _chunkB5WP4BJMcjs.cloneObject.call(void 0, item);
          delete clonedItem.oneOf;
          childSchema.allOf[i] = clonedItem;
        }
      }
    }
  }
}

// src/operation/lib/dedupe-common-parameters.ts
function dedupeCommonParameters(parameters, commonParameters) {
  return commonParameters.filter((param) => {
    return !parameters.find((param2) => {
      if (param.name && param2.name) {
        return param.name === param2.name && param.in === param2.in;
      } else if (_chunkVQBEI5WIcjs.isRef.call(void 0, param) && _chunkVQBEI5WIcjs.isRef.call(void 0, param2)) {
        return param.$ref === param2.$ref;
      }
      return false;
    });
  });
}

// src/samples/index.ts
var _jsonschemamergeallof = require('json-schema-merge-allof'); var _jsonschemamergeallof2 = _interopRequireDefault(_jsonschemamergeallof);
var _memoizee = require('memoizee'); var _memoizee2 = _interopRequireDefault(_memoizee);

// src/samples/utils.ts
function usesPolymorphism(schema) {
  if (schema.oneOf) {
    return "oneOf";
  } else if (schema.anyOf) {
    return "anyOf";
  } else if (schema.allOf) {
    return "allOf";
  }
  return false;
}
function objectify(thing) {
  if (!_chunkB5WP4BJMcjs.isObject.call(void 0, thing)) {
    return {};
  }
  return thing;
}
function normalizeArray(arr) {
  if (Array.isArray(arr)) {
    return arr;
  }
  return [arr];
}
function isFunc(thing) {
  return typeof thing === "function";
}
function deeplyStripKey(input, keyToStrip, predicate = (obj, key) => true) {
  if (typeof input !== "object" || Array.isArray(input) || input === null || !keyToStrip) {
    return input;
  }
  const obj = { ...input };
  Object.keys(obj).forEach((k) => {
    if (k === keyToStrip && predicate(obj[k], k)) {
      delete obj[k];
      return;
    }
    obj[k] = deeplyStripKey(obj[k], keyToStrip, predicate);
  });
  return obj;
}

// src/samples/index.ts
var sampleDefaults = (genericSample) => {
  return (schema) => typeof schema.default === typeof genericSample ? schema.default : genericSample;
};
var primitives = {
  string: sampleDefaults("string"),
  string_email: sampleDefaults("user@example.com"),
  "string_date-time": sampleDefaults((/* @__PURE__ */ new Date()).toISOString()),
  string_date: sampleDefaults((/* @__PURE__ */ new Date()).toISOString().substring(0, 10)),
  "string_YYYY-MM-DD": sampleDefaults((/* @__PURE__ */ new Date()).toISOString().substring(0, 10)),
  string_uuid: sampleDefaults("3fa85f64-5717-4562-b3fc-2c963f66afa6"),
  string_hostname: sampleDefaults("example.com"),
  string_ipv4: sampleDefaults("198.51.100.42"),
  string_ipv6: sampleDefaults("2001:0db8:5b96:0000:0000:426f:8e17:642a"),
  number: sampleDefaults(0),
  number_float: sampleDefaults(0),
  integer: sampleDefaults(0),
  boolean: sampleDefaults(true)
};
var primitive = (schema) => {
  const objectifiedSchema = objectify(schema);
  const { format } = objectifiedSchema;
  let { type } = objectifiedSchema;
  if (type === "null") {
    return null;
  } else if (Array.isArray(type)) {
    if (type.length === 1) {
      type = type[0];
    } else {
      if (type.includes("null")) {
        type = type.filter((t) => t !== "null");
      }
      type = type.shift();
    }
  }
  const fn = primitives[`${type}_${format}`] || primitives[type];
  if (isFunc(fn)) {
    return fn(objectifiedSchema);
  }
  return `Unknown Type: ${objectifiedSchema.type}`;
};
function sampleFromSchema(schema, opts = {}) {
  const objectifySchema = objectify(schema);
  let { type } = objectifySchema;
  const hasPolymorphism = usesPolymorphism(objectifySchema);
  if (hasPolymorphism === "allOf") {
    try {
      return sampleFromSchema(
        _jsonschemamergeallof2.default.call(void 0, objectifySchema, {
          resolvers: {
            // Ignore any unrecognized OAS-specific keywords that might be present on the schema
            // (like `xml`).
            defaultResolver: _jsonschemamergeallof2.default.options.resolvers.title
          }
        }),
        opts
      );
    } catch (e) {
      return void 0;
    }
  } else if (hasPolymorphism) {
    const samples = objectifySchema[hasPolymorphism].map((s) => {
      return sampleFromSchema(s, opts);
    });
    if (samples.length === 1) {
      return samples[0];
    } else if (samples.some((s) => s === null)) {
      return samples.find((s) => s !== null);
    }
    return samples[0];
  }
  const { example, additionalProperties, properties, items } = objectifySchema;
  const { includeReadOnly, includeWriteOnly } = opts;
  if (example !== void 0) {
    return deeplyStripKey(example, "$$ref", (val) => {
      return typeof val === "string" && val.indexOf("#") > -1;
    });
  }
  if (!type) {
    if (properties || additionalProperties) {
      type = "object";
    } else if (items) {
      type = "array";
    } else {
      return void 0;
    }
  }
  if (type === "object" || Array.isArray(type) && type.includes("object")) {
    const props = objectify(properties);
    const obj = {};
    for (const name in props) {
      if (_optionalChain([props, 'optionalAccess', _10 => _10[name], 'access', _11 => _11.deprecated])) {
        continue;
      }
      if (_optionalChain([props, 'optionalAccess', _12 => _12[name], 'access', _13 => _13.readOnly]) && !includeReadOnly) {
        continue;
      }
      if (_optionalChain([props, 'optionalAccess', _14 => _14[name], 'access', _15 => _15.writeOnly]) && !includeWriteOnly) {
        continue;
      }
      if (_optionalChain([props, 'access', _16 => _16[name], 'access', _17 => _17.examples, 'optionalAccess', _18 => _18.length])) {
        obj[name] = props[name].examples[0];
        continue;
      }
      obj[name] = sampleFromSchema(props[name], opts);
    }
    if (additionalProperties === true) {
      obj.additionalProp = {};
    } else if (additionalProperties) {
      const additionalProps = objectify(additionalProperties);
      const additionalPropVal = sampleFromSchema(additionalProps, opts);
      obj.additionalProp = additionalPropVal;
    }
    return obj;
  }
  if (type === "array" || Array.isArray(type) && type.includes("array")) {
    if (typeof items === "undefined") {
      return [];
    }
    if (Array.isArray(items.anyOf)) {
      return items.anyOf.map((i) => sampleFromSchema(i, opts));
    }
    if (Array.isArray(items.oneOf)) {
      return items.oneOf.map((i) => sampleFromSchema(i, opts));
    }
    return [sampleFromSchema(items, opts)];
  }
  if (schema.enum) {
    if (schema.default) {
      return schema.default;
    }
    return normalizeArray(schema.enum)[0];
  }
  if (type === "file") {
    return void 0;
  }
  return primitive(schema);
}
var memo = _memoizee2.default.call(void 0, sampleFromSchema);
var samples_default = memo;

// src/operation/lib/get-mediatype-examples.ts
function getMediaTypeExamples(mediaType, mediaTypeObject, opts = {}) {
  if (mediaTypeObject.example) {
    return [
      {
        value: mediaTypeObject.example
      }
    ];
  } else if (mediaTypeObject.examples) {
    const { examples } = mediaTypeObject;
    const multipleExamples = Object.keys(examples).map((key) => {
      let summary = key;
      let description;
      let example = examples[key];
      if (example !== null && typeof example === "object") {
        if ("summary" in example) {
          summary = example.summary;
        }
        if ("description" in example) {
          description = example.description;
        }
        if ("value" in example) {
          if (example.value !== null && typeof example.value === "object" && "$ref" in example.value) {
            return false;
          }
          example = example.value;
        }
      }
      const ret = { summary, title: key, value: example };
      if (description) {
        ret.description = description;
      }
      return ret;
    }).filter((item) => item !== false);
    if (multipleExamples.length) {
      return multipleExamples;
    }
  }
  if (mediaTypeObject.schema) {
    if (!_chunkB5WP4BJMcjs.matches_mimetype_default.xml(mediaType)) {
      return [
        {
          value: samples_default(JSON.parse(JSON.stringify(mediaTypeObject.schema)), opts)
        }
      ];
    }
  }
  return [];
}

// src/operation/lib/get-response-examples.ts
function getResponseExamples(operation) {
  return Object.keys(operation.responses || {}).map((status) => {
    const response = _optionalChain([operation, 'access', _19 => _19.responses, 'optionalAccess', _20 => _20[status]]);
    let onlyHeaders = false;
    if (!response || _chunkVQBEI5WIcjs.isRef.call(void 0, response)) {
      return false;
    }
    const mediaTypes = {};
    (_optionalChain([response, 'optionalAccess', _21 => _21.content]) ? Object.keys(response.content) : []).forEach((mediaType) => {
      if (!mediaType) return;
      const mediaTypeObject = _optionalChain([response, 'access', _22 => _22.content, 'optionalAccess', _23 => _23[mediaType]]);
      if (!mediaTypeObject) return;
      const examples = getMediaTypeExamples(mediaType, mediaTypeObject, {
        includeReadOnly: true,
        includeWriteOnly: false
      });
      if (examples) {
        mediaTypes[mediaType] = examples;
      }
    });
    if (response.headers && Object.keys(response.headers).length && !Object.keys(mediaTypes).length) {
      mediaTypes["*/*"] = [];
      onlyHeaders = true;
    }
    if (!Object.keys(mediaTypes).length) {
      return false;
    }
    return {
      status,
      mediaTypes,
      ...onlyHeaders ? { onlyHeaders } : {}
    };
  }).filter((item) => item !== false);
}

// src/operation/lib/get-callback-examples.ts
function getCallbackExamples(operation) {
  if (!operation.callbacks) {
    return [];
  }
  const examples = Object.keys(operation.callbacks).map((identifier) => {
    const callback = _optionalChain([operation, 'access', _24 => _24.callbacks, 'optionalAccess', _25 => _25[identifier]]);
    if (!callback || _chunkVQBEI5WIcjs.isRef.call(void 0, callback)) {
      return [];
    }
    const items = Object.keys(callback).map((expression) => {
      return Object.keys(callback[expression]).map((method) => {
        const pathItem = callback[expression];
        const example = getResponseExamples(pathItem[method]);
        if (example.length === 0) return false;
        return {
          identifier,
          expression,
          method,
          example
        };
      });
    });
    return items.flat().filter((item) => item !== false);
  });
  return examples.flat();
}

// src/operation/lib/get-example-groups.ts
var noCorrespondingResponseKey = "NoCorrespondingResponseForCustomCodeSample";
function addMatchingResponseExamples(groups, operation) {
  operation.getResponseExamples().forEach((example) => {
    Object.entries(example.mediaTypes || {}).forEach(([mediaType, mediaTypeExamples]) => {
      mediaTypeExamples.forEach((mediaTypeExample) => {
        if (mediaTypeExample.title && Object.keys(groups).includes(mediaTypeExample.title)) {
          groups[mediaTypeExample.title].response = {
            mediaType,
            mediaTypeExample,
            status: example.status
          };
          if (!groups[mediaTypeExample.title].name) {
            groups[mediaTypeExample.title].name = mediaTypeExample.summary || mediaTypeExample.title;
          }
        }
      });
    });
  });
}
function getDefaultName(sample, count) {
  return sample.name && sample.name.length > 0 ? sample.name : `Default${count[sample.language] > 1 ? ` #${count[sample.language]}` : ""}`;
}
function getExampleGroups(operation) {
  const namelessCodeSampleCounts = {};
  const groups = {};
  const codeSamples = _chunkAYA3UT4Lcjs.getExtension.call(void 0, "code-samples", operation.api, operation);
  _optionalChain([codeSamples, 'optionalAccess', _26 => _26.forEach, 'call', _27 => _27((sample, i) => {
    if (namelessCodeSampleCounts[sample.language]) {
      namelessCodeSampleCounts[sample.language] += 1;
    } else {
      namelessCodeSampleCounts[sample.language] = 1;
    }
    const name = getDefaultName(sample, namelessCodeSampleCounts);
    if (sample.correspondingExample) {
      if (_optionalChain([groups, 'access', _28 => _28[sample.correspondingExample], 'optionalAccess', _29 => _29.customCodeSamples, 'optionalAccess', _30 => _30.length])) {
        groups[sample.correspondingExample].customCodeSamples.push({ ...sample, name, originalIndex: i });
      } else if (sample.correspondingExample) {
        groups[sample.correspondingExample] = {
          name,
          customCodeSamples: [{ ...sample, name, originalIndex: i }]
        };
      }
    } else if (_optionalChain([groups, 'access', _31 => _31[noCorrespondingResponseKey], 'optionalAccess', _32 => _32.customCodeSamples, 'optionalAccess', _33 => _33.length])) {
      groups[noCorrespondingResponseKey].customCodeSamples.push({ ...sample, name, originalIndex: i });
    } else {
      groups[noCorrespondingResponseKey] = {
        name,
        customCodeSamples: [{ ...sample, name, originalIndex: i }]
      };
    }
  })]);
  if (Object.keys(groups).length) {
    addMatchingResponseExamples(groups, operation);
    return groups;
  }
  operation.getParameters().forEach((param) => {
    Object.entries(param.examples || {}).forEach(([exampleKey, paramExample]) => {
      if (_chunkVQBEI5WIcjs.isRef.call(void 0, paramExample)) {
        return;
      }
      groups[exampleKey] = {
        ...groups[exampleKey],
        name: _optionalChain([groups, 'access', _34 => _34[exampleKey], 'optionalAccess', _35 => _35.name]) || paramExample.summary || exampleKey,
        request: {
          ..._optionalChain([groups, 'access', _36 => _36[exampleKey], 'optionalAccess', _37 => _37.request]),
          [param.in]: {
            ..._optionalChain([groups, 'access', _38 => _38[exampleKey], 'optionalAccess', _39 => _39.request, 'optionalAccess', _40 => _40[param.in]]),
            [param.name]: paramExample.value
          }
        }
      };
    });
  });
  operation.getRequestBodyExamples().forEach((requestExample) => {
    requestExample.examples.forEach((mediaTypeExample) => {
      if (mediaTypeExample.title) {
        const mediaType = requestExample.mediaType === "application/x-www-form-urlencoded" ? "formData" : "body";
        groups[mediaTypeExample.title] = {
          ...groups[mediaTypeExample.title],
          name: _optionalChain([groups, 'access', _41 => _41[mediaTypeExample.title], 'optionalAccess', _42 => _42.name]) || mediaTypeExample.summary || mediaTypeExample.title,
          request: {
            ..._optionalChain([groups, 'access', _43 => _43[mediaTypeExample.title], 'optionalAccess', _44 => _44.request]),
            [mediaType]: mediaTypeExample.value
          }
        };
      }
    });
  });
  if (Object.keys(groups).length) {
    addMatchingResponseExamples(groups, operation);
  }
  Object.entries(groups).forEach(([groupId, group]) => {
    if (group.request && !group.response) {
      delete groups[groupId];
    }
  });
  return groups;
}

// src/operation/lib/get-requestbody-examples.ts
function getRequestBodyExamples(operation) {
  const requestBody = operation.requestBody;
  if (!requestBody || _chunkVQBEI5WIcjs.isRef.call(void 0, requestBody) || !requestBody.content) {
    return [];
  }
  return Object.keys(requestBody.content || {}).map((mediaType) => {
    const mediaTypeObject = requestBody.content[mediaType];
    const examples = getMediaTypeExamples(mediaType, mediaTypeObject, {
      includeReadOnly: false,
      includeWriteOnly: true
    });
    if (!examples.length) {
      return false;
    }
    return {
      mediaType,
      examples
    };
  }).filter((item) => item !== false);
}

// src/operation/lib/get-response-as-json-schema.ts
var isJSON = _chunkB5WP4BJMcjs.matches_mimetype_default.json;
function buildHeadersSchema(response, opts) {
  const headersSchema = {
    type: "object",
    properties: {}
  };
  if (response.headers) {
    Object.keys(response.headers).forEach((key) => {
      if (!_optionalChain([response, 'access', _45 => _45.headers, 'optionalAccess', _46 => _46[key]]) || _chunkVQBEI5WIcjs.isRef.call(void 0, _optionalChain([response, 'access', _47 => _47.headers, 'optionalAccess', _48 => _48[key]]))) {
        return;
      }
      if (response.headers[key].schema) {
        const header = response.headers[key];
        if (!header.schema || _chunkVQBEI5WIcjs.isRef.call(void 0, header.schema)) {
          return;
        }
        headersSchema.properties[key] = _chunkB5WP4BJMcjs.toJSONSchema.call(void 0, header.schema, {
          addEnumsToDescriptions: true,
          transformer: _optionalChain([opts, 'optionalAccess', _49 => _49.transformer])
        });
        if (header.description) {
          headersSchema.properties[key].description = header.description;
        }
      }
    });
  }
  const headersWrapper = {
    schema: headersSchema,
    type: "object",
    label: "Headers"
  };
  if (response.description && headersWrapper.schema) {
    headersWrapper.description = response.description;
  }
  return headersWrapper;
}
function getResponseAsJSONSchema(operation, api, statusCode, opts) {
  const response = operation.getResponseByStatusCode(statusCode);
  const jsonSchema = [];
  if (!response) {
    return null;
  }
  let hasCircularRefs = false;
  let hasDiscriminatorMappingRefs = false;
  function refLogger(ref, type) {
    if (type === "ref") {
      hasCircularRefs = true;
    } else {
      hasDiscriminatorMappingRefs = true;
    }
  }
  function getPreferredSchema(content, preferredContentType) {
    if (!content) {
      return null;
    }
    const contentTypes = Object.keys(content);
    if (!contentTypes.length) {
      return null;
    }
    if (preferredContentType) {
      if (contentTypes.includes(preferredContentType)) {
        const schema2 = _chunkB5WP4BJMcjs.cloneObject.call(void 0, content[preferredContentType].schema);
        if (!schema2) {
          return null;
        }
        return _chunkB5WP4BJMcjs.toJSONSchema.call(void 0, schema2, {
          addEnumsToDescriptions: true,
          refLogger,
          transformer: _optionalChain([opts, 'optionalAccess', _50 => _50.transformer])
        });
      }
      return null;
    }
    for (let i = 0; i < contentTypes.length; i++) {
      if (isJSON(contentTypes[i])) {
        const schema2 = _chunkB5WP4BJMcjs.cloneObject.call(void 0, content[contentTypes[i]].schema);
        if (!schema2) {
          return {};
        }
        return _chunkB5WP4BJMcjs.toJSONSchema.call(void 0, schema2, {
          addEnumsToDescriptions: true,
          refLogger,
          transformer: _optionalChain([opts, 'optionalAccess', _51 => _51.transformer])
        });
      }
    }
    const contentType = contentTypes.shift();
    if (!contentType) {
      return {};
    }
    const schema = _chunkB5WP4BJMcjs.cloneObject.call(void 0, content[contentType].schema);
    if (!schema) {
      return {};
    }
    return _chunkB5WP4BJMcjs.toJSONSchema.call(void 0, schema, {
      addEnumsToDescriptions: true,
      refLogger,
      transformer: _optionalChain([opts, 'optionalAccess', _52 => _52.transformer])
    });
  }
  const foundSchema = getPreferredSchema(response.content, _optionalChain([opts, 'optionalAccess', _53 => _53.contentType]));
  if (_optionalChain([opts, 'optionalAccess', _54 => _54.contentType]) && !foundSchema) {
    return null;
  }
  if (foundSchema) {
    const schema = _chunkB5WP4BJMcjs.cloneObject.call(void 0, foundSchema);
    const schemaWrapper = {
      // If there's no `type` then the root schema is a circular `$ref` that we likely won't be
      // able to render so instead of generating a JSON Schema with an `undefined` type we should
      // default to `string` so there's at least *something* the end-user can interact with.
      type: foundSchema.type || "string",
      schema: _chunkB5WP4BJMcjs.isPrimitive.call(void 0, schema) ? schema : {
        ...schema,
        $schema: _chunkB5WP4BJMcjs.getSchemaVersionString.call(void 0, schema, api)
      },
      label: "Response body"
    };
    if (response.description && schemaWrapper.schema) {
      schemaWrapper.description = response.description;
    }
    if (api.components && schemaWrapper.schema) {
      if (hasCircularRefs || hasDiscriminatorMappingRefs && _optionalChain([opts, 'optionalAccess', _55 => _55.includeDiscriminatorMappingRefs])) {
        schemaWrapper.schema.components = _chunkB5WP4BJMcjs.cloneObject.call(void 0, 
          api.components
        );
      }
    }
    jsonSchema.push(schemaWrapper);
  }
  if (response.headers) {
    jsonSchema.push(buildHeadersSchema(response, opts));
  }
  return jsonSchema.length ? jsonSchema : null;
}

// src/operation/lib/operationId.ts
function hasOperationId(operation) {
  return Boolean("operationId" in operation && _optionalChain([operation, 'access', _56 => _56.operationId, 'optionalAccess', _57 => _57.length]));
}
function getOperationId(path, method, operation, opts = {}) {
  function sanitize(id) {
    return id.replace(_optionalChain([opts, 'optionalAccess', _58 => _58.camelCase]) || _optionalChain([opts, 'optionalAccess', _59 => _59.friendlyCase]) ? /[^a-zA-Z0-9_]/g : /[^a-zA-Z0-9]/g, "-").replace(/--+/g, "-").replace(/^-|-$/g, "");
  }
  const operationIdExists = hasOperationId(operation);
  let operationId;
  if (operationIdExists) {
    operationId = operation.operationId;
  } else {
    operationId = sanitize(path).toLowerCase();
  }
  const currMethod = method.toLowerCase();
  if (_optionalChain([opts, 'optionalAccess', _60 => _60.camelCase]) || _optionalChain([opts, 'optionalAccess', _61 => _61.friendlyCase])) {
    if (_optionalChain([opts, 'optionalAccess', _62 => _62.friendlyCase])) {
      operationId = operationId.replaceAll("_", " ");
      if (!operationIdExists) {
        operationId = operationId.replace(/[^a-zA-Z0-9_]+(.)/g, (_, chr) => ` ${chr}`).split(" ").filter((word, i, arr) => word !== arr[i - 1]).join(" ");
      }
    }
    operationId = operationId.replace(/[^a-zA-Z0-9_]+(.)/g, (_, chr) => chr.toUpperCase());
    if (operationIdExists) {
      operationId = sanitize(operationId);
    }
    operationId = operationId.replace(/^[0-9]/g, (match) => `_${match}`);
    operationId = operationId.charAt(0).toLowerCase() + operationId.slice(1);
    if (operationId.startsWith(currMethod)) {
      return operationId;
    }
    if (operationIdExists) {
      return operationId;
    }
    operationId = operationId.charAt(0).toUpperCase() + operationId.slice(1);
    return `${currMethod}${operationId}`;
  } else if (operationIdExists) {
    return operationId;
  }
  return `${currMethod}_${operationId}`;
}

// src/operation/index.ts
var Operation = class {
  /**
   * Schema of the operation from the API Definition.
   */
  
  /**
   * OpenAPI API Definition that this operation originated from.
   */
  
  /**
   * Path that this operation is targeted towards.
   */
  
  /**
   * HTTP Method that this operation is targeted towards.
   */
  
  /**
   * The primary Content Type that this operation accepts.
   */
  
  /**
   * An object with groups of all example definitions (body/header/query/path/response/etc.)
   */
  
  /**
   * Request body examples for this operation.
   */
  
  /**
   * Response examples for this operation.
   */
  
  /**
   * Callback examples for this operation (if it has callbacks).
   */
  
  /**
   * Flattened out arrays of both request and response headers that are utilized on this operation.
   */
  
  /**
   * Internal storage array that the library utilizes to keep track of the times the
   * {@see Operation.dereference} has been called so that if you initiate multiple promises they'll
   * all end up returning the same data set once the initial dereference call completed.
   */
  
  /**
   * Internal storage array that the library utilizes to keep track of its `dereferencing` state so
   * it doesn't initiate multiple dereferencing processes.
   */
  
  constructor(api, path, method, operation) {
    this.schema = operation;
    this.api = api;
    this.path = path;
    this.method = method;
    this.contentType = void 0;
    this.requestBodyExamples = void 0;
    this.responseExamples = void 0;
    this.callbackExamples = void 0;
    this.exampleGroups = void 0;
    this.headers = {
      request: [],
      response: []
    };
    this.promises = [];
    this.dereferencing = {
      processing: false,
      complete: false,
      circularRefs: []
    };
  }
  /**
   * Retrieve the `summary` for this operation.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#user-content-operationsummary}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#user-content-operation-summary}
   */
  getSummary() {
    if (_optionalChain([this, 'access', _63 => _63.schema, 'optionalAccess', _64 => _64.summary]) && typeof this.schema.summary === "string") {
      return this.schema.summary;
    }
    const pathItem = _optionalChain([this, 'access', _65 => _65.api, 'access', _66 => _66.paths, 'optionalAccess', _67 => _67[this.path]]);
    if (_optionalChain([pathItem, 'optionalAccess', _68 => _68.summary]) && typeof pathItem.summary === "string") {
      return pathItem.summary;
    }
    return void 0;
  }
  /**
   * Retrieve the `description` for this operation.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#user-content-operationdescription}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#user-content-operation-description}
   */
  getDescription() {
    if (_optionalChain([this, 'access', _69 => _69.schema, 'optionalAccess', _70 => _70.description]) && typeof this.schema.description === "string") {
      return this.schema.description;
    }
    const pathItem = _optionalChain([this, 'access', _71 => _71.api, 'access', _72 => _72.paths, 'optionalAccess', _73 => _73[this.path]]);
    if (_optionalChain([pathItem, 'optionalAccess', _74 => _74.description]) && typeof pathItem.description === "string") {
      return pathItem.description;
    }
    return void 0;
  }
  /**
   * Retrieve the primary content type for this operation. If multiple exist, the first JSON-like
   * type will be returned.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#user-content-requestbodycontent}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#user-content-request-body-content}
   */
  getContentType() {
    if (this.contentType) {
      return this.contentType;
    }
    let types = [];
    if (this.schema.requestBody) {
      if (_chunkVQBEI5WIcjs.isRef.call(void 0, this.schema.requestBody)) {
        this.schema.requestBody = _chunkB5WP4BJMcjs.dereferenceRef.call(void 0, this.schema.requestBody, this.api);
      }
      if (this.schema.requestBody && "content" in this.schema.requestBody) {
        types = Object.keys(this.schema.requestBody.content);
      }
    }
    this.contentType = "application/json";
    if (_optionalChain([types, 'optionalAccess', _75 => _75.length])) {
      this.contentType = types[0];
    }
    types.forEach((t) => {
      if (_chunkB5WP4BJMcjs.matches_mimetype_default.json(t)) {
        this.contentType = t;
      }
    });
    return this.contentType;
  }
  /**
   * Checks if the current operation has a `x-www-form-urlencoded` content type payload.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#user-content-requestbodycontent}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#user-content-request-body-content}
   */
  isFormUrlEncoded() {
    return _chunkB5WP4BJMcjs.matches_mimetype_default.formUrlEncoded(this.getContentType());
  }
  /**
   * Checks if the current operation has a mutipart content type payload.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#user-content-requestbodycontent}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#user-content-request-body-content}
   */
  isMultipart() {
    return _chunkB5WP4BJMcjs.matches_mimetype_default.multipart(this.getContentType());
  }
  /**
   * Checks if the current operation has a JSON-like content type payload.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#user-content-requestbodycontent}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#user-content-request-body-content}
   */
  isJson() {
    return _chunkB5WP4BJMcjs.matches_mimetype_default.json(this.getContentType());
  }
  /**
   * Checks if the current operation has an XML content type payload.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#user-content-requestbodycontent}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#user-content-request-body-content}
   */
  isXml() {
    return _chunkB5WP4BJMcjs.matches_mimetype_default.xml(this.getContentType());
  }
  /**
   * Checks if the current operation is a webhook or not.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#oas-webhooks}
   */
  isWebhook() {
    return this instanceof Webhook;
  }
  /**
   * Returns an array of all security requirements associated wtih this operation. If none are
   * defined at the operation level, the securities for the entire API definition are returned
   * (with an empty array as a final fallback).
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#security-requirement-object}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#security-requirement-object}
   */
  getSecurity() {
    if (!_optionalChain([this, 'access', _76 => _76.api, 'optionalAccess', _77 => _77.components, 'optionalAccess', _78 => _78.securitySchemes]) || !Object.keys(this.api.components.securitySchemes).length) {
      return [];
    }
    return this.schema.security || this.api.security || [];
  }
  /**
   * Retrieve a collection of grouped security schemes. The inner array determines AND-grouped
   * security schemes, the outer array determines OR-groups.
   *
   * @see {@link https://swagger.io/docs/specification/authentication/#multiple}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#security-requirement-object}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#security-requirement-object}
   * @param filterInvalid Optional flag that, when set to `true`, filters out invalid/nonexistent
   *    security schemes, rather than returning `false`.
   */
  getSecurityWithTypes(filterInvalid = false) {
    return this.getSecurity().map((requirement) => {
      let keys;
      try {
        keys = Object.keys(requirement);
      } catch (e2) {
        return false;
      }
      const keysWithTypes = keys.map((key) => {
        let security;
        try {
          if (!_optionalChain([this, 'access', _79 => _79.api, 'access', _80 => _80.components, 'optionalAccess', _81 => _81.securitySchemes, 'optionalAccess', _82 => _82[key]]) || _chunkVQBEI5WIcjs.isRef.call(void 0, this.api.components.securitySchemes[key])) {
            return false;
          }
          security = this.api.components.securitySchemes[key];
        } catch (e3) {
          return false;
        }
        if (!security) return false;
        let type = null;
        if (security.type === "http") {
          if (security.scheme === "basic") type = "Basic";
          else if (security.scheme === "bearer") type = "Bearer";
          else type = security.type;
        } else if (security.type === "oauth2") {
          type = "OAuth2";
        } else if (security.type === "apiKey") {
          if (security.in === "query") type = "Query";
          else if (security.in === "header") type = "Header";
          else if (security.in === "cookie") type = "Cookie";
          else type = security.type;
        } else {
          return false;
        }
        return {
          type,
          security: {
            ...security,
            _key: key,
            _requirements: requirement[key]
          }
        };
      });
      if (filterInvalid) return keysWithTypes.filter((key) => key !== false);
      return keysWithTypes;
    });
  }
  /**
   * Retrieve an object where the keys are unique scheme types, and the values are arrays
   * containing each security scheme of that type.
   *
   */
  prepareSecurity() {
    return this.getSecurityWithTypes().reduce(
      (prev, securities) => {
        if (!securities) return prev;
        securities.forEach((security) => {
          if (!security) return;
          if (!prev[security.type]) prev[security.type] = [];
          const exists = prev[security.type].some((sec) => sec._key === security.security._key);
          if (!exists) {
            if (_optionalChain([security, 'access', _83 => _83.security, 'optionalAccess', _84 => _84._requirements])) delete security.security._requirements;
            prev[security.type].push(security.security);
          }
        });
        return prev;
      },
      {}
    );
  }
  /**
   * Retrieve all of the headers, request and response, that are associated with this operation.
   *
   */
  getHeaders() {
    const security = this.prepareSecurity();
    if (security.Header) {
      this.headers.request = security.Header.map((h) => {
        if (!("name" in h)) return false;
        return h.name;
      }).filter((item) => item !== false);
    }
    if (security.Bearer || security.Basic || security.OAuth2) {
      this.headers.request.push("Authorization");
    }
    if (security.Cookie) {
      this.headers.request.push("Cookie");
    }
    if (this.schema.parameters) {
      this.headers.request = this.headers.request.concat(
        this.schema.parameters.map((p) => {
          if (_chunkVQBEI5WIcjs.isRef.call(void 0, p)) {
            return void 0;
          }
          if (p.in && p.in === "header") return p.name;
          return void 0;
        }).filter((item) => item !== void 0)
      );
    }
    if (this.schema.responses) {
      this.headers.response = Object.keys(this.schema.responses).map((r) => {
        const response = _optionalChain([this, 'access', _85 => _85.schema, 'access', _86 => _86.responses, 'optionalAccess', _87 => _87[r]]);
        if (!response || _chunkVQBEI5WIcjs.isRef.call(void 0, response)) {
          return [];
        }
        return _optionalChain([response, 'optionalAccess', _88 => _88.headers]) ? Object.keys(response.headers) : [];
      }).reduce((a, b) => a.concat(b), []);
    }
    if (!this.headers.request.includes("Content-Type") && this.schema.requestBody) {
      const requestBody = this.schema.requestBody;
      if (requestBody && !_chunkVQBEI5WIcjs.isRef.call(void 0, requestBody) && "content" in requestBody && Object.keys(requestBody.content)) {
        this.headers.request.push("Content-Type");
      }
    }
    if (this.schema.responses) {
      if (Object.keys(this.schema.responses).some((r) => {
        const response = _optionalChain([this, 'access', _89 => _89.schema, 'access', _90 => _90.responses, 'optionalAccess', _91 => _91[r]]);
        if (!response || _chunkVQBEI5WIcjs.isRef.call(void 0, response)) {
          return false;
        }
        return _optionalChain([response, 'optionalAccess', _92 => _92.content]) && Object.keys(response.content).length > 0;
      })) {
        if (!this.headers.request.includes("Accept")) this.headers.request.push("Accept");
        if (!this.headers.response.includes("Content-Type")) this.headers.response.push("Content-Type");
      }
    }
    return this.headers;
  }
  /**
   * Determine if this operation has an `operationId` present in its schema. Note that if one is
   * present in the schema but is an empty string then this will return `false`.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#user-content-operationid}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#user-content-operation-id}
   */
  hasOperationId() {
    return hasOperationId(this.schema);
  }
  /**
   * Determine if an operation has an `operationId` present in its schema. Note that if one is
   * present in the schema but is an empty string then this will return `false`.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#user-content-operationid}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#user-content-operation-id}
   */
  static hasOperationId(schema) {
    return hasOperationId(schema);
  }
  /**
   * Get an `operationId` for this operation. If one is not present (it's not required by the spec!)
   * a hash of the path and method will be returned instead.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#user-content-operationid}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#user-content-operation-id}
   */
  getOperationId(opts = {}) {
    return getOperationId(this.path, this.method, this.schema, opts);
  }
  /**
   * Get an `operationId` for an operation. If one is not present (it's not required by the spec!)
   * a hash of the path and method will be returned instead.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#user-content-operationid}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#user-content-operation-id}
   */
  static getOperationId(path, method, schema, opts = {}) {
    return getOperationId(path, method, schema, opts);
  }
  /**
   * Return an array of all tags, and their metadata, that exist on this operation.
   *
   */
  getTags() {
    if (!("tags" in this.schema)) {
      return [];
    }
    const oasTagMap = /* @__PURE__ */ new Map();
    if (Array.isArray(_optionalChain([this, 'access', _93 => _93.api, 'optionalAccess', _94 => _94.tags]))) {
      this.api.tags.forEach((tag) => {
        oasTagMap.set(tag.name, tag);
      });
    }
    const oasTags = Object.fromEntries(oasTagMap);
    const tags = [];
    if (Array.isArray(this.schema.tags)) {
      this.schema.tags.forEach((tag) => {
        if (tag in oasTags) {
          tags.push(oasTags[tag]);
        } else {
          tags.push({
            name: tag
          });
        }
      });
    }
    return tags;
  }
  /**
   * Return is the operation is flagged as `deprecated` or not.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#user-content-operationdeprecated}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#user-content-operation-deprecated}
   */
  isDeprecated() {
    return Boolean("deprecated" in this.schema ? this.schema.deprecated : false);
  }
  /**
   * Determine if the operation has any (non-request body) parameters.
   *
   */
  hasParameters() {
    return !!this.getParameters().length;
  }
  /**
   * Return the parameters (non-request body) on the operation.
   *
   * @todo Add support for `ReferenceObject`
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#user-content-operationparameters}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#user-content-operation-parameters}
   */
  getParameters() {
    let parameters = (_optionalChain([this, 'access', _95 => _95.schema, 'optionalAccess', _96 => _96.parameters]) || []).filter(
      (param) => param && !_chunkVQBEI5WIcjs.isRef.call(void 0, param)
    );
    const commonParams = (_optionalChain([this, 'access', _97 => _97.api, 'optionalAccess', _98 => _98.paths, 'optionalAccess', _99 => _99[this.path], 'optionalAccess', _100 => _100.parameters]) || []).filter(
      (param) => param && !_chunkVQBEI5WIcjs.isRef.call(void 0, param)
    );
    if (commonParams.length) {
      parameters = parameters.concat(dedupeCommonParameters(parameters, commonParams) || []);
    }
    return parameters;
  }
  /**
   * Determine if this operation has any required parameters.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#user-content-operationparameters}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#user-content-operation-parameters}
   */
  hasRequiredParameters() {
    return this.getParameters().some((param) => "required" in param && param.required);
  }
  /**
   * Convert the operation into an array of JSON Schema schemas for each available type of
   * parameter available on the operation.
   *
   */
  getParametersAsJSONSchema(opts = {}) {
    return _chunkB5WP4BJMcjs.getParametersAsJSONSchema.call(void 0, this, this.api, {
      includeDiscriminatorMappingRefs: true,
      transformer: (s) => s,
      ...opts
    });
  }
  /**
   * Get a single response for this status code, formatted as JSON schema.
   *
   * @param statusCode Status code to pull a JSON Schema response for.
   * @param opts Options for schema generation.
   * @param opts.contentType Optional content-type to use. If specified and the response doesn't have
   *   this content-type, the function will return null.
   */
  getResponseAsJSONSchema(statusCode, opts = {}) {
    return getResponseAsJSONSchema(this, this.api, statusCode, {
      includeDiscriminatorMappingRefs: true,
      transformer: (s) => s,
      ...opts
    });
  }
  /**
   * Get an array of all valid response status codes for this operation.
   *
   */
  getResponseStatusCodes() {
    return this.schema.responses ? Object.keys(this.schema.responses) : [];
  }
  /**
   * Determine if the operation has any request bodies.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#user-content-operationrequestbody}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#user-content-operation-request-body}
   */
  hasRequestBody() {
    return !!this.schema.requestBody;
  }
  /**
   * Retrieve the list of all available media types that the operations request body can accept.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#media-type-object}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#media-type-object}
   */
  getRequestBodyMediaTypes() {
    if (!this.hasRequestBody()) {
      return [];
    }
    const requestBody = this.schema.requestBody;
    if (!requestBody || _chunkVQBEI5WIcjs.isRef.call(void 0, requestBody)) {
      return [];
    }
    return Object.keys(requestBody.content);
  }
  /**
   * Determine if this operation has a required request body.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#media-type-object}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#media-type-object}
   */
  hasRequiredRequestBody() {
    if (!this.hasRequestBody()) {
      return false;
    }
    const requestBody = this.schema.requestBody;
    if (!requestBody || _chunkVQBEI5WIcjs.isRef.call(void 0, requestBody)) {
      return false;
    }
    if (requestBody.required) {
      return true;
    }
    const parameters = this.getParametersAsJSONSchema();
    if (parameters === null) {
      return false;
    }
    return !!parameters.filter((js) => ["body", "formData"].includes(js.type)).find((js) => js.schema && Array.isArray(js.schema.required) && js.schema.required.length);
  }
  /**
   * Retrieve a specific request body content schema off this operation.
   *
   * If no media type is supplied this will return either the first available JSON-like request
   * body, or the first available if there are no JSON-like media types present. When this return
   * comes back it's in the form of an array with the first key being the selected media type,
   * followed by the media type object in question.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#media-type-object}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#media-type-object}
   * @param mediaType Specific request body media type to retrieve if present.
   */
  getRequestBody(mediaType) {
    if (!this.hasRequestBody()) {
      return false;
    }
    const requestBody = this.schema.requestBody;
    if (!requestBody || _chunkVQBEI5WIcjs.isRef.call(void 0, requestBody)) {
      return false;
    }
    if (mediaType) {
      if (!(mediaType in requestBody.content)) {
        return false;
      }
      return requestBody.content[mediaType];
    }
    let availableMediaType;
    const mediaTypes = this.getRequestBodyMediaTypes();
    mediaTypes.forEach((mt) => {
      if (!availableMediaType && _chunkB5WP4BJMcjs.matches_mimetype_default.json(mt)) {
        availableMediaType = mt;
      }
    });
    if (!availableMediaType) {
      mediaTypes.forEach((mt) => {
        if (!availableMediaType) {
          availableMediaType = mt;
        }
      });
    }
    if (availableMediaType) {
      return [
        availableMediaType,
        requestBody.content[availableMediaType],
        ...requestBody.description ? [requestBody.description] : []
      ];
    }
    return false;
  }
  /**
   * Retrieve an array of request body examples that this operation has.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#request-body-examples}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#request-body-examples}
   */
  getRequestBodyExamples() {
    const isRequestExampleValueDefined = typeof _optionalChain([this, 'access', _101 => _101.requestBodyExamples, 'optionalAccess', _102 => _102[0], 'optionalAccess', _103 => _103.examples, 'optionalAccess', _104 => _104[0], 'access', _105 => _105.value]) !== "undefined";
    if (this.requestBodyExamples && isRequestExampleValueDefined) {
      return this.requestBodyExamples;
    }
    this.requestBodyExamples = getRequestBodyExamples(this.schema);
    return this.requestBodyExamples;
  }
  /**
   * Return a specific response out of the operation by a given HTTP status code.
   *
   * @param statusCode Status code to pull a response object for.
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#response-object}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#response-object}
   */
  getResponseByStatusCode(statusCode) {
    if (!this.schema.responses) {
      return false;
    }
    const response = this.schema.responses[statusCode];
    if (!response || _chunkVQBEI5WIcjs.isRef.call(void 0, response)) {
      return false;
    }
    return response;
  }
  /**
   * Retrieve an array of response examples that this operation has.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#response-object-examples}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.2.md#response-object-examples}
   */
  getResponseExamples() {
    if (this.responseExamples) {
      return this.responseExamples;
    }
    this.responseExamples = getResponseExamples(this.schema);
    return this.responseExamples;
  }
  /**
   * Determine if the operation has callbacks.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#callback-object}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#callback-object}
   */
  hasCallbacks() {
    return Boolean(this.schema.callbacks);
  }
  /**
   * Retrieve a specific callback.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#callback-object}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#callback-object}
   * @param identifier Callback identifier to look for.
   * @param expression Callback expression to look for.
   * @param method HTTP Method on the callback to look for.
   */
  getCallback(identifier, expression, method) {
    if (!this.schema.callbacks) return false;
    const callbackObj = this.schema.callbacks[identifier];
    if (!callbackObj || _chunkVQBEI5WIcjs.isRef.call(void 0, callbackObj)) {
      return false;
    }
    const callback = callbackObj[expression];
    if (!callback || _chunkVQBEI5WIcjs.isRef.call(void 0, callback) || !callback[method]) {
      return false;
    }
    return new Callback(this.api, expression, method, callback[method], identifier, callback);
  }
  /**
   * Retrieve an array of operations created from each callback.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#callback-object}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#callback-object}
   */
  getCallbacks() {
    if (!this.hasCallbacks()) return [];
    const callbacks = [];
    Object.keys(this.schema.callbacks || {}).forEach((callback) => {
      const cb = _optionalChain([this, 'access', _106 => _106.schema, 'access', _107 => _107.callbacks, 'optionalAccess', _108 => _108[callback]]);
      if (!cb || _chunkVQBEI5WIcjs.isRef.call(void 0, cb)) {
        return;
      }
      Object.keys(cb).forEach((expression) => {
        const exp = cb[expression];
        if (!exp || _chunkVQBEI5WIcjs.isRef.call(void 0, exp)) {
          return;
        }
        Object.keys(exp).forEach((method) => {
          if (!_chunkO3GIPZLCcjs.supportedMethods.includes(method)) return;
          const found = this.getCallback(callback, expression, method);
          if (found) {
            callbacks.push(found);
          }
        });
      });
    });
    return callbacks;
  }
  /**
   * Retrieve an array of callback examples that this operation has.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#callback-object}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#callback-object}
   */
  getCallbackExamples() {
    if (this.callbackExamples) {
      return this.callbackExamples;
    }
    this.callbackExamples = getCallbackExamples(this.schema);
    return this.callbackExamples;
  }
  /**
   * Determine if a given a custom specification extension exists within the operation.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#specification-extensions}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#specification-extensions}
   * @param extension Specification extension to lookup.
   */
  hasExtension(extension) {
    return Boolean(this.schema && extension in this.schema);
  }
  /**
   * Retrieve a custom specification extension off of the operation.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#specification-extensions}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#specification-extensions}
   * @param extension Specification extension to lookup.
   *
   * @deprecated Use `oas.getExtension(extension, operation)` instead.
   */
  getExtension(extension) {
    return _optionalChain([this, 'access', _109 => _109.schema, 'optionalAccess', _110 => _110[extension]]);
  }
  /**
   * Returns an object with groups of all example definitions (body/header/query/path/response/etc.).
   * The examples are grouped by their key when defined via the `examples` map.
   *
   * Any custom code samples defined via the `x-readme.code-samples` extension are returned,
   * regardless of if they have a matching response example.
   *
   * For standard OAS request parameter (e.g., body/header/query/path/etc.) examples,
   * they are only present in the return object if they have a corresponding response example
   * (i.e., a response example with the same key in the `examples` map).
   */
  getExampleGroups() {
    if (this.exampleGroups) return this.exampleGroups;
    const groups = getExampleGroups(this);
    this.exampleGroups = groups;
    return groups;
  }
  /**
   * Dereference the current operation schema so it can be parsed free of worries of `$ref` schemas
   * and circular structures.
   *
   */
  async dereference(opts) {
    if (this.dereferencing.complete) {
      return Promise.resolve(true);
    }
    if (this.dereferencing.processing) {
      return new Promise((resolve, reject) => {
        this.promises.push({ resolve, reject });
      });
    }
    this.dereferencing.processing = true;
    const { children: discriminatorChildrenMap, inverted: discriminatorChildrenMapInverted } = findDiscriminatorChildren(this.api);
    const { api, schema, promises } = this;
    if (_optionalChain([api, 'optionalAccess', _111 => _111.components, 'optionalAccess', _112 => _112.schemas]) && typeof api.components.schemas === "object") {
      Object.keys(api.components.schemas).forEach((schemaName) => {
        if (_chunkB5WP4BJMcjs.isPrimitive.call(void 0, _optionalChain([api, 'access', _113 => _113.components, 'optionalAccess', _114 => _114.schemas, 'optionalAccess', _115 => _115[schemaName]])) || Array.isArray(_optionalChain([api, 'access', _116 => _116.components, 'optionalAccess', _117 => _117.schemas, 'optionalAccess', _118 => _118[schemaName]])) || _optionalChain([api, 'access', _119 => _119.components, 'optionalAccess', _120 => _120.schemas, 'optionalAccess', _121 => _121[schemaName]]) === null) {
          return;
        }
        (_optionalChain([api, 'access', _122 => _122.components, 'optionalAccess', _123 => _123.schemas, 'optionalAccess', _124 => _124[schemaName]]))["x-readme-ref-name"] = schemaName;
      });
    }
    const circularRefs = /* @__PURE__ */ new Set();
    const dereferencingOptions = _chunkB5WP4BJMcjs.getDereferencingOptions.call(void 0, circularRefs);
    const parser = new (0, _jsonschemarefparser.$RefParser)();
    return parser.dereference(
      "#/__INTERNAL__",
      {
        // Because `json-schema-ref-parser` will dereference this entire object as we only want
        // to dereference this operation schema we're attaching it to the `__INTERNAL__` key, and
        // later using that to extract our dereferenced schema. If we didn't do this then we run
        // the risk of any keyword in `schema` being overriden by `paths` and `components`.
        //
        // This solution isn't the best and still requires us to send `json-schema-ref-parser`
        // basically the entire API defintiion but because we don't know what `$ref` pointers in
        // `schema` reference, we can't know which parts of full API definition we could safely
        // exclude from this process.
        __INTERNAL__: structuredClone(schema),
        paths: _nullishCoalesce(api.paths, () => ( void 0)),
        components: _nullishCoalesce(api.components, () => ( void 0))
      },
      {
        ...dereferencingOptions,
        dereference: {
          ...dereferencingOptions.dereference,
          /**
           * Because we only want to dereference our `__INTERNAL__` schema, not the **entire**
           * API definition if the parser attemps to dereference anything but that then we
           * should bail out of that crawler.
           *
           * @fixme this may cause issues where a path references a schema within itself to be ignored.
           */
          excludedPathMatcher: (path) => {
            if (path === "#/paths" || path.startsWith("#/paths/")) {
              return true;
            }
            if (discriminatorChildrenMap.size > 0 || discriminatorChildrenMapInverted.size > 0) {
              if (path.startsWith("#/components/") && path !== "#/components/schemas" && !path.startsWith("#/components/schemas/")) {
                return true;
              }
              if (path.startsWith("#/components/schemas/")) {
                const schemaName = path.split("/").pop();
                if (schemaName && (discriminatorChildrenMap.has(schemaName) || discriminatorChildrenMapInverted.has(schemaName))) {
                  if (path === `#/components/schemas/${schemaName}` || path.startsWith(`#/components/schemas/${schemaName}/`)) {
                    return false;
                  }
                  return true;
                }
              }
              return false;
            }
            return path === "#/components" || path.startsWith("#/components/");
          }
        }
      }
    ).then((res) => {
      const dereferenced = res;
      if (_optionalChain([dereferenced, 'optionalAccess', _125 => _125.components, 'optionalAccess', _126 => _126.schemas]) && discriminatorChildrenMap.size > 0) {
        buildDiscriminatorOneOf({ components: dereferenced.components }, discriminatorChildrenMap);
      }
      this.schema = dereferenced.__INTERNAL__;
      this.promises = promises;
      this.dereferencing = {
        processing: false,
        complete: true,
        // We need to convert our `Set` to an array in order to match the typings.
        circularRefs: [...circularRefs]
      };
      if (_optionalChain([opts, 'optionalAccess', _127 => _127.cb])) {
        opts.cb();
      }
    }).then(() => {
      return this.promises.map((deferred) => deferred.resolve());
    }).catch((err) => {
      this.dereferencing.processing = false;
      this.promises.map((deferred) => deferred.reject(err));
      throw err;
    });
  }
  /**
   * Retrieve any circular `$ref` pointers that maybe present within operation schema.
   *
   * This method requires that you first dereference the definition.
   *
   * @see Operation.dereference
   */
  getCircularReferences() {
    if (!this.dereferencing.complete) {
      throw new Error(".dereference() must be called first in order for this method to obtain circular references.");
    }
    return this.dereferencing.circularRefs;
  }
};
var Callback = class extends Operation {
  /**
   * The identifier that this callback is set to.
   */
  
  /**
   * The parent path item object that this Callback exists within.
   */
  
  constructor(oas, path, method, operation, identifier, parentPathItem) {
    super(oas, path, method, operation);
    this.identifier = identifier;
    this.parentSchema = parentPathItem;
  }
  /**
   * Return the primary identifier for this callback.
   *
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#callback-object}
   * @see {@link https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#callback-object}
   */
  getIdentifier() {
    return this.identifier;
  }
  getSummary() {
    if (_optionalChain([this, 'access', _128 => _128.schema, 'optionalAccess', _129 => _129.summary]) && typeof this.schema.summary === "string") {
      return this.schema.summary;
    } else if (this.parentSchema.summary && typeof this.parentSchema.summary === "string") {
      return this.parentSchema.summary;
    }
    return void 0;
  }
  getDescription() {
    if (_optionalChain([this, 'access', _130 => _130.schema, 'optionalAccess', _131 => _131.description]) && typeof this.schema.description === "string") {
      return this.schema.description;
    } else if (this.parentSchema.description && typeof this.parentSchema.description === "string") {
      return this.parentSchema.description;
    }
    return void 0;
  }
  getParameters() {
    let parameters = _optionalChain([this, 'access', _132 => _132.schema, 'optionalAccess', _133 => _133.parameters]) || [];
    const commonParams = this.parentSchema.parameters || [];
    if (commonParams.length) {
      parameters = parameters.concat(dedupeCommonParameters(parameters, commonParams) || []);
    }
    return parameters;
  }
};
var Webhook = class extends Operation {
  getSummary() {
    if (_optionalChain([this, 'access', _134 => _134.schema, 'optionalAccess', _135 => _135.summary]) && typeof this.schema.summary === "string") {
      return this.schema.summary;
    }
    const webhookPath = _optionalChain([this, 'access', _136 => _136.api, 'access', _137 => _137.webhooks, 'optionalAccess', _138 => _138[this.path]]);
    if (webhookPath && !_chunkVQBEI5WIcjs.isRef.call(void 0, webhookPath)) {
      if (_optionalChain([webhookPath, 'optionalAccess', _139 => _139.summary]) && typeof webhookPath.summary === "string") {
        return webhookPath.summary;
      }
    }
    return void 0;
  }
  getDescription() {
    if (_optionalChain([this, 'access', _140 => _140.schema, 'optionalAccess', _141 => _141.description]) && typeof this.schema.description === "string") {
      return this.schema.description;
    }
    const webhookPath = _optionalChain([this, 'access', _142 => _142.api, 'access', _143 => _143.webhooks, 'optionalAccess', _144 => _144[this.path]]);
    if (webhookPath && !_chunkVQBEI5WIcjs.isRef.call(void 0, webhookPath)) {
      if (_optionalChain([webhookPath, 'optionalAccess', _145 => _145.description]) && typeof webhookPath.description === "string") {
        return webhookPath.description;
      }
    }
    return void 0;
  }
};







exports.findDiscriminatorChildren = findDiscriminatorChildren; exports.buildDiscriminatorOneOf = buildDiscriminatorOneOf; exports.Operation = Operation; exports.Callback = Callback; exports.Webhook = Webhook;
/**
 * Portions of this file have been extracted and modified from Swagger UI.
 *
 * @license Apache-2.0
 * @see {@link https://github.com/swagger-api/swagger-ui/blob/master/src/core/utils.js}
 */
/**
 * This file has been extracted and modified from Swagger UI.
 *
 * @license Apache-2.0
 * @see {@link https://github.com/swagger-api/swagger-ui/blob/master/src/core/plugins/samples/fn.js}
 */
