// oas/dist/extensions.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});















var _chunkAYA3UT4Lcjs = require('./chunk-AYA3UT4L.cjs');
















exports.CODE_SAMPLES = _chunkAYA3UT4Lcjs.CODE_SAMPLES; exports.DISABLE_TAG_SORTING = _chunkAYA3UT4Lcjs.DISABLE_TAG_SORTING; exports.EXPLORER_ENABLED = _chunkAYA3UT4Lcjs.EXPLORER_ENABLED; exports.HEADERS = _chunkAYA3UT4Lcjs.HEADERS; exports.INTERNAL = _chunkAYA3UT4Lcjs.INTERNAL; exports.METRICS_ENABLED = _chunkAYA3UT4Lcjs.METRICS_ENABLED; exports.OAUTH_OPTIONS = _chunkAYA3UT4Lcjs.OAUTH_OPTIONS; exports.PARAMETER_ORDERING = _chunkAYA3UT4Lcjs.PARAMETER_ORDERING; exports.PROXY_ENABLED = _chunkAYA3UT4Lcjs.PROXY_ENABLED; exports.SAMPLES_LANGUAGES = _chunkAYA3UT4Lcjs.SAMPLES_LANGUAGES; exports.SIMPLE_MODE = _chunkAYA3UT4Lcjs.SIMPLE_MODE; exports.extensionDefaults = _chunkAYA3UT4Lcjs.extensionDefaults; exports.getExtension = _chunkAYA3UT4Lcjs.getExtension; exports.hasRootExtension = _chunkAYA3UT4Lcjs.hasRootExtension; exports.validateParameterOrdering = _chunkAYA3UT4Lcjs.validateParameterOrdering;
