// oas/dist/extensions.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});















var _chunkKDRD2YJGcjs = require('./chunk-KDRD2YJG.cjs');
















exports.CODE_SAMPLES = _chunkKDRD2YJGcjs.CODE_SAMPLES; exports.DISABLE_TAG_SORTING = _chunkKDRD2YJGcjs.DISABLE_TAG_SORTING; exports.EXPLORER_ENABLED = _chunkKDRD2YJGcjs.EXPLORER_ENABLED; exports.HEADERS = _chunkKDRD2YJGcjs.HEADERS; exports.INTERNAL = _chunkKDRD2YJGcjs.INTERNAL; exports.METRICS_ENABLED = _chunkKDRD2YJGcjs.METRICS_ENABLED; exports.OAUTH_OPTIONS = _chunkKDRD2YJGcjs.OAUTH_OPTIONS; exports.PARAMETER_ORDERING = _chunkKDRD2YJGcjs.PARAMETER_ORDERING; exports.PROXY_ENABLED = _chunkKDRD2YJGcjs.PROXY_ENABLED; exports.SAMPLES_LANGUAGES = _chunkKDRD2YJGcjs.SAMPLES_LANGUAGES; exports.SIMPLE_MODE = _chunkKDRD2YJGcjs.SIMPLE_MODE; exports.extensionDefaults = _chunkKDRD2YJGcjs.extensionDefaults; exports.getExtension = _chunkKDRD2YJGcjs.getExtension; exports.hasRootExtension = _chunkKDRD2YJGcjs.hasRootExtension; exports.validateParameterOrdering = _chunkKDRD2YJGcjs.validateParameterOrdering;
