// @redocly/openapi-core/lib/lint.js
import { rootRedoclyConfigSchema } from '@redocly/config';
import { initRules } from './config/rules.js';
import { detectSpec, getMajorSpecVersion } from './detect-spec.js';
import { getTypes } from './oas-types.js';
import { BaseResolver, resolveDocument, makeDocumentFromString } from './resolve.js';
import { NoUnresolvedRefs } from './rules/common/no-unresolved-refs.js';
import { Struct } from './rules/common/struct.js';
import { normalizeTypes } from './types/index.js';
import { createConfigTypes } from './types/redocly-yaml.js';
import { normalizeVisitors, } from './visitors.js';
import { walkDocument } from './walk.js';
// FIXME: remove this once we remove `theme` from the schema
const { theme: _, ...propertiesWithoutTheme } = rootRedoclyConfigSchema.properties;
const redoclyConfigSchemaWithoutTheme = {
    ...rootRedoclyConfigSchema,
    properties: propertiesWithoutTheme,
};
export async function lint(opts) {
    const { ref, externalRefResolver = new BaseResolver(opts.config.resolve) } = opts;
    const document = (await externalRefResolver.resolveDocument(null, ref, true));
    opts.collectSpecData?.(document.parsed);
    return lintDocument({
        document,
        ...opts,
        externalRefResolver,
    });
}
export async function lintFromString(opts) {
    const { source, absoluteRef, externalRefResolver = new BaseResolver(opts.config.resolve) } = opts;
    const document = makeDocumentFromString(source, absoluteRef || '/');
    return lintDocument({
        document,
        ...opts,
        externalRefResolver,
    });
}
export async function lintDocument(opts) {
    const { document, customTypes, externalRefResolver, config } = opts;
    const specVersion = detectSpec(document.parsed);
    const rules = config.getRulesForSpecVersion(getMajorSpecVersion(specVersion));
    const types = normalizeTypes(config.extendTypes(customTypes ?? getTypes(specVersion), specVersion), config);
    const ctx = {
        problems: [],
        specVersion,
        config,
        visitorsData: {},
    };
    const preprocessors = initRules(rules, config, 'preprocessors', specVersion);
    const regularRules = initRules(rules, config, 'rules', specVersion);
    let resolvedRefMap = await resolveDocument({
        rootDocument: document,
        rootType: types.Root,
        externalRefResolver,
    });
    if (preprocessors.length > 0) {
        // Make additional pass to resolve refs defined in preprocessors.
        walkDocument({
            document,
            rootType: types.Root,
            normalizedVisitors: normalizeVisitors(preprocessors, types),
            resolvedRefMap,
            ctx,
        });
        resolvedRefMap = await resolveDocument({
            rootDocument: document,
            rootType: types.Root,
            externalRefResolver,
        });
    }
    const normalizedVisitors = normalizeVisitors(regularRules, types);
    walkDocument({
        document,
        rootType: types.Root,
        normalizedVisitors,
        resolvedRefMap,
        ctx,
    });
    return ctx.problems.map((problem) => config.addProblemToIgnore(problem));
}
export async function lintConfig(opts) {
    const { severity, externalRefResolver = new BaseResolver(), config } = opts;
    if (!config.document) {
        throw new Error('Config document is not set.');
    }
    const ctx = {
        problems: [],
        specVersion: 'config',
        config,
        visitorsData: {},
    };
    const types = normalizeTypes(opts.externalConfigTypes || createConfigTypes(redoclyConfigSchemaWithoutTheme, config));
    const rules = [
        {
            severity: severity || 'error',
            ruleId: 'configuration struct',
            visitor: Struct({ severity: 'error' }),
        },
        {
            severity: severity || 'error',
            ruleId: 'configuration no-unresolved-refs',
            visitor: NoUnresolvedRefs({ severity: 'error' }),
        },
    ];
    const normalizedVisitors = normalizeVisitors(rules, types);
    const resolvedRefMap = config.resolvedRefMap ||
        (await resolveDocument({
            rootDocument: config.document,
            rootType: types.ConfigRoot,
            externalRefResolver,
        }));
    walkDocument({
        document: config.document,
        rootType: types.ConfigRoot,
        normalizedVisitors,
        resolvedRefMap,
        ctx,
    });
    return ctx.problems;
}
