// @redocly/openapi-core/lib/oas-types.js
import { Arazzo1Types } from './types/arazzo.js';
import { Arazzo1_1Types } from './types/arazzo1_1.js';
import { AsyncApi2Types } from './types/asyncapi2.js';
import { AsyncApi3Types } from './types/asyncapi3.js';
import { Oas2Types } from './types/oas2.js';
import { Oas3Types } from './types/oas3.js';
import { Oas3_1Types } from './types/oas3_1.js';
import { Oas3_2Types } from './types/oas3_2.js';
import { OpenRpcTypes } from './types/openrpc.js';
import { Overlay1Types } from './types/overlay.js';
export const specVersions = [
    'oas2',
    'oas3_0',
    'oas3_1',
    'oas3_2',
    'async2',
    'async3',
    'arazzo1',
    'arazzo1_1',
    'overlay1',
    'openrpc1',
    'graphql',
];
/** Characters allowed in a Components Object key by the OpenAPI and AsyncAPI specs. */
export const COMPONENT_NAME_CHARS = 'a-zA-Z0-9\\.\\-_';
const typesMap = {
    oas2: Oas2Types,
    oas3_0: Oas3Types,
    oas3_1: Oas3_1Types,
    oas3_2: Oas3_2Types,
    async2: AsyncApi2Types,
    async3: AsyncApi3Types,
    arazzo1: Arazzo1Types,
    arazzo1_1: Arazzo1_1Types,
    overlay1: Overlay1Types,
    openrpc1: OpenRpcTypes,
};
export function getTypes(spec) {
    if (spec === 'graphql')
        return {}; // graphql is linted by a separate engine so it has no NodeType
    return typesMap[spec];
}
