// @kubb/plugin-faker/dist/fakerGenerator-Nu5PFwS9.cjs
const require_components = require("./components-IdP8GXXX.cjs");
let _kubb_plugin_oas = require("@kubb/plugin-oas");
let _kubb_plugin_ts = require("@kubb/plugin-ts");
let _kubb_core_hooks = require("@kubb/core/hooks");
let _kubb_plugin_oas_generators = require("@kubb/plugin-oas/generators");
let _kubb_plugin_oas_hooks = require("@kubb/plugin-oas/hooks");
let _kubb_plugin_oas_utils = require("@kubb/plugin-oas/utils");
let _kubb_react_fabric = require("@kubb/react-fabric");
let _kubb_react_fabric_jsx_runtime = require("@kubb/react-fabric/jsx-runtime");
//#region src/generators/fakerGenerator.tsx
const fakerGenerator = (0, _kubb_plugin_oas_generators.createReactGenerator)({
	name: "faker",
	Operation({ operation, generator, plugin }) {
		const { options, options: { dateParser, regexGenerator, seed, mapper } } = plugin;
		const mode = (0, _kubb_core_hooks.useMode)();
		const pluginManager = (0, _kubb_core_hooks.usePluginManager)();
		const oas = (0, _kubb_plugin_oas_hooks.useOas)();
		const { getSchemas, getFile, getGroup } = (0, _kubb_plugin_oas_hooks.useOperationManager)(generator);
		const schemaManager = (0, _kubb_plugin_oas_hooks.useSchemaManager)();
		const file = getFile(operation);
		const schemas = getSchemas(operation);
		const schemaGenerator = new _kubb_plugin_oas.SchemaGenerator(options, {
			fabric: generator.context.fabric,
			oas,
			plugin,
			events: generator.context.events,
			pluginManager,
			mode,
			override: options.override
		});
		const operationSchemas = [
			schemas.pathParams,
			schemas.queryParams,
			schemas.headerParams,
			schemas.statusCodes,
			schemas.request,
			schemas.response
		].flat().filter((x) => Boolean(x));
		const mapOperationSchema = ({ name, schema, description, ...options }) => {
			const transformedSchema = (0, _kubb_plugin_oas_utils.isParameterSchema)(name) && plugin.options.paramsCasing ? (0, _kubb_plugin_oas_utils.applyParamsCasing)(schema, plugin.options.paramsCasing) : schema;
			const tree = schemaGenerator.parse({
				schema: transformedSchema,
				name,
				parentName: null
			});
			const imports = (0, _kubb_plugin_oas_utils.getImports)(tree);
			const group = options.operation ? getGroup(options.operation) : void 0;
			const faker = {
				name: schemaManager.getName(name, { type: "function" }),
				file: schemaManager.getFile(name)
			};
			const type = {
				name: schemaManager.getName(name, {
					type: "type",
					pluginKey: [_kubb_plugin_ts.pluginTsName]
				}),
				file: schemaManager.getFile(options.operationName || name, {
					pluginKey: [_kubb_plugin_ts.pluginTsName],
					group
				})
			};
			const canOverride = tree.some(({ keyword }) => keyword === _kubb_plugin_oas.schemaKeywords.array || keyword === _kubb_plugin_oas.schemaKeywords.and || keyword === _kubb_plugin_oas.schemaKeywords.object || keyword === _kubb_plugin_oas.schemaKeywords.union || keyword === _kubb_plugin_oas.schemaKeywords.tuple || keyword === _kubb_plugin_oas.schemaKeywords.enum || keyword === _kubb_plugin_oas.schemaKeywords.ref);
			return /* @__PURE__ */ (0, _kubb_react_fabric_jsx_runtime.jsxs)(_kubb_react_fabric_jsx_runtime.Fragment, { children: [
				canOverride && /* @__PURE__ */ (0, _kubb_react_fabric_jsx_runtime.jsx)(_kubb_react_fabric.File.Import, {
					isTypeOnly: true,
					root: file.path,
					path: type.file.path,
					name: [type.name]
				}),
				imports.map((imp) => /* @__PURE__ */ (0, _kubb_react_fabric_jsx_runtime.jsx)(_kubb_react_fabric.File.Import, {
					root: file.path,
					path: imp.path,
					name: imp.name
				}, [
					imp.path,
					imp.name,
					imp.isTypeOnly
				].join("-"))),
				/* @__PURE__ */ (0, _kubb_react_fabric_jsx_runtime.jsx)(require_components.Faker, {
					name: faker.name,
					typeName: type.name,
					description,
					tree,
					regexGenerator,
					dateParser,
					mapper,
					seed,
					canOverride
				})
			] });
		};
		return /* @__PURE__ */ (0, _kubb_react_fabric_jsx_runtime.jsxs)(_kubb_react_fabric.File, {
			baseName: file.baseName,
			path: file.path,
			meta: file.meta,
			banner: (0, _kubb_plugin_oas_utils.getBanner)({
				oas,
				output: plugin.options.output,
				config: pluginManager.config
			}),
			footer: (0, _kubb_plugin_oas_utils.getFooter)({
				oas,
				output: plugin.options.output
			}),
			children: [
				/* @__PURE__ */ (0, _kubb_react_fabric_jsx_runtime.jsx)(_kubb_react_fabric.File.Import, {
					name: ["faker"],
					path: "@faker-js/faker"
				}),
				regexGenerator === "randexp" && /* @__PURE__ */ (0, _kubb_react_fabric_jsx_runtime.jsx)(_kubb_react_fabric.File.Import, {
					name: "RandExp",
					path: "randexp"
				}),
				dateParser !== "faker" && /* @__PURE__ */ (0, _kubb_react_fabric_jsx_runtime.jsx)(_kubb_react_fabric.File.Import, {
					path: dateParser,
					name: dateParser
				}),
				operationSchemas.map(mapOperationSchema)
			]
		});
	},
	Schema({ schema, plugin }) {
		const { getName, getFile } = (0, _kubb_plugin_oas_hooks.useSchemaManager)();
		const { options: { output, dateParser, regexGenerator, seed, mapper } } = plugin;
		const pluginManager = (0, _kubb_core_hooks.usePluginManager)();
		const oas = (0, _kubb_plugin_oas_hooks.useOas)();
		const imports = (0, _kubb_plugin_oas_utils.getImports)(schema.tree);
		const faker = {
			name: getName(schema.name, { type: "function" }),
			file: getFile(schema.name)
		};
		const type = {
			name: getName(schema.name, {
				type: "type",
				pluginKey: [_kubb_plugin_ts.pluginTsName]
			}),
			file: getFile(schema.name, { pluginKey: [_kubb_plugin_ts.pluginTsName] })
		};
		const canOverride = schema.tree.some(({ keyword }) => keyword === _kubb_plugin_oas.schemaKeywords.array || keyword === _kubb_plugin_oas.schemaKeywords.and || keyword === _kubb_plugin_oas.schemaKeywords.object || keyword === _kubb_plugin_oas.schemaKeywords.union || keyword === _kubb_plugin_oas.schemaKeywords.tuple || keyword === _kubb_plugin_oas.schemaKeywords.ref || keyword === _kubb_plugin_oas.schemaKeywords.enum || keyword === _kubb_plugin_oas.schemaKeywords.string || keyword === _kubb_plugin_oas.schemaKeywords.integer || keyword === _kubb_plugin_oas.schemaKeywords.number);
		return /* @__PURE__ */ (0, _kubb_react_fabric_jsx_runtime.jsxs)(_kubb_react_fabric.File, {
			baseName: faker.file.baseName,
			path: faker.file.path,
			meta: faker.file.meta,
			banner: (0, _kubb_plugin_oas_utils.getBanner)({
				oas,
				output,
				config: pluginManager.config
			}),
			footer: (0, _kubb_plugin_oas_utils.getFooter)({
				oas,
				output
			}),
			children: [
				/* @__PURE__ */ (0, _kubb_react_fabric_jsx_runtime.jsx)(_kubb_react_fabric.File.Import, {
					name: ["faker"],
					path: "@faker-js/faker"
				}),
				regexGenerator === "randexp" && /* @__PURE__ */ (0, _kubb_react_fabric_jsx_runtime.jsx)(_kubb_react_fabric.File.Import, {
					name: "RandExp",
					path: "randexp"
				}),
				dateParser !== "faker" && /* @__PURE__ */ (0, _kubb_react_fabric_jsx_runtime.jsx)(_kubb_react_fabric.File.Import, {
					path: dateParser,
					name: dateParser
				}),
				/* @__PURE__ */ (0, _kubb_react_fabric_jsx_runtime.jsx)(_kubb_react_fabric.File.Import, {
					isTypeOnly: true,
					root: faker.file.path,
					path: type.file.path,
					name: [type.name]
				}),
				imports.map((imp) => /* @__PURE__ */ (0, _kubb_react_fabric_jsx_runtime.jsx)(_kubb_react_fabric.File.Import, {
					root: faker.file.path,
					path: imp.path,
					name: imp.name
				}, [
					imp.path,
					imp.name,
					imp.isTypeOnly
				].join("-"))),
				/* @__PURE__ */ (0, _kubb_react_fabric_jsx_runtime.jsx)(require_components.Faker, {
					name: faker.name,
					typeName: type.name,
					description: schema.value.description,
					tree: schema.tree,
					regexGenerator,
					dateParser,
					mapper,
					seed,
					canOverride
				})
			]
		});
	}
});
//#endregion
Object.defineProperty(exports, "fakerGenerator", {
	enumerable: true,
	get: function() {
		return fakerGenerator;
	}
});
