// @kubb/plugin-faker/dist/index.cjs
const require_components = require('./components-BHsLXkVC.cjs');
const require_fakerGenerator = require('./fakerGenerator-BgUuFQPq.cjs');
let node_path = require("node:path");
node_path = require_components.__toESM(node_path);
let __kubb_core = require("@kubb/core");
__kubb_core = require_components.__toESM(__kubb_core);
let __kubb_core_transformers = require("@kubb/core/transformers");
__kubb_core_transformers = require_components.__toESM(__kubb_core_transformers);
let __kubb_plugin_oas = require("@kubb/plugin-oas");
__kubb_plugin_oas = require_components.__toESM(__kubb_plugin_oas);
let __kubb_plugin_ts = require("@kubb/plugin-ts");
__kubb_plugin_ts = require_components.__toESM(__kubb_plugin_ts);

//#region src/plugin.ts
const pluginFakerName = "plugin-faker";
const pluginFaker = (0, __kubb_core.createPlugin)((options) => {
	const { output = {
		path: "mocks",
		barrelType: "named"
	}, seed, group, exclude = [], include, override = [], transformers = {}, mapper = {}, unknownType = "any", emptySchemaType = unknownType, dateType = "string", dateParser = "faker", generators = [require_fakerGenerator.fakerGenerator].filter(Boolean), regexGenerator = "faker", contentType } = options;
	return {
		name: pluginFakerName,
		options: {
			output,
			transformers,
			seed,
			dateType,
			unknownType,
			emptySchemaType,
			dateParser,
			mapper,
			override,
			regexGenerator,
			group
		},
		pre: [__kubb_plugin_oas.pluginOasName, __kubb_plugin_ts.pluginTsName],
		resolvePath(baseName, pathMode, options$1) {
			const root = node_path.default.resolve(this.config.root, this.config.output.path);
			if ((pathMode ?? (0, __kubb_core.getMode)(node_path.default.resolve(root, output.path))) === "single")
 /**
			* when output is a file then we will always append to the same file(output file), see fileManager.addOrAppend
			* Other plugins then need to call addOrAppend instead of just add from the fileManager class
			*/
			return node_path.default.resolve(root, output.path);
			if (group && (options$1?.group?.path || options$1?.group?.tag)) {
				const groupName = group?.name ? group.name : (ctx) => {
					if (group?.type === "path") return `${ctx.group.split("/")[1]}`;
					return `${(0, __kubb_core_transformers.camelCase)(ctx.group)}Controller`;
				};
				return node_path.default.resolve(root, output.path, groupName({ group: group.type === "path" ? options$1.group.path : options$1.group.tag }), baseName);
			}
			return node_path.default.resolve(root, output.path, baseName);
		},
		resolveName(name, type) {
			const resolvedName = (0, __kubb_core_transformers.camelCase)(name, {
				prefix: type ? "create" : void 0,
				isFile: type === "file"
			});
			if (type) return transformers?.name?.(resolvedName, type) || resolvedName;
			return resolvedName;
		},
		async buildStart() {
			const [swaggerPlugin] = __kubb_core.PluginManager.getDependedPlugins(this.plugins, [__kubb_plugin_oas.pluginOasName]);
			const oas = await swaggerPlugin.context.getOas();
			const root = node_path.default.resolve(this.config.root, this.config.output.path);
			const mode = (0, __kubb_core.getMode)(node_path.default.resolve(root, output.path));
			const schemaFiles = await new __kubb_plugin_oas.SchemaGenerator(this.plugin.options, {
				oas,
				pluginManager: this.pluginManager,
				plugin: this.plugin,
				contentType,
				include: void 0,
				override,
				mode,
				output: output.path
			}).build(...generators);
			await this.addFile(...schemaFiles);
			const operationFiles = await new __kubb_plugin_oas.OperationGenerator(this.plugin.options, {
				oas,
				pluginManager: this.pluginManager,
				plugin: this.plugin,
				contentType,
				exclude,
				include,
				override,
				mode
			}).build(...generators);
			await this.addFile(...operationFiles);
			const barrelFiles = await (0, __kubb_core.getBarrelFiles)(this.fileManager.files, {
				type: output.barrelType ?? "named",
				root,
				output,
				meta: { pluginKey: this.plugin.key },
				logger: this.logger
			});
			await this.addFile(...barrelFiles);
		}
	};
});

//#endregion
exports.pluginFaker = pluginFaker;
exports.pluginFakerName = pluginFakerName;
