// @kubb/plugin-oas/dist/hooks.cjs
const require_chunk = require('./chunk-CUT6urMc.cjs');
const require_SchemaGenerator = require('./SchemaGenerator-DsK0wuii.cjs');
const require_Oas = require('./Oas--cvTRMTA.cjs');
const require_SchemaMapper = require('./SchemaMapper-CpAbmGiH.cjs');
require('./getSchemas-08CU8KWq.cjs');
let __kubb_react = require("@kubb/react");
__kubb_react = require_chunk.__toESM(__kubb_react);
let __kubb_core_hooks = require("@kubb/core/hooks");
__kubb_core_hooks = require_chunk.__toESM(__kubb_core_hooks);

//#region src/hooks/useOas.ts
function useOas() {
	const { oas } = (0, __kubb_react.useContext)(require_Oas.Oas.Context);
	if (!oas) throw new Error("Oas is not defined");
	return oas;
}

//#endregion
//#region src/hooks/useOperation.ts
/**
* `useOperation` will return the current `Operation`
*/
function useOperation() {
	const { operation } = (0, __kubb_react.useContext)(require_Oas.Operation.Context);
	if (!operation) throw new Error("Operation is not defined");
	return operation;
}

//#endregion
//#region src/hooks/useOperationManager.ts
/**
* `useOperationManager` will return some helper functions that can be used to get the operation file, get the operation name.
*/
function useOperationManager() {
	const plugin = (0, __kubb_core_hooks.usePlugin)();
	const pluginManager = (0, __kubb_core_hooks.usePluginManager)();
	const { generator } = (0, __kubb_react.useContext)(require_Oas.Oas.Context);
	const getName = (operation, { prefix = "", suffix = "", pluginKey = plugin.key, type }) => {
		return pluginManager.resolveName({
			name: `${prefix} ${operation.getOperationId()} ${suffix}`,
			pluginKey,
			type
		});
	};
	const getGroup = (operation) => {
		return {
			tag: operation.getTags().at(0)?.name,
			path: operation.path
		};
	};
	const getSchemas = (operation, params) => {
		if (!generator) throw new Error(`'generator' is not defined`);
		return generator.getSchemas(operation, { resolveName: (name) => pluginManager.resolveName({
			name,
			pluginKey: params?.pluginKey,
			type: params?.type
		}) });
	};
	const getFile = (operation, { prefix, suffix, pluginKey = plugin.key, extname = ".ts" } = {}) => {
		const name = getName(operation, {
			type: "file",
			pluginKey,
			prefix,
			suffix
		});
		const group = getGroup(operation);
		const file = pluginManager.getFile({
			name,
			extname,
			pluginKey,
			options: {
				type: "file",
				pluginKey,
				group
			}
		});
		return {
			...file,
			meta: {
				...file.meta,
				name,
				pluginKey,
				group
			}
		};
	};
	const groupSchemasByName = (operation, { pluginKey = plugin.key, type }) => {
		if (!generator) throw new Error(`'generator' is not defined`);
		const schemas = generator.getSchemas(operation);
		const errors = (schemas.errors || []).reduce((prev, acc) => {
			if (!acc.statusCode) return prev;
			prev[acc.statusCode] = pluginManager.resolveName({
				name: acc.name,
				pluginKey,
				type
			});
			return prev;
		}, {});
		const responses = (schemas.responses || []).reduce((prev, acc) => {
			if (!acc.statusCode) return prev;
			prev[acc.statusCode] = pluginManager.resolveName({
				name: acc.name,
				pluginKey,
				type
			});
			return prev;
		}, {});
		return {
			request: schemas.request?.name ? pluginManager.resolveName({
				name: schemas.request.name,
				pluginKey,
				type
			}) : void 0,
			parameters: {
				path: schemas.pathParams?.name ? pluginManager.resolveName({
					name: schemas.pathParams.name,
					pluginKey,
					type
				}) : void 0,
				query: schemas.queryParams?.name ? pluginManager.resolveName({
					name: schemas.queryParams.name,
					pluginKey,
					type
				}) : void 0,
				header: schemas.headerParams?.name ? pluginManager.resolveName({
					name: schemas.headerParams.name,
					pluginKey,
					type
				}) : void 0
			},
			responses: {
				...responses,
				["default"]: pluginManager.resolveName({
					name: schemas.response.name,
					pluginKey,
					type
				}),
				...errors
			},
			errors
		};
	};
	return {
		getName,
		getFile,
		getSchemas,
		groupSchemasByName,
		getGroup
	};
}

//#endregion
//#region src/hooks/useOperations.ts
/**
* `useOperations` will return all the Operations
*/
function useOperations({ method, path } = {}) {
	const { operations } = (0, __kubb_react.useContext)(require_Oas.Oas.Context);
	if (!operations) throw new Error("Operations is not defined");
	let items = operations;
	if (path) items = items.filter((item) => item.path === path);
	if (method) items = items.filter((item) => item.method === method);
	return items;
}

//#endregion
//#region src/hooks/useSchema.ts
/**
* `useSchema` will return the current `schema properties`
*/
function useSchema() {
	return (0, __kubb_react.useContext)(require_Oas.Schema.Context);
}

//#endregion
//#region src/hooks/useSchemaManager.ts
/**
* `useSchemaManager` will return some helper functions that can be used to get the schema file, get the schema name.
*/
function useSchemaManager() {
	const plugin = (0, __kubb_core_hooks.usePlugin)();
	const pluginManager = (0, __kubb_core_hooks.usePluginManager)();
	const getName = (name, { pluginKey = plugin.key, type }) => {
		return pluginManager.resolveName({
			name,
			pluginKey,
			type
		});
	};
	const getFile = (name, { mode = "split", pluginKey = plugin.key, extname = ".ts", group } = {}) => {
		const resolvedName = mode === "single" ? "" : getName(name, {
			type: "file",
			pluginKey
		});
		const file = pluginManager.getFile({
			name: resolvedName,
			extname,
			pluginKey,
			options: {
				type: "file",
				pluginKey,
				group
			}
		});
		return {
			...file,
			meta: {
				...file.meta,
				name: resolvedName,
				pluginKey
			}
		};
	};
	const getImports = (tree) => {
		return require_SchemaGenerator.SchemaGenerator.deepSearch(tree, require_SchemaMapper.schemaKeywords.ref)?.map((item) => {
			if (!item.args.path || !item.args.isImportable) return;
			return {
				name: [item.args.name],
				path: item.args.path
			};
		}).filter(Boolean);
	};
	return {
		getName,
		getFile,
		getImports
	};
}

//#endregion
exports.useOas = useOas;
exports.useOperation = useOperation;
exports.useOperationManager = useOperationManager;
exports.useOperations = useOperations;
exports.useSchema = useSchema;
exports.useSchemaManager = useSchemaManager;
