// @kubb/plugin-oas/dist/mocks.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_SchemaMapper = require("./SchemaMapper-CeavHZlp.cjs");
const schemas = {
	basic: [
		{
			name: "any",
			schema: { keyword: require_SchemaMapper.schemaKeywords.any }
		},
		{
			name: "unknown",
			schema: { keyword: require_SchemaMapper.schemaKeywords.unknown }
		},
		{
			name: "string",
			schema: { keyword: require_SchemaMapper.schemaKeywords.string }
		},
		{
			name: "number",
			schema: { keyword: require_SchemaMapper.schemaKeywords.number }
		},
		{
			name: "integer",
			schema: { keyword: require_SchemaMapper.schemaKeywords.integer }
		},
		{
			name: "boolean",
			schema: { keyword: require_SchemaMapper.schemaKeywords.boolean }
		},
		{
			name: "primitiveDate",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.date,
				args: { type: "date" }
			}
		},
		{
			name: "date",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.date,
				args: { type: "string" }
			}
		},
		{
			name: "time",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.time,
				args: { type: "string" }
			}
		},
		{
			name: "stringOffset",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.datetime,
				args: { offset: true }
			}
		},
		{
			name: "stringLocal",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.datetime,
				args: { local: true }
			}
		},
		{
			name: "datetime",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.datetime,
				args: { offset: false }
			}
		},
		{
			name: "nullable",
			schema: { keyword: require_SchemaMapper.schemaKeywords.nullable }
		},
		{
			name: "undefined",
			schema: { keyword: require_SchemaMapper.schemaKeywords.undefined }
		},
		{
			name: "min",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.min,
				args: 2
			}
		},
		{
			name: "max",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.max,
				args: 2
			}
		},
		{
			name: "matchesReg",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.matches,
				args: "/node_modules/"
			}
		},
		{
			name: "matches",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.matches,
				args: "^[A-Z]{2}$"
			}
		},
		{
			name: "const",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.const,
				args: {
					name: "",
					value: "",
					format: require_SchemaMapper.schemaKeywords.string
				}
			}
		},
		{
			name: "ref",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.ref,
				args: {
					$ref: "$ref",
					name: "Pet",
					path: "./pet.ts",
					isImportable: true
				}
			}
		},
		{
			name: "enum",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.enum,
				args: {
					name: "enum",
					typeName: "Enum",
					asConst: false,
					items: [
						{
							name: "A",
							value: "A",
							format: require_SchemaMapper.schemaKeywords.string
						},
						{
							name: "B",
							value: "B",
							format: require_SchemaMapper.schemaKeywords.string
						},
						{
							name: "C",
							value: "C",
							format: require_SchemaMapper.schemaKeywords.string
						},
						{
							name: 2,
							value: 2,
							format: require_SchemaMapper.schemaKeywords.number
						}
					]
				}
			}
		},
		{
			name: "enumLiteralBoolean",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.enum,
				args: {
					asConst: true,
					items: [{
						format: "boolean",
						name: "true",
						value: true
					}, {
						format: "boolean",
						name: "false",
						value: false
					}],
					name: "PetEnumLiteral",
					typeName: "PetEnumLiteral"
				}
			}
		},
		{
			name: "tuple",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.tuple,
				args: { items: [] }
			}
		},
		{
			name: "tupleMulti",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.tuple,
				args: { items: [{ keyword: require_SchemaMapper.schemaKeywords.string }, { keyword: require_SchemaMapper.schemaKeywords.number }] }
			}
		},
		{
			name: "array",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.array,
				args: { items: [{
					keyword: require_SchemaMapper.schemaKeywords.union,
					args: [{ keyword: require_SchemaMapper.schemaKeywords.number }, { keyword: require_SchemaMapper.schemaKeywords.string }]
				}] }
			}
		},
		{
			name: "arrayEmpty",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.array,
				args: { items: [] }
			}
		},
		{
			name: "arrayRef",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.array,
				args: { items: [{
					keyword: require_SchemaMapper.schemaKeywords.ref,
					args: {
						name: "Pet",
						$ref: "#component/schema/Pet",
						path: "./pet.ts",
						isImportable: true
					}
				}] }
			}
		},
		{
			name: "arrayAdvanced",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.array,
				args: {
					items: [
						{
							keyword: require_SchemaMapper.schemaKeywords.min,
							args: 1
						},
						{
							keyword: require_SchemaMapper.schemaKeywords.max,
							args: 10
						},
						{ keyword: require_SchemaMapper.schemaKeywords.number }
					],
					min: 3,
					max: 10
				}
			}
		},
		{
			name: "arrayRegex",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.array,
				args: {
					items: [{
						keyword: require_SchemaMapper.schemaKeywords.matches,
						args: "^[a-zA-Z0-9]{1,13}$"
					}],
					min: 3,
					max: 10
				}
			}
		},
		{
			name: "union",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.union,
				args: [{ keyword: require_SchemaMapper.schemaKeywords.string }, { keyword: require_SchemaMapper.schemaKeywords.number }]
			}
		},
		{
			name: "unionOne",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.union,
				args: [{ keyword: require_SchemaMapper.schemaKeywords.string }]
			}
		},
		{
			name: "catchall",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.object,
				args: {
					properties: {},
					additionalProperties: [{
						keyword: require_SchemaMapper.schemaKeywords.ref,
						args: {
							name: "Pet",
							$ref: "#component/schema/Pet",
							path: "./Pet.ts",
							isImportable: true
						}
					}]
				}
			}
		},
		{
			name: "and",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.and,
				args: [{
					keyword: require_SchemaMapper.schemaKeywords.object,
					args: {
						properties: { street: [{ keyword: require_SchemaMapper.schemaKeywords.string }] },
						additionalProperties: []
					}
				}, {
					keyword: require_SchemaMapper.schemaKeywords.object,
					args: {
						properties: { city: [{ keyword: require_SchemaMapper.schemaKeywords.string }] },
						additionalProperties: []
					}
				}]
			}
		},
		{
			name: "object",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.object,
				args: {
					properties: {
						firstName: [{ keyword: require_SchemaMapper.schemaKeywords.string }, {
							keyword: require_SchemaMapper.schemaKeywords.min,
							args: 2
						}],
						address: [
							{ keyword: require_SchemaMapper.schemaKeywords.string },
							{ keyword: require_SchemaMapper.schemaKeywords.nullable },
							{
								keyword: require_SchemaMapper.schemaKeywords.describe,
								args: "\"Your address\""
							}
						]
					},
					additionalProperties: []
				}
			}
		},
		{
			name: "objectOptional",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.object,
				args: {
					properties: {
						firstName: [
							{ keyword: require_SchemaMapper.schemaKeywords.string },
							{ keyword: require_SchemaMapper.schemaKeywords.optional },
							{
								keyword: require_SchemaMapper.schemaKeywords.min,
								args: 2
							}
						],
						address: [
							{ keyword: require_SchemaMapper.schemaKeywords.string },
							{ keyword: require_SchemaMapper.schemaKeywords.nullable },
							{
								keyword: require_SchemaMapper.schemaKeywords.describe,
								args: "\"Your address\""
							}
						]
					},
					additionalProperties: []
				}
			}
		},
		{
			name: "objectArray",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.object,
				args: {
					properties: { ids: [{
						keyword: require_SchemaMapper.schemaKeywords.array,
						args: {
							items: [{
								keyword: require_SchemaMapper.schemaKeywords.matches,
								args: "^[a-zA-Z0-9]{1,13}$"
							}],
							min: 3,
							max: 10
						}
					}] },
					additionalProperties: []
				}
			}
		},
		{
			name: "objectDates",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.object,
				args: {
					properties: {
						dateTime: [{
							keyword: require_SchemaMapper.schemaKeywords.datetime,
							args: { offset: true }
						}],
						date: [{
							keyword: require_SchemaMapper.schemaKeywords.date,
							args: { type: "string" }
						}],
						time: [{
							keyword: require_SchemaMapper.schemaKeywords.time,
							args: { type: "string" }
						}],
						nativeDate: [{
							keyword: require_SchemaMapper.schemaKeywords.date,
							args: { type: "date" }
						}]
					},
					additionalProperties: []
				}
			}
		},
		{
			name: "objectAnd",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.object,
				args: {
					properties: {
						firstName: [
							{ keyword: require_SchemaMapper.schemaKeywords.deprecated },
							{
								keyword: require_SchemaMapper.schemaKeywords.default,
								args: "test"
							},
							{
								keyword: require_SchemaMapper.schemaKeywords.min,
								args: 2
							},
							{ keyword: require_SchemaMapper.schemaKeywords.string }
						],
						age: [
							{
								keyword: require_SchemaMapper.schemaKeywords.example,
								args: "2"
							},
							{
								keyword: require_SchemaMapper.schemaKeywords.default,
								args: 2
							},
							{
								keyword: require_SchemaMapper.schemaKeywords.min,
								args: 3
							},
							{ keyword: require_SchemaMapper.schemaKeywords.number }
						],
						address: [
							{
								keyword: require_SchemaMapper.schemaKeywords.and,
								args: [{
									keyword: require_SchemaMapper.schemaKeywords.object,
									args: {
										properties: { street: [{ keyword: require_SchemaMapper.schemaKeywords.string }] },
										additionalProperties: []
									}
								}, {
									keyword: require_SchemaMapper.schemaKeywords.object,
									args: {
										properties: { city: [{ keyword: require_SchemaMapper.schemaKeywords.string }] },
										additionalProperties: []
									}
								}]
							},
							{ keyword: require_SchemaMapper.schemaKeywords.nullable },
							{
								keyword: require_SchemaMapper.schemaKeywords.describe,
								args: "Your address"
							}
						]
					},
					additionalProperties: []
				}
			}
		},
		{
			name: "objectEnum",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.object,
				args: {
					properties: { version: [
						{
							keyword: require_SchemaMapper.schemaKeywords.schema,
							args: {
								format: "string",
								type: "string"
							}
						},
						{
							keyword: require_SchemaMapper.schemaKeywords.enum,
							args: {
								name: "enum",
								typeName: "Enum",
								asConst: false,
								items: [
									{
										name: "A",
										value: "A",
										format: require_SchemaMapper.schemaKeywords.string
									},
									{
										name: "B",
										value: "B",
										format: require_SchemaMapper.schemaKeywords.string
									},
									{
										name: "C",
										value: "C",
										format: require_SchemaMapper.schemaKeywords.string
									},
									{
										name: 2,
										value: 2,
										format: require_SchemaMapper.schemaKeywords.number
									}
								]
							}
						},
						{
							keyword: require_SchemaMapper.schemaKeywords.min,
							args: 4
						},
						{
							keyword: require_SchemaMapper.schemaKeywords.describe,
							args: "Your address"
						}
					] },
					additionalProperties: []
				}
			}
		},
		{
			name: "objectObjectEnum",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.object,
				args: {
					properties: { prop1: [{
						keyword: require_SchemaMapper.schemaKeywords.object,
						args: {
							properties: { prop2: [{
								keyword: require_SchemaMapper.schemaKeywords.schema,
								args: {
									format: "string",
									type: "string"
								}
							}, {
								keyword: require_SchemaMapper.schemaKeywords.enum,
								args: {
									name: "enum",
									typeName: "Enum",
									asConst: false,
									items: [{
										name: "A",
										value: "A",
										format: require_SchemaMapper.schemaKeywords.string
									}, {
										name: "B",
										value: "B",
										format: require_SchemaMapper.schemaKeywords.string
									}]
								}
							}] },
							additionalProperties: []
						}
					}] },
					additionalProperties: []
				}
			}
		},
		{
			name: "objectArrayObject",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.object,
				args: {
					properties: { ids: [{
						keyword: require_SchemaMapper.schemaKeywords.array,
						args: {
							items: [{
								keyword: require_SchemaMapper.schemaKeywords.object,
								args: {
									properties: { enum: [{
										keyword: require_SchemaMapper.schemaKeywords.schema,
										args: {
											format: "string",
											type: "string"
										}
									}, {
										keyword: require_SchemaMapper.schemaKeywords.enum,
										args: {
											name: "enum",
											typeName: "Enum",
											asConst: false,
											items: [
												{
													name: "A",
													value: "A",
													format: require_SchemaMapper.schemaKeywords.string
												},
												{
													name: "B",
													value: "B",
													format: require_SchemaMapper.schemaKeywords.string
												},
												{
													name: "C",
													value: "C",
													format: require_SchemaMapper.schemaKeywords.string
												},
												{
													name: 2,
													value: 2,
													format: require_SchemaMapper.schemaKeywords.number
												}
											]
										}
									}] },
									additionalProperties: []
								}
							}],
							min: 3,
							max: 10
						}
					}] },
					additionalProperties: []
				}
			}
		},
		{
			name: "objectEmpty",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.object,
				args: {
					properties: {},
					additionalProperties: []
				}
			}
		},
		{
			name: "default",
			schema: { keyword: require_SchemaMapper.schemaKeywords.default }
		},
		{
			name: "default",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.default,
				args: "default"
			}
		},
		{
			name: "blob",
			schema: { keyword: require_SchemaMapper.schemaKeywords.blob }
		},
		{
			name: "nullableAdditionalProperties",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.object,
				args: {
					properties: {},
					additionalProperties: [
						{ keyword: require_SchemaMapper.schemaKeywords.string },
						{
							args: {
								format: void 0,
								type: require_SchemaMapper.schemaKeywords.string
							},
							keyword: require_SchemaMapper.schemaKeywords.schema
						},
						{ keyword: require_SchemaMapper.schemaKeywords.nullable }
					]
				}
			}
		},
		{
			name: "andRefEnum",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.and,
				args: [{
					keyword: require_SchemaMapper.schemaKeywords.ref,
					args: {
						$ref: "#/components/schemas/IssueCategory",
						name: "createIssueCategory",
						path: "./createIssueCategory.ts",
						isImportable: true
					}
				}]
			}
		},
		{
			name: "objectWithNullableEnumRef",
			schema: {
				keyword: require_SchemaMapper.schemaKeywords.object,
				args: {
					properties: { category: [{
						keyword: require_SchemaMapper.schemaKeywords.and,
						args: [{
							keyword: require_SchemaMapper.schemaKeywords.ref,
							args: {
								$ref: "#/components/schemas/IssueCategory",
								name: "createIssueCategory",
								path: "./createIssueCategory.ts",
								isImportable: true
							}
						}]
					}] },
					additionalProperties: []
				}
			}
		}
	],
	full: [
		{
			name: "Upload",
			schema: [{ keyword: require_SchemaMapper.schemaKeywords.blob }]
		},
		{
			name: "PageSizeNumber",
			schema: [{ keyword: require_SchemaMapper.schemaKeywords.number }, {
				keyword: require_SchemaMapper.schemaKeywords.default,
				args: 10
			}]
		},
		{
			name: "PageSizeInteger",
			schema: [{ keyword: require_SchemaMapper.schemaKeywords.integer }, {
				keyword: require_SchemaMapper.schemaKeywords.default,
				args: 10
			}]
		},
		{
			name: "Object",
			schema: [
				{ keyword: require_SchemaMapper.schemaKeywords.nullable },
				{
					keyword: require_SchemaMapper.schemaKeywords.describe,
					args: "Your address"
				},
				{
					keyword: require_SchemaMapper.schemaKeywords.object,
					args: {
						properties: {
							firstName: [
								{ keyword: require_SchemaMapper.schemaKeywords.deprecated },
								{
									keyword: require_SchemaMapper.schemaKeywords.default,
									args: "test"
								},
								{
									keyword: require_SchemaMapper.schemaKeywords.min,
									args: 2
								},
								{ keyword: require_SchemaMapper.schemaKeywords.string }
							],
							age: [
								{
									keyword: require_SchemaMapper.schemaKeywords.example,
									args: "2"
								},
								{
									keyword: require_SchemaMapper.schemaKeywords.default,
									args: 2
								},
								{
									keyword: require_SchemaMapper.schemaKeywords.min,
									args: 2
								},
								{ keyword: require_SchemaMapper.schemaKeywords.number }
							],
							address: [
								{
									keyword: require_SchemaMapper.schemaKeywords.and,
									args: [{ keyword: require_SchemaMapper.schemaKeywords.string }, {
										keyword: require_SchemaMapper.schemaKeywords.object,
										args: {
											properties: {
												street: [{ keyword: require_SchemaMapper.schemaKeywords.string }],
												city: [{ keyword: require_SchemaMapper.schemaKeywords.string }]
											},
											additionalProperties: []
										}
									}]
								},
								{ keyword: require_SchemaMapper.schemaKeywords.nullable },
								{
									keyword: require_SchemaMapper.schemaKeywords.describe,
									args: "Your address"
								}
							]
						},
						additionalProperties: []
					}
				}
			]
		},
		{
			name: "Order",
			schema: [{
				keyword: require_SchemaMapper.schemaKeywords.schema,
				args: { type: "object" }
			}, {
				keyword: require_SchemaMapper.schemaKeywords.object,
				args: {
					properties: { status: [{
						keyword: require_SchemaMapper.schemaKeywords.enum,
						args: {
							name: "orderStatus",
							asConst: false,
							typeName: "OrderStatus",
							items: [{
								name: "Placed",
								value: "placed",
								format: "string"
							}, {
								name: "Approved",
								value: "approved",
								format: "string"
							}]
						}
					}] },
					additionalProperties: []
				}
			}]
		},
		{
			name: "nullableAdditionalProperties",
			schema: [{
				keyword: require_SchemaMapper.schemaKeywords.object,
				args: {
					properties: {},
					additionalProperties: [
						{ keyword: require_SchemaMapper.schemaKeywords.string },
						{
							args: {
								format: void 0,
								type: require_SchemaMapper.schemaKeywords.string
							},
							keyword: require_SchemaMapper.schemaKeywords.schema
						},
						{ keyword: require_SchemaMapper.schemaKeywords.number }
					]
				}
			}]
		},
		{
			name: "Record",
			schema: [
				{
					keyword: require_SchemaMapper.schemaKeywords.object,
					args: {
						properties: {},
						additionalProperties: [
							{ keyword: require_SchemaMapper.schemaKeywords.integer },
							{
								keyword: require_SchemaMapper.schemaKeywords.schema,
								args: {
									type: "integer",
									format: "int32"
								}
							},
							{ keyword: require_SchemaMapper.schemaKeywords.optional }
						]
					}
				},
				{
					keyword: require_SchemaMapper.schemaKeywords.schema,
					args: {
						type: "integer",
						format: "int32"
					}
				},
				{ keyword: require_SchemaMapper.schemaKeywords.optional }
			]
		}
	]
};
//#endregion
exports.schemas = schemas;
