// @kubb/plugin-oas/dist/SchemaMapper-CqMkO2T1.js
import "./chunk--u3MIqq1.js";
//#region src/SchemaMapper.ts
const schemaKeywords = {
	any: "any",
	unknown: "unknown",
	number: "number",
	integer: "integer",
	bigint: "bigint",
	string: "string",
	boolean: "boolean",
	undefined: "undefined",
	nullable: "nullable",
	null: "null",
	nullish: "nullish",
	array: "array",
	tuple: "tuple",
	enum: "enum",
	union: "union",
	datetime: "datetime",
	date: "date",
	email: "email",
	uuid: "uuid",
	url: "url",
	void: "void",
	default: "default",
	const: "const",
	and: "and",
	describe: "describe",
	min: "min",
	max: "max",
	exclusiveMinimum: "exclusiveMinimum",
	exclusiveMaximum: "exclusiveMaximum",
	optional: "optional",
	readOnly: "readOnly",
	writeOnly: "writeOnly",
	object: "object",
	ref: "ref",
	matches: "matches",
	firstName: "firstName",
	lastName: "lastName",
	password: "password",
	phone: "phone",
	blob: "blob",
	deprecated: "deprecated",
	example: "example",
	schema: "schema",
	catchall: "catchall",
	time: "time",
	name: "name",
	interface: "interface"
};
function isKeyword(meta, keyword) {
	return meta.keyword === keyword;
}
//#endregion
export { schemaKeywords as n, isKeyword as t };
