// @kubb/plugin-oas/dist/utils.js
import { n as getBanner, t as getFooter } from "./getFooter-CvTKkK7-.js";
import { n as getSchemaFactory, t as getSchemas } from "./getSchemas-CvUX0jWa.js";
import { t as parseFromConfig } from "./parseFromConfig-DSI6Meg0.js";
import transformers, { camelCase, isValidVarName } from "@kubb/core/transformers";
import { URLPath } from "@kubb/core/utils";
import { isOptional, isParameterObject } from "@kubb/oas";

//#region src/utils/getComments.ts
function getComments(operation) {
	return [
		operation.getDescription() && `@description ${operation.getDescription()}`,
		operation.getSummary() && `@summary ${operation.getSummary()}`,
		operation.path && `{@link ${new URLPath(operation.path).URL}}`,
		operation.isDeprecated() && "@deprecated"
	].filter(Boolean).map((text) => transformers.trim(text));
}

//#endregion
//#region src/utils/getParams.ts
/**
*
* @deprecated
* TODO move to operationManager hook
*/
function getASTParams(operationSchema, { typed = false, override } = {}) {
	if (!operationSchema || !operationSchema.schema.properties || !operationSchema.name) return [];
	return Object.entries(operationSchema.schema.properties).map(([name, schema]) => {
		const isParam = isParameterObject(schema);
		const data = {
			name,
			enabled: !!name,
			required: isParam ? schema.required : true,
			type: typed ? `${operationSchema.name}["${name}"]` : void 0
		};
		return override ? override(data) : data;
	});
}
function getPathParams(operationSchema, options = {}) {
	return getASTParams(operationSchema, options).reduce((acc, curr) => {
		if (curr.name && curr.enabled) {
			let name = isValidVarName(curr.name) ? curr.name : camelCase(curr.name);
			if (options.casing === "camelcase") name = camelCase(name);
			acc[name] = {
				default: curr.default,
				type: curr.type,
				optional: !curr.required
			};
		}
		return acc;
	}, {});
}

//#endregion
export { getBanner, getComments, getFooter, getPathParams, getSchemaFactory, getSchemas, isOptional, parseFromConfig };
