// @redocly/config/lib/asyncapi-config-schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.asyncapiConfigSchema = void 0;
const feedback_config_schema_1 = require("./feedback-config-schema");
const redoc_config_schema_1 = require("./redoc-config-schema");
exports.asyncapiConfigSchema = {
    type: 'object',
    properties: {
        downloadUrls: redoc_config_schema_1.downloadUrlsSchema,
        apiLogo: {
            type: 'object',
            properties: {
                imageUrl: {
                    type: 'string',
                },
                href: {
                    type: 'string',
                },
                altText: {
                    type: 'string',
                },
                backgroundColor: {
                    type: 'string',
                },
            },
        },
        jsonSamplesDepth: {
            type: 'number',
        },
        samplesMaxInlineArgs: {
            type: 'number',
        },
        fieldExpandLevel: {
            type: 'number',
        },
        baseUrlPath: {
            type: 'string',
        },
        metadata: {
            type: 'object',
            properties: {
                apiId: {
                    type: 'string',
                },
            },
            additionalProperties: true,
        },
        feedback: feedback_config_schema_1.feedbackConfigSchema,
        layout: { type: 'string', enum: ['stacked', 'three-panel'] },
    },
    additionalProperties: false,
};
