// @redocly/openapi-core/lib/rules/oas3/no-invalid-media-type-examples.js
import { isRef } from '../../ref-utils.js';
import { isDefined } from '../../utils/is-defined.js';
import { isPlainObject } from '../../utils/is-plain-object.js';
import { validateExample } from '../utils.js';
export const ValidContentExamples = (opts) => {
    const skip = (mediaType) => {
        return mediaType.schema === undefined;
    };
    const leave = (context) => (mediaType, ctx) => {
        const { location, resolve } = ctx;
        if (isDefined(mediaType.example)) {
            resolveAndValidateExample(mediaType.example, location.child('example'));
        }
        else if (isPlainObject(mediaType.examples)) {
            for (const exampleName of Object.keys(mediaType.examples)) {
                resolveAndValidateExample(mediaType.examples[exampleName], location.child(['examples', exampleName, 'value']), true);
            }
        }
        function resolveAndValidateExample(example, location, isMultiple) {
            if (isRef(example)) {
                const resolved = resolve(example);
                if (!resolved.location)
                    return;
                location = isMultiple ? resolved.location.child('value') : resolved.location;
                example = resolved.node;
            }
            if (isMultiple && typeof example?.value === 'undefined') {
                return;
            }
            validateExample(isMultiple ? example.value : example, mediaType.schema, {
                location,
                ctx,
                allowAdditionalProperties: !!opts.allowAdditionalProperties,
                ajvContext: context,
            });
        }
    };
    return {
        Parameter: {
            MediaType: {
                skip,
                leave: leave({ apiContext: 'request' }),
            },
        },
        RequestBody: {
            MediaType: {
                skip,
                leave: leave({ apiContext: 'request' }),
            },
        },
        Response: {
            MediaType: {
                skip,
                leave: leave({ apiContext: 'response' }),
            },
        },
    };
};
