// @redocly/openapi-core/lib/rules/oas3/request-mime-type.js
import { validateMimeTypeOAS3 } from '../../utils/validate-mime-type.js';
export const RequestMimeType = ({ allowedValues }) => {
    return {
        Paths: {
            RequestBody: {
                leave(requestBody, ctx) {
                    validateMimeTypeOAS3({
                        type: 'consumes',
                        value: requestBody,
                        ctx,
                        allowedValues,
                        reference: 'https://redocly.com/docs/cli/rules/oas/request-mime-type',
                    });
                },
            },
            Callback: {
                RequestBody() { },
                Response: {
                    leave(response, ctx) {
                        validateMimeTypeOAS3({
                            type: 'consumes',
                            value: response,
                            ctx,
                            allowedValues,
                            reference: 'https://redocly.com/docs/cli/rules/oas/request-mime-type',
                        });
                    },
                },
            },
        },
        WebhooksMap: {
            Response: {
                leave(response, ctx) {
                    validateMimeTypeOAS3({
                        type: 'consumes',
                        value: response,
                        ctx,
                        allowedValues,
                        reference: 'https://redocly.com/docs/cli/rules/oas/request-mime-type',
                    });
                },
            },
        },
    };
};
