// @redocly/openapi-core/lib/rules/oas3/response-mime-type.js
import { validateMimeTypeOAS3 } from '../../utils/validate-mime-type.js';
export const ResponseMimeType = ({ allowedValues }) => {
    return {
        Paths: {
            Response: {
                leave(response, ctx) {
                    validateMimeTypeOAS3({
                        type: 'produces',
                        value: response,
                        ctx,
                        allowedValues,
                        reference: 'https://redocly.com/docs/cli/rules/oas/response-mime-type',
                    });
                },
            },
            Callback: {
                Response() { },
                RequestBody: {
                    leave(requestBody, ctx) {
                        validateMimeTypeOAS3({
                            type: 'produces',
                            value: requestBody,
                            ctx,
                            allowedValues,
                            reference: 'https://redocly.com/docs/cli/rules/oas/response-mime-type',
                        });
                    },
                },
            },
        },
        WebhooksMap: {
            RequestBody: {
                leave(requestBody, ctx) {
                    validateMimeTypeOAS3({
                        type: 'produces',
                        value: requestBody,
                        ctx,
                        allowedValues,
                        reference: 'https://redocly.com/docs/cli/rules/oas/response-mime-type',
                    });
                },
            },
        },
    };
};
