// @redocly/openapi-core/lib/rules/oas3/spec-components-invalid-map-name.js
import { COMPONENT_NAME_CHARS } from '../../oas-types.js';
export const SpecComponentsInvalidMapName = () => {
    const componentNamePattern = `^[${COMPONENT_NAME_CHARS}]+$`;
    function validateKey(key, report, location, component) {
        if (!new RegExp(componentNamePattern).test(key)) {
            report({
                message: `The map key in ${component} "${key}" does not match the regular expression "${componentNamePattern}"`,
                location: location.key(),
                reference: 'https://redocly.com/docs/cli/rules/oas/spec-components-invalid-map-name',
            });
        }
    }
    return {
        NamedSchemas: {
            Schema(_node, { key, report, location }) {
                validateKey(key, report, location, 'schemas');
            },
        },
        NamedParameters: {
            Parameter(_node, { key, report, location }) {
                validateKey(key, report, location, 'parameters');
            },
        },
        NamedResponses: {
            Response(_node, { key, report, location }) {
                validateKey(key, report, location, 'responses');
            },
        },
        NamedExamples: {
            Example(_node, { key, report, location }) {
                validateKey(key, report, location, 'examples');
            },
        },
        NamedRequestBodies: {
            RequestBody(_node, { key, report, location }) {
                validateKey(key, report, location, 'requestBodies');
            },
        },
        NamedHeaders: {
            Header(_node, { key, report, location }) {
                validateKey(key, report, location, 'headers');
            },
        },
        NamedSecuritySchemes: {
            SecurityScheme(_node, { key, report, location }) {
                validateKey(key, report, location, 'securitySchemes');
            },
        },
        NamedLinks: {
            Link(_node, { key, report, location }) {
                validateKey(key, report, location, 'links');
            },
        },
        NamedCallbacks: {
            Callback(_node, { key, report, location }) {
                validateKey(key, report, location, 'callbacks');
            },
        },
    };
};
