// @kubb/oas/dist/index.js
import { findSchemaDefinition, matchesMimeType, matchesMimeType as matchesMimeType$1 } from "oas/utils";
import { isRef } from "oas/types";
import OASNormalize from "oas-normalize";
import { isPlainObject, mergeDeep } from "remeda";
import swagger2openapi from "swagger2openapi";
import jsonpointer from "jsonpointer";
import BaseOas from "oas";

//#region src/types.ts
const HttpMethods = {
	GET: "get",
	POST: "post",
	PUT: "put",
	PATCH: "patch",
	DELETE: "delete",
	HEAD: "head",
	OPTIONS: "options",
	TRACE: "trace"
};

//#endregion
//#region src/Oas.ts
var Oas = class extends BaseOas {
	#options = { discriminator: "strict" };
	document = void 0;
	constructor({ oas, user }) {
		if (typeof oas === "string") oas = JSON.parse(oas);
		super(oas, user);
		this.document = oas;
	}
	setOptions(options) {
		this.#options = options;
	}
	get options() {
		return this.#options;
	}
	get($ref) {
		const origRef = $ref;
		$ref = $ref.trim();
		if ($ref === "") return false;
		if ($ref.startsWith("#")) $ref = globalThis.decodeURIComponent($ref.substring(1));
		else return null;
		const current = jsonpointer.get(this.api, $ref);
		if (!current) throw new Error(`Could not find a definition for ${origRef}.`);
		return current;
	}
	getKey($ref) {
		const key = $ref.split("/").pop();
		return key === "" ? void 0 : key;
	}
	set($ref, value) {
		$ref = $ref.trim();
		if ($ref === "") return false;
		if ($ref.startsWith("#")) {
			$ref = globalThis.decodeURIComponent($ref.substring(1));
			jsonpointer.set(this.api, $ref, value);
		}
	}
	getDiscriminator(schema) {
		if (!isDiscriminator(schema)) return;
		const { mapping = {}, propertyName } = schema.discriminator;
		if (this.#options.discriminator === "inherit") Object.entries(mapping).forEach(([mappingKey, mappingValue]) => {
			if (mappingValue) {
				const childSchema = this.get(mappingValue);
				if (!childSchema.properties) childSchema.properties = {};
				const property = childSchema.properties[propertyName];
				if (childSchema.properties) {
					childSchema.properties[propertyName] = {
						...childSchema.properties ? childSchema.properties[propertyName] : {},
						enum: [...property?.enum?.filter((value) => value !== mappingKey) ?? [], mappingKey]
					};
					childSchema.required = [...childSchema.required ?? [], propertyName];
					this.set(mappingValue, childSchema);
				}
			}
		});
		if (schema.oneOf) schema.oneOf.forEach((schema$1) => {
			if (isReference(schema$1)) {
				const key = this.getKey(schema$1.$ref);
				const propertySchema = this.get(schema$1.$ref).properties?.[propertyName];
				const canAdd = key && !Object.values(mapping).includes(schema$1.$ref);
				if (canAdd && propertySchema?.enum?.length === 1) mapping[propertySchema.enum[0]] = schema$1.$ref;
				else if (canAdd) mapping[key] = schema$1.$ref;
			}
		});
		if (schema.anyOf) schema.anyOf.forEach((schema$1) => {
			if (isReference(schema$1)) {
				const key = this.getKey(schema$1.$ref);
				const propertySchema = this.get(schema$1.$ref).properties?.[propertyName];
				const canAdd = key && !Object.values(mapping).includes(schema$1.$ref);
				if (canAdd && propertySchema?.enum?.length === 1) mapping[propertySchema.enum[0]] = schema$1.$ref;
				else if (canAdd) mapping[key] = schema$1.$ref;
			}
		});
		return {
			...schema.discriminator,
			mapping
		};
	}
	dereferenceWithRef(schema) {
		if (isReference(schema)) return {
			...schema,
			...this.get(schema.$ref),
			$ref: schema.$ref
		};
		return schema;
	}
	/**
	* Oas does not have a getResponseBody(contentType)
	*/
	#getResponseBodyFactory(responseBody) {
		function hasResponseBody(res = responseBody) {
			return !!res;
		}
		return (contentType) => {
			if (!hasResponseBody(responseBody)) return false;
			if (isReference(responseBody)) return false;
			if (!responseBody.content) return false;
			if (contentType) {
				if (!(contentType in responseBody.content)) return false;
				return responseBody.content[contentType];
			}
			let availablecontentType;
			const contentTypes = Object.keys(responseBody.content);
			contentTypes.forEach((mt) => {
				if (!availablecontentType && matchesMimeType$1.json(mt)) availablecontentType = mt;
			});
			if (!availablecontentType) contentTypes.forEach((mt) => {
				if (!availablecontentType) availablecontentType = mt;
			});
			if (availablecontentType) return [
				availablecontentType,
				responseBody.content[availablecontentType],
				...responseBody.description ? [responseBody.description] : []
			];
			return false;
		};
	}
	getResponseSchema(operation, statusCode) {
		if (operation.schema.responses) Object.keys(operation.schema.responses).forEach((key) => {
			const schema$1 = operation.schema.responses[key];
			const $ref = isReference(schema$1) ? schema$1.$ref : void 0;
			if (schema$1 && $ref) operation.schema.responses[key] = this.get($ref);
		});
		const getResponseBody = this.#getResponseBodyFactory(operation.getResponseByStatusCode(statusCode));
		const { contentType } = this.#options;
		const responseBody = getResponseBody(contentType);
		if (responseBody === false) return {};
		const schema = Array.isArray(responseBody) ? responseBody[1].schema : responseBody.schema;
		if (!schema) return {};
		return this.dereferenceWithRef(schema);
	}
	getRequestSchema(operation) {
		const { contentType } = this.#options;
		if (operation.schema.requestBody) operation.schema.requestBody = this.dereferenceWithRef(operation.schema.requestBody);
		const requestBody = operation.getRequestBody(contentType);
		if (requestBody === false) return;
		const schema = Array.isArray(requestBody) ? requestBody[1].schema : requestBody.schema;
		if (!schema) return;
		return this.dereferenceWithRef(schema);
	}
	getParametersSchema(operation, inKey) {
		const { contentType = operation.getContentType() } = this.#options;
		const params = operation.getParameters().map((schema) => {
			return this.dereferenceWithRef(schema);
		}).filter((v) => v.in === inKey);
		if (!params.length) return null;
		return params.reduce((schema, pathParameters) => {
			const property = pathParameters.content?.[contentType]?.schema ?? pathParameters.schema;
			const required = [...schema.required || [], pathParameters.required ? pathParameters.name : void 0].filter(Boolean);
			return {
				...schema,
				description: schema.description,
				deprecated: schema.deprecated,
				example: schema.example,
				required,
				properties: {
					...schema.properties,
					[pathParameters.name]: {
						description: pathParameters.description,
						...property
					}
				}
			};
		}, {
			type: "object",
			required: [],
			properties: {}
		});
	}
	async valdiate() {
		return new OASNormalize(this.api, {
			enablePaths: true,
			colorizeErrors: true
		}).validate({ parser: { validate: { errors: { colorize: true } } } });
	}
};

//#endregion
//#region src/utils.ts
function isOpenApiV2Document(doc) {
	return doc && isPlainObject(doc) && !("openapi" in doc);
}
function isOpenApiV3_1Document(doc) {
	return doc && isPlainObject(doc) && "openapi" in doc && doc.openapi.startsWith("3.1");
}
function isParameterObject(obj) {
	return obj && "in" in obj;
}
/**
* Determines if a schema is nullable, considering both the standard `nullable` property and the legacy `x-nullable` extension.
*
* @param schema - The schema object to check.
* @returns `true` if the schema is marked as nullable; otherwise, `false`.
*/
function isNullable(schema) {
	return schema?.nullable ?? schema?.["x-nullable"] ?? false;
}
/**
* Determines if the given object is an OpenAPI ReferenceObject.
*
* @returns True if {@link obj} is a ReferenceObject; otherwise, false.
*/
function isReference(obj) {
	return !!obj && isRef(obj);
}
/**
* Determines if the given object is a SchemaObject with a discriminator property of type DiscriminatorObject.
*
* @returns True if {@link obj} is a SchemaObject containing a non-string {@link discriminator} property.
*/
function isDiscriminator(obj) {
	return !!obj && obj?.["discriminator"] && typeof obj.discriminator !== "string";
}
/**
* Determines whether a schema is required.
*
* Returns true if the schema has a non-empty {@link SchemaObject.required} array or a truthy {@link SchemaObject.required} property.
*
* @param schema - The schema object to check.
* @returns True if the schema is required; otherwise, false.
*/
function isRequired(schema) {
	if (!schema) return false;
	return Array.isArray(schema.required) ? !!schema.required?.length : !!schema.required;
}
function isOptional(schema) {
	return !isRequired(schema);
}
async function parse(pathOrApi, { oasClass = Oas, canBundle = true, enablePaths = true } = {}) {
	const { loadConfig, bundle } = await import("@redocly/openapi-core");
	if (typeof pathOrApi === "string" && canBundle) return parse((await bundle({
		ref: pathOrApi,
		config: await loadConfig(),
		base: pathOrApi
	})).bundle.parsed);
	const document = await new OASNormalize(pathOrApi, {
		enablePaths,
		colorizeErrors: true
	}).load();
	if (isOpenApiV2Document(document)) {
		const { openapi } = await swagger2openapi.convertObj(document, { anchors: true });
		return new oasClass({ oas: openapi });
	}
	return new oasClass({ oas: document });
}
async function merge(pathOrApi, { oasClass = Oas } = {}) {
	const instances = await Promise.all(pathOrApi.map((p) => parse(p, {
		oasClass,
		enablePaths: false,
		canBundle: false
	})));
	if (instances.length === 0) throw new Error("No OAS instances provided for merging.");
	return parse(instances.reduce((acc, current) => {
		return mergeDeep(acc, current.document);
	}, {
		openapi: "3.0.0",
		info: {
			title: "Merged API",
			version: "1.0.0"
		},
		paths: {},
		components: { schemas: {} }
	}), { oasClass });
}

//#endregion
export { HttpMethods, Oas, findSchemaDefinition, isDiscriminator, isNullable, isOpenApiV3_1Document, isOptional, isParameterObject, isReference, isRequired, matchesMimeType, merge, parse };
