// @kubb/cli/dist/agent-DFz0qvI9.cjs
require("./chunk-ByKO4r7w.cjs");
const require_define = require("./define-D6Kfm7-Z.cjs");
const require_package = require("./package-g4_RFuAW.cjs");
const require_constants = require("./constants-BTUap0zs.cjs");
//#region src/commands/agent/start.ts
const command$1 = require_define.defineCommand({
	name: "start",
	description: "Start the Agent server",
	options: {
		config: {
			type: "string",
			description: "Path to the Kubb config",
			short: "c"
		},
		port: {
			type: "string",
			description: `Port for the server (default: ${require_constants.agentDefaults.port})`,
			short: "p"
		},
		host: {
			type: "string",
			description: "Host for the server",
			default: require_constants.agentDefaults.host
		},
		"allow-write": {
			type: "boolean",
			description: "Allow writing generated files to the filesystem. When not set, no files are written and the config patch is not persisted.",
			default: false
		},
		"allow-all": {
			type: "boolean",
			description: "Grant all permissions (implies --allow-write).",
			default: false
		}
	},
	async run({ values }) {
		const { runAgentStart } = await Promise.resolve().then(() => require("./agent-C6o_6GSJ.cjs"));
		await runAgentStart({
			port: values.port !== void 0 ? values.port : void 0,
			host: values.host,
			configPath: values.config,
			allowWrite: values["allow-write"],
			allowAll: values["allow-all"],
			version: require_package.version
		});
	}
});
//#endregion
//#region src/commands/agent.ts
const command = require_define.defineCommand({
	name: "agent",
	description: "Manage the Kubb Agent server",
	subCommands: [command$1]
});
//#endregion
exports.command = command;
