// @kubb/cli/dist/agent-C6o_6GSJ.cjs
const require_chunk = require("./chunk-ByKO4r7w.cjs");
const require_telemetry = require("./telemetry-Cn9X1I5B.cjs");
const require_shell = require("./shell-7HPrTCJ5.cjs");
const require_constants = require("./constants-BTUap0zs.cjs");
let node_util = require("node:util");
let node_path = require("node:path");
node_path = require_chunk.__toESM(node_path);
let node_process = require("node:process");
node_process = require_chunk.__toESM(node_process);
let node_net = require("node:net");
node_net = require_chunk.__toESM(node_net);
let node_url = require("node:url");
let _clack_prompts = require("@clack/prompts");
_clack_prompts = require_chunk.__toESM(_clack_prompts);
//#region src/runners/agent.ts
function isPortAvailable(port, host) {
	return new Promise((resolve) => {
		const server = node_net.default.createServer();
		server.once("error", () => resolve(false));
		server.once("listening", () => {
			server.close();
			resolve(true);
		});
		server.listen(port, host);
	});
}
async function runAgentStart({ port, host, configPath, allowWrite, allowAll, version }) {
	const hrStart = node_process.hrtime();
	try {
		try {
			node_process.loadEnvFile();
		} catch {}
		const agentPkgPath = (0, node_url.fileURLToPath)({}.resolve("@kubb/agent/package.json"));
		const agentDir = node_path.default.dirname(agentPkgPath);
		const serverPath = node_path.default.join(agentDir, require_constants.agentDefaults.serverEntryPath);
		const PORT = port !== void 0 ? port : node_process.env.PORT ?? require_constants.agentDefaults.port;
		const HOST = host !== require_constants.agentDefaults.host ? host : node_process.env.HOST ?? require_constants.agentDefaults.host;
		const KUBB_AGENT_ROOT = node_process.env.KUBB_AGENT_ROOT ?? node_process.cwd();
		const KUBB_AGENT_CONFIG = node_path.default.resolve(node_process.cwd(), configPath || node_process.env.KUBB_AGENT_CONFIG || require_constants.agentDefaults.configFile);
		const KUBB_AGENT_ALLOW_WRITE = allowAll || allowWrite ? "true" : node_process.env.KUBB_AGENT_ALLOW_WRITE ?? "false";
		const KUBB_AGENT_ALLOW_ALL = allowAll ? "true" : node_process.env.KUBB_AGENT_ALLOW_ALL ?? "false";
		const KUBB_AGENT_TOKEN = node_process.env.KUBB_AGENT_TOKEN;
		const KUBB_AGENT_RETRY_TIMEOUT = node_process.env.KUBB_AGENT_RETRY_TIMEOUT ?? require_constants.agentDefaults.retryTimeout;
		const KUBB_STUDIO_URL = node_process.env.KUBB_STUDIO_URL ?? require_constants.agentDefaults.studioUrl;
		const env = {
			...node_process.env,
			PORT,
			HOST,
			KUBB_AGENT_ROOT,
			KUBB_AGENT_CONFIG,
			KUBB_AGENT_ALLOW_WRITE,
			KUBB_AGENT_ALLOW_ALL,
			KUBB_AGENT_TOKEN,
			KUBB_AGENT_RETRY_TIMEOUT,
			KUBB_STUDIO_URL
		};
		_clack_prompts.log.step((0, node_util.styleText)("cyan", "Starting agent server..."));
		_clack_prompts.log.info((0, node_util.styleText)("dim", `Config: ${KUBB_AGENT_CONFIG}`));
		_clack_prompts.log.info((0, node_util.styleText)("dim", `Host: ${HOST}`));
		_clack_prompts.log.info((0, node_util.styleText)("dim", `Port: ${PORT}`));
		if (!KUBB_AGENT_ALLOW_WRITE && !KUBB_AGENT_ALLOW_ALL) _clack_prompts.log.warn((0, node_util.styleText)("yellow", "Filesystem writes disabled. Use --allow-write or --allow-all to enable."));
		if (!await isPortAvailable(Number(PORT), HOST)) {
			_clack_prompts.log.error((0, node_util.styleText)("red", `Port ${PORT} is already in use. Stop the existing process or choose a different port with --port.`));
			node_process.exit(1);
		}
		console.log(env);
		await require_shell.spawnAsync("node", [serverPath], {
			env,
			cwd: node_process.cwd()
		});
		await require_telemetry.sendTelemetry(require_telemetry.buildTelemetryEvent({
			command: "agent",
			kubbVersion: version,
			hrStart,
			status: "success"
		}));
	} catch (error) {
		await require_telemetry.sendTelemetry(require_telemetry.buildTelemetryEvent({
			command: "agent",
			kubbVersion: version,
			hrStart,
			status: "failed"
		}));
		_clack_prompts.log.error((0, node_util.styleText)("red", "Failed to start agent server"));
		console.error(error);
		node_process.exit(1);
	}
}
//#endregion
exports.runAgentStart = runAgentStart;
