// @kubb/cli/dist/agent-BKut5yJS.js
import "./chunk--u3MIqq1.js";
import { n as defineCommand } from "./define--M_JMcDC.js";
import { t as version } from "./package-DaHFLAeG.js";
import { a as agentDefaults } from "./constants-CM3dJzjK.js";
//#endregion
//#region src/commands/agent.ts
const command = defineCommand({
	name: "agent",
	description: "Manage the Kubb Agent server",
	subCommands: [defineCommand({
		name: "start",
		description: "Start the Agent server",
		options: {
			config: {
				type: "string",
				description: "Path to the Kubb config",
				short: "c"
			},
			port: {
				type: "string",
				description: `Port for the server (default: ${agentDefaults.port})`,
				short: "p"
			},
			host: {
				type: "string",
				description: "Host for the server",
				default: agentDefaults.host
			},
			"allow-write": {
				type: "boolean",
				description: "Allow writing generated files to the filesystem. When not set, no files are written and the config patch is not persisted.",
				default: false
			},
			"allow-all": {
				type: "boolean",
				description: "Grant all permissions (implies --allow-write).",
				default: false
			}
		},
		async run({ values }) {
			const { runAgentStart } = await import("./agent-L50VNhXv.js");
			await runAgentStart({
				port: values.port !== void 0 ? values.port : void 0,
				host: values.host,
				configPath: values.config,
				allowWrite: values["allow-write"],
				allowAll: values["allow-all"],
				version
			});
		}
	})]
});
//#endregion
export { command };
