// @kubb/cli/dist/agent-L50VNhXv.js
import "./chunk--u3MIqq1.js";
import { r as sendTelemetry, t as buildTelemetryEvent } from "./telemetry-DxiR7clS.js";
import { t as spawnAsync } from "./shell-DqqWsHCD.js";
import { a as agentDefaults } from "./constants-CM3dJzjK.js";
import { styleText } from "node:util";
import path from "node:path";
import * as process$1 from "node:process";
import net from "node:net";
import { fileURLToPath } from "node:url";
import * as clack from "@clack/prompts";
//#region src/runners/agent.ts
function isPortAvailable(port, host) {
	return new Promise((resolve) => {
		const server = net.createServer();
		server.once("error", () => resolve(false));
		server.once("listening", () => {
			server.close();
			resolve(true);
		});
		server.listen(port, host);
	});
}
async function runAgentStart({ port, host, configPath, allowWrite, allowAll, version }) {
	const hrStart = process$1.hrtime();
	try {
		try {
			process$1.loadEnvFile();
		} catch {}
		const agentPkgPath = fileURLToPath(import.meta.resolve("@kubb/agent/package.json"));
		const agentDir = path.dirname(agentPkgPath);
		const serverPath = path.join(agentDir, agentDefaults.serverEntryPath);
		const PORT = port !== void 0 ? port : process$1.env.PORT ?? agentDefaults.port;
		const HOST = host !== agentDefaults.host ? host : process$1.env.HOST ?? agentDefaults.host;
		const KUBB_AGENT_ROOT = process$1.env.KUBB_AGENT_ROOT ?? process$1.cwd();
		const KUBB_AGENT_CONFIG = path.resolve(process$1.cwd(), configPath || process$1.env.KUBB_AGENT_CONFIG || agentDefaults.configFile);
		const KUBB_AGENT_ALLOW_WRITE = allowAll || allowWrite ? "true" : process$1.env.KUBB_AGENT_ALLOW_WRITE ?? "false";
		const KUBB_AGENT_ALLOW_ALL = allowAll ? "true" : process$1.env.KUBB_AGENT_ALLOW_ALL ?? "false";
		const KUBB_AGENT_TOKEN = process$1.env.KUBB_AGENT_TOKEN;
		const KUBB_AGENT_RETRY_TIMEOUT = process$1.env.KUBB_AGENT_RETRY_TIMEOUT ?? agentDefaults.retryTimeout;
		const KUBB_STUDIO_URL = process$1.env.KUBB_STUDIO_URL ?? agentDefaults.studioUrl;
		const env = {
			...process$1.env,
			PORT,
			HOST,
			KUBB_AGENT_ROOT,
			KUBB_AGENT_CONFIG,
			KUBB_AGENT_ALLOW_WRITE,
			KUBB_AGENT_ALLOW_ALL,
			KUBB_AGENT_TOKEN,
			KUBB_AGENT_RETRY_TIMEOUT,
			KUBB_STUDIO_URL
		};
		clack.log.step(styleText("cyan", "Starting agent server..."));
		clack.log.info(styleText("dim", `Config: ${KUBB_AGENT_CONFIG}`));
		clack.log.info(styleText("dim", `Host: ${HOST}`));
		clack.log.info(styleText("dim", `Port: ${PORT}`));
		if (!KUBB_AGENT_ALLOW_WRITE && !KUBB_AGENT_ALLOW_ALL) clack.log.warn(styleText("yellow", "Filesystem writes disabled. Use --allow-write or --allow-all to enable."));
		if (!await isPortAvailable(Number(PORT), HOST)) {
			clack.log.error(styleText("red", `Port ${PORT} is already in use. Stop the existing process or choose a different port with --port.`));
			process$1.exit(1);
		}
		console.log(env);
		await spawnAsync("node", [serverPath], {
			env,
			cwd: process$1.cwd()
		});
		await sendTelemetry(buildTelemetryEvent({
			command: "agent",
			kubbVersion: version,
			hrStart,
			status: "success"
		}));
	} catch (error) {
		await sendTelemetry(buildTelemetryEvent({
			command: "agent",
			kubbVersion: version,
			hrStart,
			status: "failed"
		}));
		clack.log.error(styleText("red", "Failed to start agent server"));
		console.error(error);
		process$1.exit(1);
	}
}
//#endregion
export { runAgentStart };
