// @kubb/cli/dist/constants-BTUap0zs.cjs
//#region src/constants.ts
/** NPM registry endpoint used to check for @kubb/cli updates. */
const KUBB_NPM_PACKAGE_URL = "https://registry.npmjs.org/@kubb/cli/latest";
/** OpenTelemetry ingestion endpoint for anonymous usage telemetry. */
const OTLP_ENDPOINT = "https://otlp.kubb.dev";
/** Horizontal rule rendered above/below the plain-logger generation summary. */
const SUMMARY_SEPARATOR = "─".repeat(27);
/** Glob pattern for paths the file watcher ignores. */
const WATCHER_IGNORED_PATHS = "**/{.git,node_modules}/**";
/** Default runtime values for the `agent start` command. */
const agentDefaults = {
	port: "3000",
	host: "localhost",
	configFile: "kubb.config.ts",
	retryTimeout: "30000",
	studioUrl: "https://studio.kubb.dev",
	serverEntryPath: ".output/server/index.mjs"
};
/** Default values used during interactive `init` scaffolding. */
const initDefaults = {
	inputPath: "./openapi.yaml",
	outputPath: "./src/gen",
	plugins: ["plugin-oas", "plugin-ts"]
};
/**
* Maps each plugin value to the default config snippet inserted by `init`.
* The `satisfies` constraint ensures all values remain plain strings while
* `as const` keeps the object deeply immutable.
*/
const pluginDefaultConfigs = {
	"plugin-oas": "pluginOas()",
	"plugin-ts": `pluginTs({
      output: { path: 'models' },
    })`,
	"plugin-client": `pluginClient({
      output: { path: 'clients' },
    })`,
	"plugin-react-query": `pluginReactQuery({
      output: { path: 'hooks' },
    })`,
	"plugin-solid-query": `pluginSolidQuery({
      output: { path: 'hooks' },
    })`,
	"plugin-svelte-query": `pluginSvelteQuery({
      output: { path: 'hooks' },
    })`,
	"plugin-vue-query": `pluginVueQuery({
      output: { path: 'hooks' },
    })`,
	"plugin-swr": `pluginSwr({
      output: { path: 'hooks' },
    })`,
	"plugin-zod": `pluginZod({
      output: { path: 'zod' },
    })`,
	"plugin-faker": `pluginFaker({
      output: { path: 'mocks' },
    })`,
	"plugin-msw": `pluginMsw({
      output: { path: 'msw' },
    })`
};
/** Color palette used by randomCliColor() for deterministic plugin name coloring. */
//#endregion
Object.defineProperty(exports, "KUBB_NPM_PACKAGE_URL", {
	enumerable: true,
	get: function() {
		return KUBB_NPM_PACKAGE_URL;
	}
});
Object.defineProperty(exports, "OTLP_ENDPOINT", {
	enumerable: true,
	get: function() {
		return OTLP_ENDPOINT;
	}
});
Object.defineProperty(exports, "SUMMARY_SEPARATOR", {
	enumerable: true,
	get: function() {
		return SUMMARY_SEPARATOR;
	}
});
Object.defineProperty(exports, "WATCHER_IGNORED_PATHS", {
	enumerable: true,
	get: function() {
		return WATCHER_IGNORED_PATHS;
	}
});
Object.defineProperty(exports, "agentDefaults", {
	enumerable: true,
	get: function() {
		return agentDefaults;
	}
});
Object.defineProperty(exports, "initDefaults", {
	enumerable: true,
	get: function() {
		return initDefaults;
	}
});
Object.defineProperty(exports, "pluginDefaultConfigs", {
	enumerable: true,
	get: function() {
		return pluginDefaultConfigs;
	}
});
