// @kubb/cli/dist/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
require("./chunk-ByKO4r7w.cjs");
const require_define = require("./define-D6Kfm7-Z.cjs");
const require_telemetry = require("./telemetry-Cn9X1I5B.cjs");
const require_package = require("./package-g4_RFuAW.cjs");
let node_util = require("node:util");
//#region ../../internals/utils/src/cli/schema.ts
/**
* Serializes `CommandDefinition[]` to a plain, JSON-serializable structure.
* Use to expose CLI capabilities to AI agents or MCP tools.
*/
function getCommandSchema(defs) {
	return defs.map(serializeCommand);
}
function serializeCommand(def) {
	return {
		name: def.name,
		description: def.description,
		arguments: def.arguments,
		options: serializeOptions(def.options ?? {}),
		subCommands: def.subCommands ? def.subCommands.map(serializeCommand) : []
	};
}
function serializeOptions(options) {
	return Object.entries(options).map(([name, opt]) => {
		return {
			name,
			flags: `${opt.short ? `-${opt.short}, ` : ""}--${name}${opt.type === "string" ? ` <${opt.hint ?? name}>` : ""}`,
			type: opt.type,
			description: opt.description,
			...opt.default !== void 0 ? { default: opt.default } : {},
			...opt.hint ? { hint: opt.hint } : {},
			...opt.enum ? { enum: opt.enum } : {},
			...opt.required ? { required: opt.required } : {}
		};
	});
}
//#endregion
//#region ../../internals/utils/src/cli/help.ts
/** Prints formatted help output for a command using its `CommandDefinition`. */
function renderHelp(def, parentName) {
	const schema = getCommandSchema([def])[0];
	const programName = parentName ? `${parentName} ${schema.name}` : schema.name;
	const argsPart = schema.arguments?.length ? ` ${schema.arguments.join(" ")}` : "";
	const subCmdPart = schema.subCommands.length ? " <command>" : "";
	console.log(`\n${(0, node_util.styleText)("bold", "Usage:")} ${programName}${argsPart}${subCmdPart} [options]\n`);
	if (schema.description) console.log(`  ${schema.description}\n`);
	if (schema.subCommands.length) {
		console.log((0, node_util.styleText)("bold", "Commands:"));
		for (const sub of schema.subCommands) console.log(`  ${(0, node_util.styleText)("cyan", sub.name.padEnd(16))}${sub.description}`);
		console.log();
	}
	const options = [...schema.options, {
		name: "help",
		flags: "-h, --help",
		type: "boolean",
		description: "Show help"
	}];
	console.log((0, node_util.styleText)("bold", "Options:"));
	for (const opt of options) {
		const flags = (0, node_util.styleText)("cyan", opt.flags.padEnd(30));
		const defaultPart = opt.default !== void 0 ? (0, node_util.styleText)("dim", ` (default: ${opt.default})`) : "";
		console.log(`  ${flags}${opt.description}${defaultPart}`);
	}
	console.log();
}
//#endregion
//#region ../../internals/utils/src/cli/adapters/nodeAdapter.ts
function buildParseOptions(def) {
	const result = { help: {
		type: "boolean",
		short: "h"
	} };
	for (const [name, opt] of Object.entries(def.options ?? {})) result[name] = {
		type: opt.type,
		...opt.short ? { short: opt.short } : {},
		...opt.default !== void 0 ? { default: opt.default } : {}
	};
	return result;
}
async function runCommand(def, argv, parentName) {
	const parseOptions = buildParseOptions(def);
	let parsed;
	try {
		const result = (0, node_util.parseArgs)({
			args: argv,
			options: parseOptions,
			allowPositionals: true,
			strict: false
		});
		parsed = {
			values: result.values,
			positionals: result.positionals
		};
	} catch {
		renderHelp(def, parentName);
		process.exit(1);
	}
	if (parsed.values["help"]) {
		renderHelp(def, parentName);
		process.exit(0);
	}
	for (const [name, opt] of Object.entries(def.options ?? {})) if (opt.required && parsed.values[name] === void 0) {
		console.error((0, node_util.styleText)("red", `Error: --${name} is required`));
		renderHelp(def, parentName);
		process.exit(1);
	}
	if (!def.run) {
		renderHelp(def, parentName);
		process.exit(0);
	}
	try {
		await def.run(parsed);
	} catch (err) {
		console.error((0, node_util.styleText)("red", `Error: ${err instanceof Error ? err.message : String(err)}`));
		renderHelp(def, parentName);
		process.exit(1);
	}
}
function printRootHelp(programName, version, defs) {
	console.log(`\n${(0, node_util.styleText)("bold", "Usage:")} ${programName} <command> [options]\n`);
	console.log(`  Kubb generation — v${version}\n`);
	console.log((0, node_util.styleText)("bold", "Commands:"));
	for (const def of defs) console.log(`  ${(0, node_util.styleText)("cyan", def.name.padEnd(16))}${def.description}`);
	console.log();
	console.log((0, node_util.styleText)("bold", "Options:"));
	console.log(`  ${(0, node_util.styleText)("cyan", "-v, --version".padEnd(30))}Show version number`);
	console.log(`  ${(0, node_util.styleText)("cyan", "-h, --help".padEnd(30))}Show help`);
	console.log();
	console.log(`Run ${(0, node_util.styleText)("cyan", `${programName} <command> --help`)} for command-specific help.\n`);
}
/** CLI adapter using `node:util parseArgs`. No external dependencies. */
const nodeAdapter = require_define.defineCLIAdapter({
	renderHelp(def, parentName) {
		renderHelp(def, parentName);
	},
	async run(defs, argv, opts) {
		const { programName, defaultCommandName, version } = opts;
		const args = argv.length >= 2 && argv[0]?.includes("node") ? argv.slice(2) : argv;
		if (args[0] === "--version" || args[0] === "-v") {
			console.log(version);
			process.exit(0);
		}
		if (args[0] === "--help" || args[0] === "-h") {
			printRootHelp(programName, version, defs);
			process.exit(0);
		}
		if (args.length === 0) {
			const defaultDef = defs.find((d) => d.name === defaultCommandName);
			if (defaultDef?.run) await runCommand(defaultDef, [], programName);
			else printRootHelp(programName, version, defs);
			return;
		}
		const [first, ...rest] = args;
		const isKnownSubcommand = defs.some((d) => d.name === first);
		let def;
		let commandArgv;
		let parentName;
		if (isKnownSubcommand) {
			def = defs.find((d) => d.name === first);
			commandArgv = rest;
			parentName = programName;
		} else {
			def = defs.find((d) => d.name === defaultCommandName);
			commandArgv = args;
			parentName = programName;
		}
		if (!def) {
			console.error(`Unknown command: ${first}`);
			printRootHelp(programName, version, defs);
			process.exit(1);
		}
		if (def.subCommands?.length) {
			const [subName, ...subRest] = commandArgv;
			const subDef = def.subCommands.find((s) => s.name === subName);
			if (subName === "--help" || subName === "-h") {
				renderHelp(def, parentName);
				process.exit(0);
			}
			if (!subDef) {
				renderHelp(def, parentName);
				process.exit(subName ? 1 : 0);
			}
			await runCommand(subDef, subRest, `${parentName} ${def.name}`);
			return;
		}
		await runCommand(def, commandArgv, parentName);
	}
});
//#endregion
//#region ../../internals/utils/src/cli/parse.ts
/**
* Create a CLI runner bound to a specific adapter.
* Defaults to the built-in `nodeAdapter` (Node.js `node:util parseArgs`).
*/
function createCLI(options) {
	const adapter = options?.adapter ?? nodeAdapter;
	return { run(commands, argv, opts) {
		return adapter.run(commands, argv, opts);
	} };
}
//#endregion
//#region src/index.ts
const cli = createCLI();
function shouldShowTelemetryNotice(argv) {
	if (require_telemetry.isTelemetryDisabled()) return false;
	const quietFlags = new Set([
		"--help",
		"-h",
		"--version",
		"-v"
	]);
	if (argv.some((arg) => quietFlags.has(arg))) return false;
	if (!process.stdout.isTTY) return false;
	return true;
}
async function run(argv = process.argv) {
	if (shouldShowTelemetryNotice(argv)) console.log(`${(0, node_util.styleText)("yellow", "Notice:")} Kubb collects anonymous telemetry data to help improve the tool. No personal data or file contents are collected. \nTo disable, set ${(0, node_util.styleText)("cyan", "KUBB_DISABLE_TELEMETRY=1")}.\n`);
	const [{ command: generateCommand }, { command: validateCommand }, { command: mcpCommand }, { command: agentCommand }, { command: initCommand }] = await Promise.all([
		Promise.resolve().then(() => require("./generate-Dcy4wBJb.cjs")),
		Promise.resolve().then(() => require("./validate-DDk_Kw27.cjs")),
		Promise.resolve().then(() => require("./mcp-BbxbxGxa.cjs")),
		Promise.resolve().then(() => require("./agent-DFz0qvI9.cjs")),
		Promise.resolve().then(() => require("./init-DG67Dx80.cjs"))
	]);
	await cli.run([
		generateCommand,
		validateCommand,
		mcpCommand,
		agentCommand,
		initCommand
	], argv, {
		programName: "kubb",
		defaultCommandName: "generate",
		version: require_package.version
	});
}
//#endregion
exports.run = run;
