// @kubb/cli/dist/index.js
import { defineCommand, runCommand, runMain } from "citty";
import consola from "consola";
import gradientString from "gradient-string";
import getLatestVersion from "latest-version";
import { lt } from "semver";

//#region package.json
var version = "4.3.1";

//#endregion
//#region src/index.ts
const main = defineCommand({
	meta: {
		name: "kubb",
		version,
		description: "Kubb generation"
	},
	args: { version: {
		type: "boolean",
		alias: "v",
		description: "Show version number"
	} },
	async setup({ rawArgs, args }) {
		if (args.version) {
			console.log(version);
			process.exit(0);
		}
		try {
			consola.log(gradientString([
				"#F58517",
				"#F5A217",
				"#F55A17"
			])("Kubb CLI:"));
			const latestVersion = await getLatestVersion("@kubb/cli");
			if (lt(version, latestVersion)) consola.box({
				title: "Update available for `Kubb` ",
				message: `\`v${version}\` → \`v${latestVersion}\`
Run \`npm install -g @kubb/cli\` to update`,
				style: {
					padding: 2,
					borderColor: "yellow",
					borderStyle: "rounded"
				}
			});
		} catch (_e) {}
		if (![
			"generate",
			"validate",
			"mcp"
		].includes(rawArgs[0])) {
			console.log(rawArgs[0]);
			await runCommand(await import("./generate-MZHwytlj.js").then((r) => r.default), { rawArgs });
			process.exit(0);
		}
	},
	subCommands: {
		generate: () => import("./generate-MZHwytlj.js").then((r) => r.default),
		validate: () => import("./validate-8i6chavn.js").then((r) => r.default),
		mcp: () => import("./mcp-C4KjGtQv.js").then((r) => r.default)
	}
});
async function run(_argv) {
	await runMain(main);
}

//#endregion
export { run };
