// @kubb/cli/dist/init-C-InrmSY.js
import "./chunk--u3MIqq1.js";
import { t as spawnAsync } from "./shell-DqqWsHCD.js";
import { o as initDefaults, s as pluginDefaultConfigs } from "./constants-CM3dJzjK.js";
import { styleText } from "node:util";
import fs, { existsSync, readFileSync } from "node:fs";
import path, { join } from "node:path";
import process$1 from "node:process";
import * as clack from "@clack/prompts";
//#region ../../internals/utils/src/packageManager.ts
const packageManagers = {
	pnpm: {
		name: "pnpm",
		lockFile: "pnpm-lock.yaml",
		installCommand: ["add", "-D"]
	},
	yarn: {
		name: "yarn",
		lockFile: "yarn.lock",
		installCommand: ["add", "-D"]
	},
	bun: {
		name: "bun",
		lockFile: "bun.lockb",
		installCommand: ["add", "-d"]
	},
	npm: {
		name: "npm",
		lockFile: "package-lock.json",
		installCommand: ["install", "--save-dev"]
	}
};
/**
* Detects the active package manager for the given directory.
* Resolution order: `packageManager` field in `package.json`, then presence of a lock file.
* Falls back to npm when no signal is found.
*/
function detectPackageManager(cwd = process.cwd()) {
	const packageJsonPath = join(cwd, "package.json");
	if (existsSync(packageJsonPath)) try {
		const pmField = JSON.parse(readFileSync(packageJsonPath, "utf-8")).packageManager;
		if (typeof pmField === "string") {
			const name = pmField.split("@")[0];
			if (name && name in packageManagers) return packageManagers[name];
		}
	} catch {}
	for (const pm of Object.values(packageManagers)) if (existsSync(join(cwd, pm.lockFile))) return pm;
	return packageManagers.npm;
}
//#endregion
//#region src/utils/packageManager.ts
function hasPackageJson(cwd = process.cwd()) {
	return fs.existsSync(path.join(cwd, "package.json"));
}
async function initPackageJson(cwd, packageManager) {
	await spawnAsync(packageManager.name, {
		npm: ["init", "-y"],
		pnpm: ["init"],
		yarn: ["init", "-y"],
		bun: ["init", "-y"]
	}[packageManager.name], { cwd });
}
async function installPackages(packages, packageManager, cwd = process.cwd()) {
	await spawnAsync(packageManager.name, [...packageManager.installCommand, ...packages], { cwd });
}
//#endregion
//#region src/runners/init.ts
const availablePlugins = [
	{
		value: "plugin-oas",
		label: "OpenAPI Parser",
		hint: "Required",
		packageName: "@kubb/plugin-oas",
		importName: "pluginOas",
		category: "core"
	},
	{
		value: "plugin-ts",
		label: "TypeScript",
		hint: "Recommended",
		packageName: "@kubb/plugin-ts",
		importName: "pluginTs",
		category: "typescript"
	},
	{
		value: "plugin-client",
		label: "Client (Fetch/Axios)",
		packageName: "@kubb/plugin-client",
		importName: "pluginClient",
		category: "typescript"
	},
	{
		value: "plugin-react-query",
		label: "React Query / TanStack Query",
		packageName: "@kubb/plugin-react-query",
		importName: "pluginReactQuery",
		category: "query"
	},
	{
		value: "plugin-solid-query",
		label: "Solid Query",
		packageName: "@kubb/plugin-solid-query",
		importName: "pluginSolidQuery",
		category: "query"
	},
	{
		value: "plugin-svelte-query",
		label: "Svelte Query",
		packageName: "@kubb/plugin-svelte-query",
		importName: "pluginSvelteQuery",
		category: "query"
	},
	{
		value: "plugin-vue-query",
		label: "Vue Query",
		packageName: "@kubb/plugin-vue-query",
		importName: "pluginVueQuery",
		category: "query"
	},
	{
		value: "plugin-swr",
		label: "SWR",
		packageName: "@kubb/plugin-swr",
		importName: "pluginSwr",
		category: "query"
	},
	{
		value: "plugin-zod",
		label: "Zod Schemas",
		packageName: "@kubb/plugin-zod",
		importName: "pluginZod",
		category: "validation"
	},
	{
		value: "plugin-faker",
		label: "Faker.js Mocks",
		packageName: "@kubb/plugin-faker",
		importName: "pluginFaker",
		category: "mocking"
	},
	{
		value: "plugin-msw",
		label: "MSW Handlers",
		packageName: "@kubb/plugin-msw",
		importName: "pluginMsw",
		category: "mocking"
	},
	{
		value: "plugin-cypress",
		label: "Cypress Tests",
		packageName: "@kubb/plugin-cypress",
		importName: "pluginCypress",
		category: "testing"
	},
	{
		value: "plugin-redoc",
		label: "ReDoc Documentation",
		packageName: "@kubb/plugin-redoc",
		importName: "pluginRedoc",
		category: "docs"
	}
];
function generateConfigFile(selectedPlugins, inputPath, outputPath) {
	return `import { defineConfig } from '@kubb/core'
${selectedPlugins.map((plugin) => `import { ${plugin.importName} } from '${plugin.packageName}'`).join("\n")}

export default defineConfig({
  root: '.',
  input: {
    path: '${inputPath}',
  },
  output: {
    path: '${outputPath}',
    clean: true,
  },
  plugins: [
${selectedPlugins.map((plugin) => {
		return `    ${pluginDefaultConfigs[plugin.value] ?? `${plugin.importName}()`},`;
	}).join("\n")}
  ],
})
`;
}
function cancelAndExit(message = "Operation cancelled.") {
	clack.cancel(message);
	process$1.exit(0);
}
async function runInit({ yes, version }) {
	const cwd = process$1.cwd();
	clack.intro(styleText("bgCyan", styleText("black", " Kubb Init ")));
	try {
		let packageManager;
		if (!hasPackageJson(cwd)) {
			if (!yes) {
				const shouldInit = await clack.confirm({
					message: "No package.json found. Would you like to create one?",
					initialValue: true
				});
				if (clack.isCancel(shouldInit) || !shouldInit) cancelAndExit();
			}
			packageManager = detectPackageManager(cwd);
			const spinner = clack.spinner();
			spinner.start(`Initializing package.json with ${packageManager.name}`);
			await initPackageJson(cwd, packageManager);
			spinner.stop(`Created package.json with ${packageManager.name}`);
		} else {
			packageManager = detectPackageManager(cwd);
			clack.log.info(`Detected package manager: ${styleText("cyan", packageManager.name)}`);
		}
		let inputPath;
		if (yes) {
			inputPath = initDefaults.inputPath;
			clack.log.info(`Using input path: ${styleText("cyan", inputPath)}`);
		} else {
			const inputPathResult = await clack.text({
				message: "Where is your OpenAPI specification located?",
				placeholder: initDefaults.inputPath,
				defaultValue: initDefaults.inputPath,
				validate: (value) => {
					if (!value) return "Input path is required";
				}
			});
			if (clack.isCancel(inputPathResult)) cancelAndExit();
			inputPath = inputPathResult;
		}
		let outputPath;
		if (yes) {
			outputPath = initDefaults.outputPath;
			clack.log.info(`Using output path: ${styleText("cyan", outputPath)}`);
		} else {
			const outputPathResult = await clack.text({
				message: "Where should the generated files be output?",
				placeholder: initDefaults.outputPath,
				defaultValue: initDefaults.outputPath,
				validate: (value) => {
					if (!value) return "Output path is required";
				}
			});
			if (clack.isCancel(outputPathResult)) cancelAndExit();
			outputPath = outputPathResult;
		}
		let selectedPlugins;
		if (yes) {
			selectedPlugins = availablePlugins.filter((plugin) => initDefaults.plugins.includes(plugin.value));
			clack.log.info(`Using plugins: ${styleText("cyan", selectedPlugins.map((p) => p.label).join(", "))}`);
		} else {
			const selectedPluginValues = await clack.multiselect({
				message: "Select plugins to use:",
				options: availablePlugins.map((plugin) => ({
					value: plugin.value,
					label: plugin.label,
					hint: plugin.hint
				})),
				initialValues: [...initDefaults.plugins],
				required: true
			});
			if (clack.isCancel(selectedPluginValues)) cancelAndExit();
			selectedPlugins = availablePlugins.filter((plugin) => selectedPluginValues.includes(plugin.value));
		}
		if (!selectedPlugins.find((p) => p.value === "plugin-oas")) selectedPlugins.unshift(availablePlugins.find((p) => p.value === "plugin-oas"));
		const packagesToInstall = [
			"@kubb/core",
			"@kubb/cli",
			"@kubb/agent",
			...selectedPlugins.map((p) => p.packageName)
		];
		const spinner = clack.spinner();
		spinner.start(`Installing ${packagesToInstall.length} packages with ${packageManager.name}`);
		try {
			await installPackages(packagesToInstall, packageManager, cwd);
			spinner.stop(`Installed ${packagesToInstall.length} packages`);
		} catch (error) {
			spinner.stop("Installation failed");
			throw error;
		}
		const configSpinner = clack.spinner();
		configSpinner.start("Creating kubb.config.ts");
		const configContent = generateConfigFile(selectedPlugins, inputPath, outputPath);
		const configPath = path.join(cwd, "kubb.config.ts");
		if (fs.existsSync(configPath)) {
			configSpinner.stop("kubb.config.ts already exists");
			if (!yes) {
				const shouldOverwrite = await clack.confirm({
					message: "kubb.config.ts already exists. Overwrite?",
					initialValue: false
				});
				if (clack.isCancel(shouldOverwrite) || !shouldOverwrite) cancelAndExit("Keeping existing configuration. Packages have been installed.");
			}
			configSpinner.start("Overwriting kubb.config.ts");
		}
		fs.writeFileSync(configPath, configContent, "utf-8");
		configSpinner.stop("Created kubb.config.ts");
		clack.outro(styleText("green", "✓ All set!") + "\n\n" + styleText("dim", "Next steps:") + "\n" + styleText("cyan", `  1. Make sure your OpenAPI spec is at: ${inputPath}`) + "\n" + styleText("cyan", "  2. Generate code with: npx kubb generate") + "\n" + styleText("cyan", "     Or start a stream server with: npx kubb start") + "\n" + styleText("cyan", `  3. Find generated files in: ${outputPath}`) + "\n\n" + styleText("dim", `Using ${packageManager.name} • Kubb v${version}`));
	} catch (error) {
		clack.log.error(styleText("red", "An error occurred during initialization"));
		if (error instanceof Error) clack.log.error(error.message);
		process$1.exit(1);
	}
}
//#endregion
export { runInit };
