// @kubb/cli/dist/init-hmolV6B4.cjs
const require_chunk = require("./chunk-ByKO4r7w.cjs");
const require_shell = require("./shell-7HPrTCJ5.cjs");
const require_constants = require("./constants-BTUap0zs.cjs");
let node_util = require("node:util");
let node_fs = require("node:fs");
node_fs = require_chunk.__toESM(node_fs);
let node_path = require("node:path");
node_path = require_chunk.__toESM(node_path);
let node_process = require("node:process");
node_process = require_chunk.__toESM(node_process);
let _clack_prompts = require("@clack/prompts");
_clack_prompts = require_chunk.__toESM(_clack_prompts);
//#region ../../internals/utils/src/packageManager.ts
const packageManagers = {
	pnpm: {
		name: "pnpm",
		lockFile: "pnpm-lock.yaml",
		installCommand: ["add", "-D"]
	},
	yarn: {
		name: "yarn",
		lockFile: "yarn.lock",
		installCommand: ["add", "-D"]
	},
	bun: {
		name: "bun",
		lockFile: "bun.lockb",
		installCommand: ["add", "-d"]
	},
	npm: {
		name: "npm",
		lockFile: "package-lock.json",
		installCommand: ["install", "--save-dev"]
	}
};
/**
* Detects the active package manager for the given directory.
* Resolution order: `packageManager` field in `package.json`, then presence of a lock file.
* Falls back to npm when no signal is found.
*/
function detectPackageManager(cwd = process.cwd()) {
	const packageJsonPath = (0, node_path.join)(cwd, "package.json");
	if ((0, node_fs.existsSync)(packageJsonPath)) try {
		const pmField = JSON.parse((0, node_fs.readFileSync)(packageJsonPath, "utf-8")).packageManager;
		if (typeof pmField === "string") {
			const name = pmField.split("@")[0];
			if (name && name in packageManagers) return packageManagers[name];
		}
	} catch {}
	for (const pm of Object.values(packageManagers)) if ((0, node_fs.existsSync)((0, node_path.join)(cwd, pm.lockFile))) return pm;
	return packageManagers.npm;
}
//#endregion
//#region src/utils/packageManager.ts
function hasPackageJson(cwd = process.cwd()) {
	return node_fs.default.existsSync(node_path.default.join(cwd, "package.json"));
}
async function initPackageJson(cwd, packageManager) {
	await require_shell.spawnAsync(packageManager.name, {
		npm: ["init", "-y"],
		pnpm: ["init"],
		yarn: ["init", "-y"],
		bun: ["init", "-y"]
	}[packageManager.name], { cwd });
}
async function installPackages(packages, packageManager, cwd = process.cwd()) {
	await require_shell.spawnAsync(packageManager.name, [...packageManager.installCommand, ...packages], { cwd });
}
//#endregion
//#region src/runners/init.ts
const availablePlugins = [
	{
		value: "plugin-oas",
		label: "OpenAPI Parser",
		hint: "Required",
		packageName: "@kubb/plugin-oas",
		importName: "pluginOas",
		category: "core"
	},
	{
		value: "plugin-ts",
		label: "TypeScript",
		hint: "Recommended",
		packageName: "@kubb/plugin-ts",
		importName: "pluginTs",
		category: "typescript"
	},
	{
		value: "plugin-client",
		label: "Client (Fetch/Axios)",
		packageName: "@kubb/plugin-client",
		importName: "pluginClient",
		category: "typescript"
	},
	{
		value: "plugin-react-query",
		label: "React Query / TanStack Query",
		packageName: "@kubb/plugin-react-query",
		importName: "pluginReactQuery",
		category: "query"
	},
	{
		value: "plugin-solid-query",
		label: "Solid Query",
		packageName: "@kubb/plugin-solid-query",
		importName: "pluginSolidQuery",
		category: "query"
	},
	{
		value: "plugin-svelte-query",
		label: "Svelte Query",
		packageName: "@kubb/plugin-svelte-query",
		importName: "pluginSvelteQuery",
		category: "query"
	},
	{
		value: "plugin-vue-query",
		label: "Vue Query",
		packageName: "@kubb/plugin-vue-query",
		importName: "pluginVueQuery",
		category: "query"
	},
	{
		value: "plugin-swr",
		label: "SWR",
		packageName: "@kubb/plugin-swr",
		importName: "pluginSwr",
		category: "query"
	},
	{
		value: "plugin-zod",
		label: "Zod Schemas",
		packageName: "@kubb/plugin-zod",
		importName: "pluginZod",
		category: "validation"
	},
	{
		value: "plugin-faker",
		label: "Faker.js Mocks",
		packageName: "@kubb/plugin-faker",
		importName: "pluginFaker",
		category: "mocking"
	},
	{
		value: "plugin-msw",
		label: "MSW Handlers",
		packageName: "@kubb/plugin-msw",
		importName: "pluginMsw",
		category: "mocking"
	},
	{
		value: "plugin-cypress",
		label: "Cypress Tests",
		packageName: "@kubb/plugin-cypress",
		importName: "pluginCypress",
		category: "testing"
	},
	{
		value: "plugin-redoc",
		label: "ReDoc Documentation",
		packageName: "@kubb/plugin-redoc",
		importName: "pluginRedoc",
		category: "docs"
	}
];
function generateConfigFile(selectedPlugins, inputPath, outputPath) {
	return `import { defineConfig } from '@kubb/core'
${selectedPlugins.map((plugin) => `import { ${plugin.importName} } from '${plugin.packageName}'`).join("\n")}

export default defineConfig({
  root: '.',
  input: {
    path: '${inputPath}',
  },
  output: {
    path: '${outputPath}',
    clean: true,
  },
  plugins: [
${selectedPlugins.map((plugin) => {
		return `    ${require_constants.pluginDefaultConfigs[plugin.value] ?? `${plugin.importName}()`},`;
	}).join("\n")}
  ],
})
`;
}
function cancelAndExit(message = "Operation cancelled.") {
	_clack_prompts.cancel(message);
	node_process.default.exit(0);
}
async function runInit({ yes, version }) {
	const cwd = node_process.default.cwd();
	_clack_prompts.intro((0, node_util.styleText)("bgCyan", (0, node_util.styleText)("black", " Kubb Init ")));
	try {
		let packageManager;
		if (!hasPackageJson(cwd)) {
			if (!yes) {
				const shouldInit = await _clack_prompts.confirm({
					message: "No package.json found. Would you like to create one?",
					initialValue: true
				});
				if (_clack_prompts.isCancel(shouldInit) || !shouldInit) cancelAndExit();
			}
			packageManager = detectPackageManager(cwd);
			const spinner = _clack_prompts.spinner();
			spinner.start(`Initializing package.json with ${packageManager.name}`);
			await initPackageJson(cwd, packageManager);
			spinner.stop(`Created package.json with ${packageManager.name}`);
		} else {
			packageManager = detectPackageManager(cwd);
			_clack_prompts.log.info(`Detected package manager: ${(0, node_util.styleText)("cyan", packageManager.name)}`);
		}
		let inputPath;
		if (yes) {
			inputPath = require_constants.initDefaults.inputPath;
			_clack_prompts.log.info(`Using input path: ${(0, node_util.styleText)("cyan", inputPath)}`);
		} else {
			const inputPathResult = await _clack_prompts.text({
				message: "Where is your OpenAPI specification located?",
				placeholder: require_constants.initDefaults.inputPath,
				defaultValue: require_constants.initDefaults.inputPath,
				validate: (value) => {
					if (!value) return "Input path is required";
				}
			});
			if (_clack_prompts.isCancel(inputPathResult)) cancelAndExit();
			inputPath = inputPathResult;
		}
		let outputPath;
		if (yes) {
			outputPath = require_constants.initDefaults.outputPath;
			_clack_prompts.log.info(`Using output path: ${(0, node_util.styleText)("cyan", outputPath)}`);
		} else {
			const outputPathResult = await _clack_prompts.text({
				message: "Where should the generated files be output?",
				placeholder: require_constants.initDefaults.outputPath,
				defaultValue: require_constants.initDefaults.outputPath,
				validate: (value) => {
					if (!value) return "Output path is required";
				}
			});
			if (_clack_prompts.isCancel(outputPathResult)) cancelAndExit();
			outputPath = outputPathResult;
		}
		let selectedPlugins;
		if (yes) {
			selectedPlugins = availablePlugins.filter((plugin) => require_constants.initDefaults.plugins.includes(plugin.value));
			_clack_prompts.log.info(`Using plugins: ${(0, node_util.styleText)("cyan", selectedPlugins.map((p) => p.label).join(", "))}`);
		} else {
			const selectedPluginValues = await _clack_prompts.multiselect({
				message: "Select plugins to use:",
				options: availablePlugins.map((plugin) => ({
					value: plugin.value,
					label: plugin.label,
					hint: plugin.hint
				})),
				initialValues: [...require_constants.initDefaults.plugins],
				required: true
			});
			if (_clack_prompts.isCancel(selectedPluginValues)) cancelAndExit();
			selectedPlugins = availablePlugins.filter((plugin) => selectedPluginValues.includes(plugin.value));
		}
		if (!selectedPlugins.find((p) => p.value === "plugin-oas")) selectedPlugins.unshift(availablePlugins.find((p) => p.value === "plugin-oas"));
		const packagesToInstall = [
			"@kubb/core",
			"@kubb/cli",
			"@kubb/agent",
			...selectedPlugins.map((p) => p.packageName)
		];
		const spinner = _clack_prompts.spinner();
		spinner.start(`Installing ${packagesToInstall.length} packages with ${packageManager.name}`);
		try {
			await installPackages(packagesToInstall, packageManager, cwd);
			spinner.stop(`Installed ${packagesToInstall.length} packages`);
		} catch (error) {
			spinner.stop("Installation failed");
			throw error;
		}
		const configSpinner = _clack_prompts.spinner();
		configSpinner.start("Creating kubb.config.ts");
		const configContent = generateConfigFile(selectedPlugins, inputPath, outputPath);
		const configPath = node_path.default.join(cwd, "kubb.config.ts");
		if (node_fs.default.existsSync(configPath)) {
			configSpinner.stop("kubb.config.ts already exists");
			if (!yes) {
				const shouldOverwrite = await _clack_prompts.confirm({
					message: "kubb.config.ts already exists. Overwrite?",
					initialValue: false
				});
				if (_clack_prompts.isCancel(shouldOverwrite) || !shouldOverwrite) cancelAndExit("Keeping existing configuration. Packages have been installed.");
			}
			configSpinner.start("Overwriting kubb.config.ts");
		}
		node_fs.default.writeFileSync(configPath, configContent, "utf-8");
		configSpinner.stop("Created kubb.config.ts");
		_clack_prompts.outro((0, node_util.styleText)("green", "✓ All set!") + "\n\n" + (0, node_util.styleText)("dim", "Next steps:") + "\n" + (0, node_util.styleText)("cyan", `  1. Make sure your OpenAPI spec is at: ${inputPath}`) + "\n" + (0, node_util.styleText)("cyan", "  2. Generate code with: npx kubb generate") + "\n" + (0, node_util.styleText)("cyan", "     Or start a stream server with: npx kubb start") + "\n" + (0, node_util.styleText)("cyan", `  3. Find generated files in: ${outputPath}`) + "\n\n" + (0, node_util.styleText)("dim", `Using ${packageManager.name} • Kubb v${version}`));
	} catch (error) {
		_clack_prompts.log.error((0, node_util.styleText)("red", "An error occurred during initialization"));
		if (error instanceof Error) _clack_prompts.log.error(error.message);
		node_process.default.exit(1);
	}
}
//#endregion
exports.runInit = runInit;
