// @kubb/cli/dist/telemetry-Cn9X1I5B.cjs
const require_chunk = require("./chunk-ByKO4r7w.cjs");
const require_constants = require("./constants-BTUap0zs.cjs");
let node_crypto = require("node:crypto");
let node_dns = require("node:dns");
let node_os = require("node:os");
node_os = require_chunk.__toESM(node_os);
let node_process = require("node:process");
node_process = require_chunk.__toESM(node_process);
//#region ../../internals/utils/src/env.ts
/**
* Returns `true` when running inside a GitHub Actions workflow.
*/
function isGitHubActions() {
	return !!process.env.GITHUB_ACTIONS;
}
/**
* Returns `true` when the process is running in a CI environment.
* Covers GitHub Actions, GitLab CI, CircleCI, Travis CI, Jenkins, Bitbucket,
* TeamCity, Buildkite, and Azure Pipelines.
*/
function isCIEnvironment() {
	return !!(process.env.CI || process.env.GITHUB_ACTIONS || process.env.GITLAB_CI || process.env.BITBUCKET_BUILD_NUMBER || process.env.JENKINS_URL || process.env.CIRCLECI || process.env.TRAVIS || process.env.TEAMCITY_VERSION || process.env.BUILDKITE || process.env.TF_BUILD);
}
/**
* Returns `true` when the process has an interactive TTY and is not running in CI.
*/
function canUseTTY() {
	return !!process.stdout.isTTY && !isCIEnvironment();
}
//#endregion
//#region ../../internals/utils/src/network.ts
/** Well-known stable domains used as DNS probes to check internet connectivity. */
const TEST_DOMAINS = [
	"dns.google.com",
	"cloudflare.com",
	"one.one.one.one"
];
/**
* Returns `true` when the system has internet connectivity.
* Uses DNS resolution against well-known stable domains as a lightweight probe.
*/
async function isOnline() {
	for (const domain of TEST_DOMAINS) try {
		await node_dns.promises.resolve(domain);
		return true;
	} catch {}
	return false;
}
/**
* Executes `fn` only when the system is online. Returns `null` when offline or on error.
*/
async function executeIfOnline(fn) {
	if (!await isOnline()) return null;
	try {
		return await fn();
	} catch {
		return null;
	}
}
//#endregion
//#region src/utils/telemetry.ts
/**
* Detect whether the current process is running inside a CI environment.
* Delegates to the canonical isCIEnvironment() from envDetection.
*/
function isCi() {
	return isCIEnvironment();
}
/**
* Check if telemetry is disabled via DO_NOT_TRACK or KUBB_DISABLE_TELEMETRY.
* Respects the standard DO_NOT_TRACK convention used across development tools.
*/
function isTelemetryDisabled() {
	return node_process.default.env["DO_NOT_TRACK"] === "1" || node_process.default.env["DO_NOT_TRACK"] === "true" || node_process.default.env["KUBB_DISABLE_TELEMETRY"] === "1" || node_process.default.env["KUBB_DISABLE_TELEMETRY"] === "true";
}
/**
* Convert a TelemetryEvent into an OTLP-compatible JSON trace payload.
* See https://opentelemetry.io/docs/languages/sdk-configuration/otlp-exporter/
*/
function buildOtlpPayload(event) {
	const traceId = (0, node_crypto.randomBytes)(16).toString("hex");
	const spanId = (0, node_crypto.randomBytes)(8).toString("hex");
	const endTimeNs = BigInt(Date.now()) * 1000000n;
	const startTimeNs = endTimeNs - BigInt(event.duration) * 1000000n;
	const attributes = [
		{
			key: "kubb.command",
			value: { stringValue: event.command }
		},
		{
			key: "kubb.version",
			value: { stringValue: event.kubbVersion }
		},
		{
			key: "kubb.node_version",
			value: { stringValue: event.nodeVersion }
		},
		{
			key: "kubb.platform",
			value: { stringValue: event.platform }
		},
		{
			key: "kubb.ci",
			value: { boolValue: event.ci }
		},
		{
			key: "kubb.files_created",
			value: { intValue: event.filesCreated }
		},
		{
			key: "kubb.status",
			value: { stringValue: event.status }
		},
		{
			key: "kubb.plugins",
			value: { arrayValue: { values: event.plugins.map((p) => ({ kvlistValue: { values: [{
				key: "name",
				value: { stringValue: p.name }
			}, {
				key: "options",
				value: { stringValue: JSON.stringify({
					...p.options,
					usedEnumNames: void 0
				}) }
			}] } })) } }
		}
	];
	return { resourceSpans: [{
		resource: { attributes: [
			{
				key: "service.name",
				value: { stringValue: "kubb-cli" }
			},
			{
				key: "service.version",
				value: { stringValue: event.kubbVersion }
			},
			{
				key: "telemetry.sdk.language",
				value: { stringValue: "nodejs" }
			}
		] },
		scopeSpans: [{
			scope: {
				name: "kubb-cli",
				version: event.kubbVersion
			},
			spans: [{
				traceId,
				spanId,
				name: event.command,
				kind: 1,
				startTimeUnixNano: String(startTimeNs),
				endTimeUnixNano: String(endTimeNs),
				attributes,
				status: { code: event.status === "success" ? 1 : 2 }
			}]
		}]
	}] };
}
/**
* Send an anonymous telemetry event to the Kubb OTLP endpoint.
* Respects DO_NOT_TRACK and KUBB_DISABLE_TELEMETRY environment variables.
* Fails silently to never interrupt the generation process.
*/
async function sendTelemetry(event) {
	if (isTelemetryDisabled()) return;
	await executeIfOnline(async () => {
		try {
			await fetch(`${require_constants.OTLP_ENDPOINT}/v1/traces`, {
				method: "POST",
				headers: {
					"Content-Type": "application/json",
					"Kubb-Telemetry-Version": "1",
					"Kubb-Telemetry-Source": "kubb-cli"
				},
				body: JSON.stringify(buildOtlpPayload(event)),
				signal: AbortSignal.timeout(5e3)
			});
		} catch (_e) {}
	});
}
/**
* Build an anonymous telemetry payload from a completed generation run.
* No file paths, OpenAPI specs, or secrets are included.
*/
function buildTelemetryEvent(options) {
	const [seconds, nanoseconds] = node_process.default.hrtime(options.hrStart);
	const duration = Math.round(seconds * 1e3 + nanoseconds / 1e6);
	return {
		command: options.command,
		kubbVersion: options.kubbVersion,
		nodeVersion: node_process.default.versions.node.split(".")[0],
		platform: node_os.default.platform(),
		ci: isCi(),
		plugins: options.plugins ?? [],
		duration,
		filesCreated: options.filesCreated ?? 0,
		status: options.status
	};
}
//#endregion
Object.defineProperty(exports, "buildTelemetryEvent", {
	enumerable: true,
	get: function() {
		return buildTelemetryEvent;
	}
});
Object.defineProperty(exports, "canUseTTY", {
	enumerable: true,
	get: function() {
		return canUseTTY;
	}
});
Object.defineProperty(exports, "executeIfOnline", {
	enumerable: true,
	get: function() {
		return executeIfOnline;
	}
});
Object.defineProperty(exports, "isGitHubActions", {
	enumerable: true,
	get: function() {
		return isGitHubActions;
	}
});
Object.defineProperty(exports, "isTelemetryDisabled", {
	enumerable: true,
	get: function() {
		return isTelemetryDisabled;
	}
});
Object.defineProperty(exports, "sendTelemetry", {
	enumerable: true,
	get: function() {
		return sendTelemetry;
	}
});
