// @kubb/plugin-zod/dist/index.js
import "./components-BUkyR7lK.js";
import { n as zodGenerator, t as operationsGenerator } from "./generators-8AD8PdD3.js";
import path from "node:path";
import { PackageManager, PluginManager, createPlugin, getBarrelFiles, getMode } from "@kubb/core";
import { camelCase, pascalCase } from "@kubb/core/transformers";
import { OperationGenerator, SchemaGenerator, pluginOasName } from "@kubb/plugin-oas";
import { pluginTsName } from "@kubb/plugin-ts";

//#region src/plugin.ts
const pluginZodName = "plugin-zod";
const pluginZod = createPlugin((options) => {
	const { output = {
		path: "zod",
		barrelType: "named"
	}, group, exclude = [], include, override = [], transformers: transformers$1 = {}, dateType = "string", unknownType = "any", emptySchemaType = unknownType, typed = false, mapper = {}, operations = false, version = new PackageManager().isValidSync("zod", ">=4") ? "4" : "3", importPath = version === "4" ? "zod/v4" : "zod", coercion = false, inferred = false, generators = [zodGenerator, operations ? operationsGenerator : void 0].filter(Boolean), wrapOutput = void 0, contentType } = options;
	return {
		name: pluginZodName,
		options: {
			output,
			transformers: transformers$1,
			include,
			exclude,
			override,
			typed,
			dateType,
			unknownType,
			emptySchemaType,
			mapper,
			importPath,
			coercion,
			operations,
			inferred,
			group,
			wrapOutput,
			version
		},
		pre: [pluginOasName, typed ? pluginTsName : void 0].filter(Boolean),
		resolvePath(baseName, pathMode, options$1) {
			const root = path.resolve(this.config.root, this.config.output.path);
			if ((pathMode ?? getMode(path.resolve(root, output.path))) === "single")
 /**
			* when output is a file then we will always append to the same file(output file), see fileManager.addOrAppend
			* Other plugins then need to call addOrAppend instead of just add from the fileManager class
			*/
			return path.resolve(root, output.path);
			if (group && (options$1?.group?.path || options$1?.group?.tag)) {
				const groupName = group?.name ? group.name : (ctx) => {
					if (group?.type === "path") return `${ctx.group.split("/")[1]}`;
					return `${camelCase(ctx.group)}Controller`;
				};
				return path.resolve(root, output.path, groupName({ group: group.type === "path" ? options$1.group.path : options$1.group.tag }), baseName);
			}
			return path.resolve(root, output.path, baseName);
		},
		resolveName(name, type) {
			let resolvedName = camelCase(name, {
				suffix: type ? "schema" : void 0,
				isFile: type === "file"
			});
			if (type === "type") resolvedName = pascalCase(resolvedName);
			if (type) return transformers$1?.name?.(resolvedName, type) || resolvedName;
			return resolvedName;
		},
		async buildStart() {
			const [swaggerPlugin] = PluginManager.getDependedPlugins(this.plugins, [pluginOasName]);
			const oas = await swaggerPlugin.context.getOas();
			const root = path.resolve(this.config.root, this.config.output.path);
			const mode = getMode(path.resolve(root, output.path));
			const schemaFiles = await new SchemaGenerator(this.plugin.options, {
				oas,
				pluginManager: this.pluginManager,
				plugin: this.plugin,
				contentType,
				include: void 0,
				override,
				mode,
				output: output.path
			}).build(...generators);
			await this.addFile(...schemaFiles);
			const operationFiles = await new OperationGenerator(this.plugin.options, {
				oas,
				pluginManager: this.pluginManager,
				plugin: this.plugin,
				contentType,
				exclude,
				include,
				override,
				mode
			}).build(...generators);
			await this.addFile(...operationFiles);
			const barrelFiles = await getBarrelFiles(this.fileManager.files, {
				type: output.barrelType ?? "named",
				root,
				output,
				meta: { pluginKey: this.plugin.key },
				logger: this.logger
			});
			await this.addFile(...barrelFiles);
		}
	};
});

//#endregion
export { pluginZod, pluginZodName };
