// @kubb/plugin-zod/dist/index.cjs
const require_components = require('./components-MHUlEVjU.cjs');
const require_generators = require('./generators-Bo3q-m_p.cjs');
let node_path = require("node:path");
node_path = require_components.__toESM(node_path);
let __kubb_core = require("@kubb/core");
__kubb_core = require_components.__toESM(__kubb_core);
let __kubb_core_transformers = require("@kubb/core/transformers");
__kubb_core_transformers = require_components.__toESM(__kubb_core_transformers);
let __kubb_plugin_oas = require("@kubb/plugin-oas");
__kubb_plugin_oas = require_components.__toESM(__kubb_plugin_oas);
let __kubb_plugin_ts = require("@kubb/plugin-ts");
__kubb_plugin_ts = require_components.__toESM(__kubb_plugin_ts);

//#region src/plugin.ts
const pluginZodName = "plugin-zod";
const pluginZod = (0, __kubb_core.createPlugin)((options) => {
	const { output = {
		path: "zod",
		barrelType: "named"
	}, group, exclude = [], include, override = [], transformers = {}, dateType = "string", unknownType = "any", emptySchemaType = unknownType, typed = false, mapper = {}, operations = false, version = new __kubb_core.PackageManager().isValidSync("zod", ">=4") ? "4" : "3", importPath = version === "4" ? "zod/v4" : "zod", coercion = false, inferred = false, generators = [require_generators.zodGenerator, operations ? require_generators.operationsGenerator : void 0].filter(Boolean), wrapOutput = void 0, contentType } = options;
	return {
		name: pluginZodName,
		options: {
			output,
			transformers,
			include,
			exclude,
			override,
			typed,
			dateType,
			unknownType,
			emptySchemaType,
			mapper,
			importPath,
			coercion,
			operations,
			inferred,
			group,
			wrapOutput,
			version
		},
		pre: [__kubb_plugin_oas.pluginOasName, typed ? __kubb_plugin_ts.pluginTsName : void 0].filter(Boolean),
		resolvePath(baseName, pathMode, options$1) {
			const root = node_path.default.resolve(this.config.root, this.config.output.path);
			if ((pathMode ?? (0, __kubb_core.getMode)(node_path.default.resolve(root, output.path))) === "single")
 /**
			* when output is a file then we will always append to the same file(output file), see fileManager.addOrAppend
			* Other plugins then need to call addOrAppend instead of just add from the fileManager class
			*/
			return node_path.default.resolve(root, output.path);
			if (group && (options$1?.group?.path || options$1?.group?.tag)) {
				const groupName = group?.name ? group.name : (ctx) => {
					if (group?.type === "path") return `${ctx.group.split("/")[1]}`;
					return `${(0, __kubb_core_transformers.camelCase)(ctx.group)}Controller`;
				};
				return node_path.default.resolve(root, output.path, groupName({ group: group.type === "path" ? options$1.group.path : options$1.group.tag }), baseName);
			}
			return node_path.default.resolve(root, output.path, baseName);
		},
		resolveName(name, type) {
			let resolvedName = (0, __kubb_core_transformers.camelCase)(name, {
				suffix: type ? "schema" : void 0,
				isFile: type === "file"
			});
			if (type === "type") resolvedName = (0, __kubb_core_transformers.pascalCase)(resolvedName);
			if (type) return transformers?.name?.(resolvedName, type) || resolvedName;
			return resolvedName;
		},
		async buildStart() {
			const [swaggerPlugin] = __kubb_core.PluginManager.getDependedPlugins(this.plugins, [__kubb_plugin_oas.pluginOasName]);
			const oas = await swaggerPlugin.context.getOas();
			const root = node_path.default.resolve(this.config.root, this.config.output.path);
			const mode = (0, __kubb_core.getMode)(node_path.default.resolve(root, output.path));
			const schemaFiles = await new __kubb_plugin_oas.SchemaGenerator(this.plugin.options, {
				oas,
				pluginManager: this.pluginManager,
				plugin: this.plugin,
				contentType,
				include: void 0,
				override,
				mode,
				output: output.path
			}).build(...generators);
			await this.addFile(...schemaFiles);
			const operationFiles = await new __kubb_plugin_oas.OperationGenerator(this.plugin.options, {
				oas,
				pluginManager: this.pluginManager,
				plugin: this.plugin,
				contentType,
				exclude,
				include,
				override,
				mode
			}).build(...generators);
			await this.addFile(...operationFiles);
			const barrelFiles = await (0, __kubb_core.getBarrelFiles)(this.fileManager.files, {
				type: output.barrelType ?? "named",
				root,
				output,
				meta: { pluginKey: this.plugin.key },
				logger: this.logger
			});
			await this.addFile(...barrelFiles);
		}
	};
});

//#endregion
exports.pluginZod = pluginZod;
exports.pluginZodName = pluginZodName;
