// @redocly/openapi-core/lib/rules/common/info-license-strict.js
import { detectSpec } from '../../detect-spec.js';
import { validateDefinedAndNonEmpty, validateOneOfDefinedAndNonEmpty } from '../utils.js';
export const InfoLicenseStrict = () => {
    let specVersion;
    return {
        Root: {
            enter(root) {
                specVersion = detectSpec(root);
            },
            License: {
                leave(license, ctx) {
                    if (specVersion === 'oas3_1' || specVersion === 'oas3_2') {
                        validateOneOfDefinedAndNonEmpty({
                            fieldNames: ['url', 'identifier'],
                            value: license,
                            ctx,
                            reference: 'https://redocly.com/docs/cli/rules/oas/info-license-strict',
                        });
                    }
                    else {
                        validateDefinedAndNonEmpty({
                            fieldName: 'url',
                            value: license,
                            ctx,
                            reference: 'https://redocly.com/docs/cli/rules/oas/info-license-strict',
                        });
                    }
                },
            },
        },
    };
};
