// @redocly/openapi-core/lib/rules/common/no-invalid-parameter-examples.js
import { isDefined } from '../../utils/is-defined.js';
import { isPlainObject } from '../../utils/is-plain-object.js';
import { AjvValidator } from '../ajv.js';
import { validateExample } from '../utils.js';
export const NoInvalidParameterExamples = (opts) => {
    const validator = new AjvValidator();
    return {
        Parameter: {
            leave(parameter, ctx) {
                if (isDefined(parameter.example)) {
                    validateExample({
                        example: parameter.example,
                        schema: parameter.schema,
                        options: {
                            location: ctx.location.child('example'),
                            ctx,
                            validator,
                            allowAdditionalProperties: !!opts.allowAdditionalProperties,
                            ajvContext: { apiContext: 'request' },
                        },
                        reference: 'https://redocly.com/docs/cli/rules/oas/no-invalid-parameter-examples',
                    });
                }
                if (isPlainObject(parameter.examples)) {
                    for (const [key, example] of Object.entries(parameter.examples)) {
                        if (isPlainObject(example) && 'value' in example) {
                            validateExample({
                                example: example.value,
                                schema: parameter.schema,
                                options: {
                                    location: ctx.location.child(['examples', key]),
                                    ctx,
                                    validator,
                                    allowAdditionalProperties: !!opts.allowAdditionalProperties,
                                    ajvContext: { apiContext: 'request' },
                                },
                                reference: 'https://redocly.com/docs/cli/rules/oas/no-invalid-parameter-examples',
                            });
                        }
                    }
                }
            },
        },
    };
};
