// @redocly/openapi-core/lib/rules/common/no-invalid-schema-examples.js
import { isDefined } from '../../utils/is-defined.js';
import { AjvValidator } from '../ajv.js';
import { validateExample } from '../utils.js';
export const NoInvalidSchemaExamples = (opts) => {
    const validator = new AjvValidator();
    return {
        Schema: {
            leave(schema, ctx) {
                const examples = schema.examples;
                if (Array.isArray(examples)) {
                    for (const example of examples) {
                        validateExample({
                            example,
                            schema,
                            options: {
                                location: ctx.location.child(['examples', examples.indexOf(example)]),
                                ctx,
                                validator,
                                allowAdditionalProperties: !!opts.allowAdditionalProperties,
                            },
                            reference: 'https://redocly.com/docs/cli/rules/oas/no-invalid-schema-examples',
                        });
                    }
                }
                if (isDefined(schema.example)) {
                    // Handle nullable example for OAS3
                    if (schema.nullable === true &&
                        schema.example === null &&
                        schema.type !== undefined) {
                        return;
                    }
                    validateExample({
                        example: schema.example,
                        schema,
                        options: {
                            location: ctx.location.child('example'),
                            ctx,
                            validator,
                            allowAdditionalProperties: !!opts.allowAdditionalProperties,
                        },
                        reference: 'https://redocly.com/docs/cli/rules/oas/no-invalid-schema-examples',
                    });
                }
            },
        },
    };
};
