// @redocly/openapi-core/lib/rules/common/operation-parameters-unique.js
export const OperationParametersUnique = () => {
    let seenPathParams;
    let seenOperationParams;
    return {
        PathItem: {
            enter() {
                seenPathParams = new Set();
            },
            Parameter(parameter, { report, key, parentLocations }) {
                const paramId = `${parameter.in}___${parameter.name}`;
                if (seenPathParams.has(paramId)) {
                    report({
                        message: `Paths must have unique \`name\` + \`in\` parameters.\nRepeats of \`in:${parameter.in}\` + \`name:${parameter.name}\`.`,
                        location: parentLocations.PathItem.child(['parameters', key]),
                        reference: 'https://redocly.com/docs/cli/rules/oas/operation-parameters-unique',
                    });
                }
                seenPathParams.add(`${parameter.in}___${parameter.name}`);
            },
            Operation: {
                enter() {
                    seenOperationParams = new Set();
                },
                Parameter(parameter, { report, key, parentLocations }) {
                    const paramId = `${parameter.in}___${parameter.name}`;
                    if (seenOperationParams.has(paramId)) {
                        report({
                            message: `Operations must have unique \`name\` + \`in\` parameters. Repeats of \`in:${parameter.in}\` + \`name:${parameter.name}\`.`,
                            location: parentLocations.Operation.child(['parameters', key]),
                            reference: 'https://redocly.com/docs/cli/rules/oas/operation-parameters-unique',
                        });
                    }
                    seenOperationParams.add(paramId);
                },
            },
        },
    };
};
