// @redocly/openapi-core/lib/rules/common/operation-2xx-response.js
import { validateResponseCodes } from '../utils.js';
export const Operation2xxResponse = ({ validateWebhooks }) => {
    return {
        Paths: {
            Responses(responses, { report }) {
                const codes = Object.keys(responses || {});
                validateResponseCodes({
                    responseCodes: codes,
                    codeRange: '2XX',
                    report: report,
                    reference: 'https://redocly.com/docs/cli/rules/oas/operation-2xx-response',
                });
            },
        },
        WebhooksMap: {
            Responses(responses, { report }) {
                if (!validateWebhooks)
                    return;
                const codes = Object.keys(responses || {});
                validateResponseCodes({
                    responseCodes: codes,
                    codeRange: '2XX',
                    report: report,
                    reference: 'https://redocly.com/docs/cli/rules/oas/operation-2xx-response',
                });
            },
        },
    };
};
