// @kubb/plugin-client/dist/index.js
import "./Operations-ktPlsW1u.js";
import { n as operationsGenerator, r as clientGenerator, t as groupedClientGenerator } from "./generators-BqXV2IXC.js";
import path from "node:path";
import { PluginManager, createPlugin, getBarrelFiles, getMode } from "@kubb/core";
import { camelCase } from "@kubb/core/transformers";
import { OperationGenerator, pluginOasName } from "@kubb/plugin-oas";
import { pluginZodName } from "@kubb/plugin-zod";

//#region src/plugin.ts
const pluginClientName = "plugin-client";
const pluginClient = createPlugin((options) => {
	const { output = {
		path: "clients",
		barrelType: "named"
	}, group, urlType = false, exclude = [], include, override = [], transformers = {}, dataReturnType = "data", paramsType = "inline", pathParamsType = paramsType === "object" ? "object" : options.pathParamsType || "inline", operations = false, baseURL, paramsCasing, generators = [
		clientGenerator,
		group ? groupedClientGenerator : void 0,
		operations ? operationsGenerator : void 0
	].filter(Boolean), parser = "client", client = "axios", importPath = client === "fetch" ? "@kubb/plugin-client/clients/fetch" : "@kubb/plugin-client/clients/axios", contentType } = options;
	return {
		name: pluginClientName,
		options: {
			output,
			group,
			parser,
			dataReturnType,
			importPath,
			paramsType,
			paramsCasing,
			pathParamsType,
			baseURL,
			urlType
		},
		pre: [pluginOasName, parser === "zod" ? pluginZodName : void 0].filter(Boolean),
		resolvePath(baseName, pathMode, options$1) {
			const root = path.resolve(this.config.root, this.config.output.path);
			if ((pathMode ?? getMode(path.resolve(root, output.path))) === "single")
 /**
			* when output is a file then we will always append to the same file(output file), see fileManager.addOrAppend
			* Other plugins then need to call addOrAppend instead of just add from the fileManager class
			*/
			return path.resolve(root, output.path);
			if (group && (options$1?.group?.path || options$1?.group?.tag)) {
				const groupName = group?.name ? group.name : (ctx) => {
					if (group?.type === "path") return `${ctx.group.split("/")[1]}`;
					return `${camelCase(ctx.group)}Controller`;
				};
				return path.resolve(root, output.path, groupName({ group: group.type === "path" ? options$1.group.path : options$1.group.tag }), baseName);
			}
			return path.resolve(root, output.path, baseName);
		},
		resolveName(name, type) {
			const resolvedName = camelCase(name, { isFile: type === "file" });
			if (type) return transformers?.name?.(resolvedName, type) || resolvedName;
			return resolvedName;
		},
		async buildStart() {
			const [swaggerPlugin] = PluginManager.getDependedPlugins(this.plugins, [pluginOasName]);
			const oas = await swaggerPlugin.context.getOas();
			const root = path.resolve(this.config.root, this.config.output.path);
			const mode = getMode(path.resolve(root, output.path));
			const baseURL$1 = await swaggerPlugin.context.getBaseURL();
			const files = await new OperationGenerator(baseURL$1 ? {
				...this.plugin.options,
				baseURL: baseURL$1
			} : this.plugin.options, {
				oas,
				pluginManager: this.pluginManager,
				plugin: this.plugin,
				contentType,
				exclude,
				include,
				override,
				mode
			}).build(...generators);
			await this.addFile(...files);
			const barrelFiles = await getBarrelFiles(this.fileManager.files, {
				type: output.barrelType ?? "named",
				root,
				output,
				meta: { pluginKey: this.plugin.key },
				logger: this.logger
			});
			await this.addFile(...barrelFiles);
		}
	};
});

//#endregion
export { pluginClient, pluginClientName };
