// @kubb/plugin-client/dist/index.cjs
const require_chunk = require('./chunk-CUT6urMc.cjs');
require('./Operations-CT9m1j9w.cjs');
const require_generators = require('./generators-BSxTToF6.cjs');
let node_path = require("node:path");
node_path = require_chunk.__toESM(node_path);
let __kubb_core = require("@kubb/core");
__kubb_core = require_chunk.__toESM(__kubb_core);
let __kubb_core_transformers = require("@kubb/core/transformers");
__kubb_core_transformers = require_chunk.__toESM(__kubb_core_transformers);
let __kubb_plugin_oas = require("@kubb/plugin-oas");
__kubb_plugin_oas = require_chunk.__toESM(__kubb_plugin_oas);
let __kubb_plugin_zod = require("@kubb/plugin-zod");
__kubb_plugin_zod = require_chunk.__toESM(__kubb_plugin_zod);

//#region src/plugin.ts
const pluginClientName = "plugin-client";
const pluginClient = (0, __kubb_core.createPlugin)((options) => {
	const { output = {
		path: "clients",
		barrelType: "named"
	}, group, urlType = false, exclude = [], include, override = [], transformers = {}, dataReturnType = "data", paramsType = "inline", pathParamsType = paramsType === "object" ? "object" : options.pathParamsType || "inline", operations = false, baseURL, paramsCasing, generators = [
		require_generators.clientGenerator,
		group ? require_generators.groupedClientGenerator : void 0,
		operations ? require_generators.operationsGenerator : void 0
	].filter(Boolean), parser = "client", client = "axios", importPath = client === "fetch" ? "@kubb/plugin-client/clients/fetch" : "@kubb/plugin-client/clients/axios", contentType } = options;
	return {
		name: pluginClientName,
		options: {
			output,
			group,
			parser,
			dataReturnType,
			importPath,
			paramsType,
			paramsCasing,
			pathParamsType,
			baseURL,
			urlType
		},
		pre: [__kubb_plugin_oas.pluginOasName, parser === "zod" ? __kubb_plugin_zod.pluginZodName : void 0].filter(Boolean),
		resolvePath(baseName, pathMode, options$1) {
			const root = node_path.default.resolve(this.config.root, this.config.output.path);
			if ((pathMode ?? (0, __kubb_core.getMode)(node_path.default.resolve(root, output.path))) === "single")
 /**
			* when output is a file then we will always append to the same file(output file), see fileManager.addOrAppend
			* Other plugins then need to call addOrAppend instead of just add from the fileManager class
			*/
			return node_path.default.resolve(root, output.path);
			if (group && (options$1?.group?.path || options$1?.group?.tag)) {
				const groupName = group?.name ? group.name : (ctx) => {
					if (group?.type === "path") return `${ctx.group.split("/")[1]}`;
					return `${(0, __kubb_core_transformers.camelCase)(ctx.group)}Controller`;
				};
				return node_path.default.resolve(root, output.path, groupName({ group: group.type === "path" ? options$1.group.path : options$1.group.tag }), baseName);
			}
			return node_path.default.resolve(root, output.path, baseName);
		},
		resolveName(name, type) {
			const resolvedName = (0, __kubb_core_transformers.camelCase)(name, { isFile: type === "file" });
			if (type) return transformers?.name?.(resolvedName, type) || resolvedName;
			return resolvedName;
		},
		async buildStart() {
			const [swaggerPlugin] = __kubb_core.PluginManager.getDependedPlugins(this.plugins, [__kubb_plugin_oas.pluginOasName]);
			const oas = await swaggerPlugin.context.getOas();
			const root = node_path.default.resolve(this.config.root, this.config.output.path);
			const mode = (0, __kubb_core.getMode)(node_path.default.resolve(root, output.path));
			const baseURL$1 = await swaggerPlugin.context.getBaseURL();
			const files = await new __kubb_plugin_oas.OperationGenerator(baseURL$1 ? {
				...this.plugin.options,
				baseURL: baseURL$1
			} : this.plugin.options, {
				oas,
				pluginManager: this.pluginManager,
				plugin: this.plugin,
				contentType,
				exclude,
				include,
				override,
				mode
			}).build(...generators);
			await this.addFile(...files);
			const barrelFiles = await (0, __kubb_core.getBarrelFiles)(this.fileManager.files, {
				type: output.barrelType ?? "named",
				root,
				output,
				meta: { pluginKey: this.plugin.key },
				logger: this.logger
			});
			await this.addFile(...barrelFiles);
		}
	};
});

//#endregion
exports.pluginClient = pluginClient;
exports.pluginClientName = pluginClientName;
