// @redocly/config/lib/constants/entities.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NODE_TYPE_NAMES = exports.ENTITY_NODE_TYPE_NAMES = exports.ENTITY_RELATION_TYPES = void 0;
const enum_1 = require("./enum");
exports.ENTITY_RELATION_TYPES = [
    'partOf',
    'hasParts',
    'creates',
    'createdBy',
    'owns',
    'ownedBy',
    'implements',
    'implementedBy',
    'dependsOn',
    'dependencyOf',
    'uses',
    'usedBy',
    'produces',
    'consumes',
    'linksTo',
    'supersedes',
    'supersededBy',
    'compatibleWith',
    'extends',
    'extendedBy',
    'relatesTo',
    'hasMember',
    'memberOf',
    'triggers',
    'triggeredBy',
    'returns',
    'returnedBy',
];
exports.ENTITY_NODE_TYPE_NAMES = (0, enum_1.createEnum)({
    UserEntity: 'UserEntity',
    UserEntityMetadata: 'UserEntityMetadata',
    ApiDescriptionEntity: 'ApiDescriptionEntity',
    ApiDescriptionEntityMetadata: 'ApiDescriptionEntityMetadata',
    ApiOperationEntity: 'ApiOperationEntity',
    ApiOperationEntityMetadata: 'ApiOperationEntityMetadata',
    DataSchemaEntity: 'DataSchemaEntity',
    DataSchemaEntityMetadata: 'DataSchemaEntityMetadata',
    ServiceEntity: 'ServiceEntity',
    DomainEntity: 'DomainEntity',
    TeamEntity: 'TeamEntity',
    Entity: 'Entity',
    EntityMetadata: 'EntityMetadata',
    EntityLinkList: 'EntityLinkList',
    EntityLink: 'EntityLink',
    EntityRelation: 'EntityRelation',
    EntityRelationList: 'EntityRelationList',
    EntityContact: 'EntityContact',
    EntitySlackContact: 'EntitySlackContact',
    EntitySlackChannel: 'EntitySlackChannel',
});
// FIXME: remove this once the config is updated in openapi-core and it's updated in the monorepo (https://github.com/Redocly/redocly-cli/pull/2604)
exports.NODE_TYPE_NAMES = exports.ENTITY_NODE_TYPE_NAMES;
