// @redocly/openapi-core/lib/rules/arazzo/sourceDescription-type.js
export const SourceDescriptionType = () => {
    return {
        SourceDescriptions: {
            enter(sourceDescriptions, { report, location, specVersion }) {
                if (!sourceDescriptions.length)
                    return;
                const allowedTypes = specVersion === 'arazzo1' ? ['openapi', 'arazzo'] : ['openapi', 'arazzo', 'asyncapi'];
                for (const sourceDescription of sourceDescriptions) {
                    if (!allowedTypes.includes(sourceDescription?.type)) {
                        report({
                            message: `The \`type\` property of the \`sourceDescription\` object must be either ${allowedTypes
                                .map((t) => `\`${t}\``)
                                .join(' or ')}.`,
                            location: location.child([sourceDescriptions.indexOf(sourceDescription)]),
                            reference: 'https://redocly.com/docs/cli/rules/arazzo/sourcedescription-type',
                        });
                    }
                }
            },
        },
    };
};
