// @redocly/openapi-core/lib/rules/arazzo/step-onFailure-unique.js
export const StepOnFailureUnique = () => {
    return {
        OnFailureActionList: {
            enter(onFailureActionList, { report, location }) {
                if (!onFailureActionList)
                    return;
                const seenFailureActions = new Set();
                for (const onFailureAction of onFailureActionList) {
                    if (seenFailureActions.has(onFailureAction?.name)) {
                        report({
                            message: 'The action `name` must be unique amongst listed `onFailure` actions.',
                            location: location.child([onFailureActionList.indexOf(onFailureAction)]),
                            reference: 'https://redocly.com/docs/cli/rules/arazzo/step-onfailure-unique',
                        });
                    }
                    if (seenFailureActions.has(onFailureAction?.reference)) {
                        report({
                            message: 'The action `reference` must be unique amongst listed `onFailure` actions.',
                            location: location.child([onFailureActionList.indexOf(onFailureAction)]),
                            reference: 'https://redocly.com/docs/cli/rules/arazzo/step-onfailure-unique',
                        });
                    }
                    seenFailureActions.add(onFailureAction.name ?? onFailureAction.reference);
                }
            },
        },
    };
};
